{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "149adffa",
   "metadata": {},
   "source": [
    "# **Authors:** Seth M. Woodbury (woodbuse@uw.edu) & Donghyo Kim (donghyo@uw.edu)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "971b750c",
   "metadata": {},
   "source": [
    "# **Organization & Notebook Setup/Initialization (Run Everytime)** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aef2ba95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### PROJECT zn_hydrolase NOTEBOOK SUCCESSFULLY INITIALIZED ON 2025-12-03 AT TIME 13:14:30 ### \n",
      "\n",
      "Example Variables & Paths:\n",
      "   working_dir   = /home/woodbuse/publication/metallohydrolase/github/Computational_Design_of_Metallohydrolases_PrivateGitHub/Metallohydrolase_Design_Pipeline_Tutorial/working_dir/\n",
      "   theozymes_dir = /home/woodbuse/publication/metallohydrolase/github/Computational_Design_of_Metallohydrolases_PrivateGitHub/Metallohydrolase_Design_Pipeline_Tutorial/inputs/theozymes/\n",
      "   output_dir    = /home/woodbuse/publication/metallohydrolase/github/Computational_Design_of_Metallohydrolases_PrivateGitHub/Metallohydrolase_Design_Pipeline_Tutorial/outputs/ \n",
      "\n"
     ]
    }
   ],
   "source": [
    "######################################################################\n",
    "### IMPORT PACKAGES & SETUP NOTEBOOK (RUN AT BEGINNING EVERY TIME) ###\n",
    "######################################################################\n",
    "\n",
    "project_name     = 'zn_hydrolase'\n",
    "\n",
    "### MANDATORY USER SPECIFICATIONS ###\n",
    "github_repo_dir  = '/home/woodbuse/publication/metallohydrolase/github/Computational_Design_of_Metallohydrolases_PrivateGitHub' # Cloned GitHub repo directory path\n",
    "\n",
    "### SUBFOLDERS/VARIABLES (OPTIONAL: ADD EXTRA TO LIST OF DEFAULTS) ###\n",
    "# 1.) subfolders to build in notebook directory\n",
    "wrk_dirs_list = [\n",
    "    'cmds', 'logs', 'slurm_submit', 'important_dfs', 'rfd2_configs', 'fixed_residue_jsonls',  # optionally add more input subdirectories\n",
    "]\n",
    "\n",
    "# 2.) subfolders to build in output directory\n",
    "out_dirs_list = [\n",
    "    'rfdiffusion2_out', 'predesign_out', 'inpainting_out', 'mpnn_out', 'af2_out', 'fastMPNNdesign',  \n",
    "    'placer_input_structures','placer_out', # optionally add more output subdirectories\n",
    "]\n",
    "\n",
    "###################################################### AUTO SETUP ######################################################\n",
    "### STANDARD LIBRARY IMPORTS (MAKE SURE TO USE THE SE3nv KERNEL) ###\n",
    "import shlex, glob, json, math, os, random, copy, re, shutil, statistics, string, subprocess, sys\n",
    "import textwrap, warnings, concurrent.futures, time, multiprocessing, itertools, operator\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "from collections import Counter\n",
    "\n",
    "### DERIVED PATHS (CHANGE AT USER DISCRETION) ###\n",
    "scripts_dir   = f'{github_repo_dir}/Scripts/'\n",
    "software_dir  = f'{github_repo_dir}/Software/'\n",
    "notebook_dir  = f'{github_repo_dir}/Metallohydrolase_Design_Pipeline_Tutorial/'\n",
    "working_dir   = f'{notebook_dir}working_dir/'\n",
    "output_dir    = f'{notebook_dir}outputs/'\n",
    "theozymes_dir = f'{notebook_dir}inputs/theozymes/' # theozymes from quantum chemistry\n",
    "invrot_dir    = f'{notebook_dir}inputs/invrot/'    # inverse rotamers turning quantum chemistry theozymes into PDBs while sampling different histidine ring flips and coordinations\n",
    "\n",
    "for p in [working_dir, output_dir, scripts_dir, software_dir]: \n",
    "    Path(p).mkdir(parents=True, exist_ok=True) # ensure base dirs exist so downstream cells never fail on missing paths\n",
    "\n",
    "### 3RD PARTY IMPORTS (MAKE SURE TO USE THE SE3nv KERNEL) ###\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import pairwise2\n",
    "from Bio.PDB import PDBParser, PPBuilder\n",
    "\n",
    "### CUSTOM IMPORTS ###\n",
    "if scripts_dir not in sys.path:\n",
    "    sys.path.append(scripts_dir)\n",
    "import General_NoteBook_Functions as functions\n",
    "\n",
    "### SETUP SUBDIRECTORIES & CREATE VARIABLES ###\n",
    "functions.setup_directories(working_dir, wrk_dirs_list, export_globals=True, globals_dict=globals())\n",
    "functions.setup_directories(output_dir,  out_dirs_list,  export_globals=True, globals_dict=globals())\n",
    "\n",
    "### OPTIONALS ###\n",
    "functions.set_pandas_display(all_on=True) # pandas display comfort defaults (toggle here if desired)\n",
    "os.chdir(working_dir) # move into working dir for relative IO\n",
    "\n",
    "### PRINTS ###\n",
    "print(f\"### PROJECT {project_name} NOTEBOOK SUCCESSFULLY INITIALIZED ON {datetime.now().strftime('%Y-%m-%d AT TIME %H:%M:%S')} ### \")\n",
    "print(f\"\\nExample Variables & Paths:\")\n",
    "print(f\"   working_dir   = {working_dir}\")\n",
    "print(f\"   theozymes_dir = {theozymes_dir}\")\n",
    "print(f\"   output_dir    = {output_dir}\", f\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dc2e672-df75-48f2-ab55-e6429b2dc8cb",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# **I. Generate Backbone Structures using RFD2**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5974f36-dde0-48cd-876a-778ea477e85a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Setup & Filter RFD2"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8fa64aee-9e86-4e94-83d8-ab7c6d1c52dc",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "To use the rf_diffusion repository in your Python environment, you need to add it to your PYTHONPATH.\n",
    "\n",
    "üìç Steps:\n",
    "1. Open a terminal.\n",
    "2. `cd` to the directory containing this notebook.\n",
    "3. Add the rf_diffusion repository (assumed to be located at ./../Software/RFdiffusion2) to your PYTHONPATH by running:\n",
    "\n",
    "$ export PYTHONPATH=\"$PYTHONPATH:$PWD/../Software/RFdiffusion2\"\n",
    "\n",
    "üí° This command temporarily sets the PYTHONPATH for your current terminal session.\n",
    "To make it permanent, you can add it to your shell config file (e.g., ~/.bashrc or ~/.zshrc)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d21b8bd6-87ac-4558-ba95-b9e6765d5966",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to run the script:\n",
      "python ./../Software/RFdiffusion2/rf_diffusion/run_inference.py --config-dir=./working_dir/rfd2_configs/ --config-name=PSZ_H_H_H_1_63.yaml \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###################\n",
    "### SET UP RFD2 ###\n",
    "###################\n",
    "\n",
    "### SET OUTPUT DIRECTORY FOR GENERATED .PDB STRUCTURES & INPUT MOTIFS + LIG NAME ###\n",
    "further_prefix = '' #want to add extra prefix before '.pdb' ? If not, write ''\n",
    "\n",
    "### SET UP THEOZYME FOR RFD2 ###\n",
    "input_motif_with_residues = [f'{invrot_dir}phenylacetate_4mu_HHH_ZnO_binding_S_enantio/PSZ_H_H_H_1_63.pdb']\n",
    "lig_name = \"PSZ\"\n",
    "\n",
    "### SET UP CONTIGS FOR RFD2 ###\n",
    "contigs = ['130-275,A1-1,B2-2,C3-3'] # ['min_length-max_length,chainXresidueY-chainXresidueY+n,...']\n",
    "contig_atoms = '{\"A1\":\"NE2,CD2,CG,CB,ND1,CE1\",\"B2\":\"NE2,CD2,CG,CB,ND1,CE1\",\"C3\":\"NE2,CD2,CG,CB,ND1,CE1\"}' #fixed tip atoms\n",
    "length_range = \"130-275\" # min-max length of protein, should match first range of contigs\n",
    "\n",
    "### ESSENTIAL RFD2 PARAMETERS ###\n",
    "timesteps = 75 # basically the smaller this number is, the bigger the denoising steps taken\n",
    "final_step = 1 # DECIDE IF 25-50, higher = less compute time but cuts number of timesteps\n",
    "\n",
    "nstruct = 10 # total outputs you want per topology\n",
    "\n",
    "### BONUS FEATURES TO TOGGLE ###\n",
    "guidepost_bonds = True\n",
    "align_motif = False\n",
    "str_self_cond = True\n",
    "contig_as_guidepost = True\n",
    "remove_guideposts_from_output = True\n",
    "infer_guidepost_positions = True\n",
    "guidepost_xyz_as_design = True\n",
    "flexible_ligand = False\n",
    "model_state_dict = 'model_state_dict'\n",
    "recenter_xt = True\n",
    "recenter_xt_dropt = '3'\n",
    "deterministic = False # MAKE THIS FALSE OR ALL OUTPUTS WILL BE SAME ONLY FOR BENCHMARKING\n",
    "\n",
    "### SET POTENTIALS? THIS MIGHT BE OUTDATED BUT CAN HELP WITH LIGAND CONTACT ###\n",
    "include_potentials = False  # Set to True if you want to include the potentials section\n",
    "potentials_list = ['ligand_ncontacts'] \n",
    "guide_decay = 'linear' #linear or quadratic, quadratic seems preferred these days\n",
    "guide_scale = 2\n",
    "weight = 0.65 #Previously 0.6\n",
    "\n",
    "### SET SCRIPT & CHECKPOINT ###\n",
    "script = './../Software/RFdiffusion2/rf_diffusion/run_inference.py' # SHOULD BE UPDATED\n",
    "checkpoint = './../Software/RFdiffusion2/rf_diffusion/model_weights/RFD_140.pt'\n",
    "\n",
    "### GENERATE OUR CONFIGS & WRITE OUR COMMANDS ###\n",
    "def generate_config_content(base_name, iteration, include_potentials):\n",
    "    # Start with the common configuration content\n",
    "    config_content = textwrap.dedent(f\"\"\"\n",
    "    defaults:\n",
    "      - base\n",
    "\n",
    "    diffuser:\n",
    "      T: {timesteps}\n",
    "      \n",
    "    guidepost_bonds: {guidepost_bonds}\n",
    "\n",
    "    inference:\n",
    "      align_motif: {align_motif}\n",
    "      ckpt_path: {checkpoint}\n",
    "      num_designs: {nstruct}\n",
    "      output_prefix: {rfdiffusion2_out_dir}{project_name}_{base_name}_{iteration}{further_prefix}\n",
    "      input_pdb: {os.path.join(os.path.dirname(input_motif_with_residues[0]), f\"{base_name}.pdb\")}\n",
    "      final_step: {final_step}\n",
    "      ligand: {lig_name}\n",
    "      contig_as_guidepost: {contig_as_guidepost}\n",
    "      remove_guideposts_from_output: {remove_guideposts_from_output}\n",
    "      infer_guidepost_positions: {infer_guidepost_positions}\n",
    "      guidepost_xyz_as_design: {guidepost_xyz_as_design}\n",
    "      flexible_ligand: {flexible_ligand}\n",
    "      state_dict_to_load: {model_state_dict}\n",
    "      recenter_xt: {recenter_xt}\n",
    "      deterministic: {deterministic}\n",
    "      recenter_xt_dropt: {recenter_xt_dropt}\n",
    "      idealize_sidechain_outputs: True\n",
    "\n",
    "    contigmap:\n",
    "      contigs: {contigs}\n",
    "      length: '{length_range}'\n",
    "      contig_atoms: '{contig_atoms}'\n",
    "    \"\"\")\n",
    "\n",
    "    # Add the potentials section only if potentials are defined\n",
    "    if include_potentials == True:\n",
    "        potentials_section = textwrap.dedent(f\"\"\"\n",
    "        potentials:\n",
    "          guiding_potentials: [\"type:{potentials_list[0]},weight:{weight}\"]\n",
    "          guide_scale: {guide_scale}\n",
    "          guide_decay: {guide_decay}\n",
    "        \"\"\")\n",
    "        config_content += potentials_section\n",
    "    return config_content\n",
    "\n",
    "# Make sure to use the correct variable (include_potentials) when calling generate_config_content\n",
    "print (\"Command to run the script:\")\n",
    "for pdb in input_motif_with_residues:\n",
    "    base_name = os.path.basename(pdb).replace('.pdb', '')\n",
    "    config_file = f'{rfd2_configs_dir}{base_name}.yaml'\n",
    "    with open(config_file, 'w') as config_f:\n",
    "        config_txt = generate_config_content(base_name, 0, include_potentials)\n",
    "        config_f.write(config_txt)\n",
    "\n",
    "    cmd = f'python {script} --config-dir={rfd2_configs_dir} --config-name={base_name}.yaml \\n'\n",
    "    print(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2d60116e-506c-4930-9b46-2fbd979c2d56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to run the script:\n",
      "python {scripts_dir}process_RFD2_outputs.py --pdb ./output/rfdiffusion2_out/*PSZ*.pdb --ref ./input/invrot/COMBINED_PHENYLACETATE_HHH/*PSZ*.pdb --outdir ./output/rfdiffusion2_out/ --SASA_lower_limit 0 --loop_limit 0.3 --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --rog 32.0 --longest_helix 32 --bondlen_dev 0.3 --loop_catres --fix \n",
      "mv ./output/rfdiffusion2_out/RFD2_analysis.sc ./output/rfdiffusion2_out/RFD2_analysis_PSZ.sc\n",
      "\n"
     ]
    }
   ],
   "source": [
    "################################\n",
    "### RUN PROCESS RFD2 OUTPUTS ###\n",
    "################################\n",
    "\n",
    "# Set the multiple ligand names\n",
    "multiple_ligand_names = {\"PSZ\"}\n",
    "\n",
    "### INPUT STRUCTURE DIRECTORIES ###\n",
    "combined_input_ligands_DIR = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/\"\n",
    "params_path_DIR = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/\"\n",
    "\n",
    "### OUTPUT DIRECTORIES ###\n",
    "outdir = \"./output/rfdiffusion2_out/\"\n",
    "\n",
    "### INPUT VARIABLES ###\n",
    "SASA_lower_limit = \"0\"\n",
    "loop_limit = \"0.3\"\n",
    "rog = \"32.0\"\n",
    "longest_helix = \"32\"\n",
    "bondlen_dev = 0.3\n",
    "\n",
    "loop_catres = \"--loop_catres\" # Default = True, if enabled structs with cat residues on loopy res are filtered\n",
    "fix = \"--fix\"\n",
    "\n",
    "# Mostly Constant\n",
    "script = f\"{scripts_dir}process_RFD2_outputs.py\"\n",
    "nproc = 19 # Number of CPUs to use\n",
    "\n",
    "# Iterate over each ligand name to generate and print the command\n",
    "for ligand_name in multiple_ligand_names:\n",
    "    if len(glob.glob(f\"{rfdiffusion2_out_dir}/*{ligand_name}*.pdb\")) == 0:\n",
    "        continue\n",
    "    ### INPUT STRUCTURES ###\n",
    "    pdb_path = f\"{rfdiffusion2_out_dir}*{ligand_name}*.pdb\"\n",
    "    ref_path = f\"{combined_input_ligands_DIR}*{ligand_name}*.pdb\"\n",
    "    params_path = f\"{params_path_DIR}{ligand_name}.params\"\n",
    "    \n",
    "    # Compile the command for each ligand\n",
    "    command = (f\"python {script} \"\n",
    "               f\"--pdb {pdb_path} \"\n",
    "               f\"--ref {ref_path} \"\n",
    "               f\"--outdir {outdir} \"\n",
    "               f\"--SASA_lower_limit {SASA_lower_limit} \"\n",
    "               f\"--loop_limit {loop_limit} \"\n",
    "               f\"--params {params_path} \"\n",
    "               f\"--rog {rog} \"\n",
    "               f\"--longest_helix {longest_helix} \"\n",
    "               f\"--bondlen_dev {bondlen_dev} \"\n",
    "               f\"{loop_catres} \"\n",
    "               f\"{fix} \"\n",
    "              )\n",
    "    \n",
    "    # Print the command for the current ligand\n",
    "    print (\"Command to run the script:\")\n",
    "    print(command)\n",
    "    print(f\"mv {outdir}RFD2_analysis.sc {outdir}RFD2_analysis_{ligand_name}.sc\")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5acf29be-9608-48c1-be1e-246d9d8ea82f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Fix Side Chains Using Predesign on RFD2 Backbones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0f1127e2-8403-4949-9198-0b4493473740",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to run the script:\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/rfdiffusion2_out/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/filtered_RFD2 --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/rfdiffusion2_out/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/filtered_RFD2 --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/rfdiffusion2_out/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/filtered_RFD2 --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/rfdiffusion2_out/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_8-atomized-bb-False.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/filtered_RFD2 --scoring {scripts_dir}scoring/esterase_scoring.py\n"
     ]
    }
   ],
   "source": [
    "#############################################################\n",
    "### RUN PREDESIGN SCRIPT TO FIX SIDE CHAINS & MINIMIZE BB ###\n",
    "#############################################################\n",
    "\n",
    "# NOTE: YOU NEED TO RUN THIS IN A SEPARATE DIRECTORY\n",
    "\n",
    "### INPUTS ###\n",
    "commands_name = \"predesign_commands_after_filtered_rfd2\"\n",
    "ligand_names = [\"PSZ\"]\n",
    "\n",
    "specific_predesign_output_dir = f'{predesign_out_dir}filtered_RFD2'\n",
    "os.makedirs(specific_predesign_output_dir, exist_ok=True)\n",
    "\n",
    "input_pdb_structures_dir = os.path.join(rfdiffusion2_out_dir, \"filtered_structures/\")\n",
    "params_files = {ligand: glob.glob(os.path.join(theozymes_dir, \"phenylacetate_4MU_substrate/HHH_to_zinc/\", # Narrow it down\n",
    "                                               f\"**/*{ligand}*.params\"), recursive=True)[0]\n",
    "                for ligand in ligand_names}\n",
    "\n",
    "cst_files = {\n",
    "    ligand: os.path.join(theozymes_dir, \"cst_files\", f\"{ligand}_3H_NO_DEPROT_RES.cst\")\n",
    "    for ligand in ligand_names}\n",
    "\n",
    "### Constants ###\n",
    "script_path = f\"{scripts_dir}predesign_FMD.py\"\n",
    "scoring_path = f\"{scripts_dir}scoring/esterase_scoring.py\"\n",
    "\n",
    "### Generate Commands ###\n",
    "print (\"Command to run the script:\")\n",
    "for ligand_name in ligand_names:\n",
    "    pdb_files = glob.glob(os.path.join(input_pdb_structures_dir, f\"*{ligand_name}*.pdb\"))\n",
    "    params_file = params_files[ligand_name]\n",
    "    cst_file = cst_files[ligand_name]\n",
    "\n",
    "    for pdb_file in pdb_files:\n",
    "        cmd = (\n",
    "            f\"python {script_path} --pdb {pdb_file} \"\n",
    "            f\"--params {params_file} --cstfile {cst_file} --output_dir {specific_predesign_output_dir} --scoring {scoring_path}\"\n",
    "        )\n",
    "        print (cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c4c5dab3-f7e1-496c-8512-b8a057a8a77c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The real path of the combined file is: /home/donghyo/git/de_novo_metalloenzyme_design/output/predesign_out/filtered_RFD2/scores/cumulative_predesign_analysis_scores.sc\n"
     ]
    }
   ],
   "source": [
    "###########################################\n",
    "### COMBINE ALL PREDESIGN SCORE OUTPUTS ###\n",
    "###########################################\n",
    "\n",
    "### INPUT & OUTPUT ###\n",
    "scores_dir = os.path.join(predesign_out_dir, 'filtered_RFD2/scores')\n",
    "output_file = os.path.join(scores_dir, \"cumulative_predesign_analysis_scores.sc\")\n",
    "\n",
    "sc_files = [os.path.join(scores_dir, f) for f in os.listdir(scores_dir) if f.endswith('.sc')]\n",
    "if not sc_files:\n",
    "    raise ValueError(\"No .sc files found in the current directory.\")\n",
    "\n",
    "data_frames = []\n",
    "for file in sc_files:\n",
    "    if os.path.basename(file) == \"cumulative_predesign_analysis_scores.sc\": continue\n",
    "    df = pd.read_csv(file, sep=r'\\s+')\n",
    "    data_frames.append(df)\n",
    "    \n",
    "combined_df = pd.concat(data_frames, ignore_index=True)\n",
    "combined_df.to_csv(output_file, sep=' ', index=False)\n",
    "\n",
    "# Print the real path of the output file\n",
    "real_path = os.path.realpath(output_file)\n",
    "print(f\"The real path of the combined file is: {real_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6022971a-f7f4-4741-b0f6-8b7060be50b9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Inpainting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79ed658-15e4-4e11-a064-3bc224fe5d8b",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Setup Inpainting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "54dae2cb-2a39-489f-a3b9-4ff6cef845a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copy .trb files to predesign folder:\n",
      "cp ./output/rfdiffusion2_out/filtered_structures/*.trb ./output/predesign_out/filtered_RFD2/\n",
      "\n",
      "Change the name of .trb files:\n",
      "for f in ./output/predesign_out/filtered_RFD2/*.trb ; do mv $f ${f/.trb/_0.trb} ; done\n"
     ]
    }
   ],
   "source": [
    "### COPY ALL .trb TO PREDESIGN FOLDER (OR WHEREVER) ###\n",
    "print(\"Copy .trb files to predesign folder:\")\n",
    "print(\"cp ./output/rfdiffusion2_out/filtered_structures/*.trb ./output/predesign_out/filtered_RFD2/\")\n",
    "print()\n",
    "print(\"Change the name of .trb files:\")\n",
    "print(\"for f in ./output/predesign_out/filtered_RFD2/*.trb ; do mv $f ${f/.trb/_0.trb} ; done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bd4d7e6b-6a6b-49b3-8270-6e806cebffc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to run the script:\n",
      "python {scripts_dir}setup_inpaint_from_rfd2.py --pdb ./output/predesign_out/filtered_RFD2/*PSZ*.pdb --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --group 10 --var --ref_catres A1 B2 C3 --trb ./output/predesign_out/filtered_RFD2/*PSZ*.trb --nstruct 3 --outdir ./output/inpainting_out/\n",
      "\n",
      "Command to move the .json file:\n",
      "for f in *.json ; do mv $f ./working_dir/cmds/${f/cmds_/cmdsPSZ_} ; done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##################################################\n",
    "### GENERATE .JSON FILES TO RUN FOR INPAINTING ###\n",
    "##################################################\n",
    "\n",
    "# Set the multiple ligand names\n",
    "multiple_ligand_names = {\"PSZ\"}\n",
    "\n",
    "### INPUT VARIABLES ###\n",
    "group = \"10\"  # Define your group value here\n",
    "var_flag = \"--var\"  # Enabling variable contig lengths\n",
    "ref_catres = \"A1 B2 C3\"  # Example format for catalytic residues\n",
    "nstruct = 3 # Minimum num of structures per one\n",
    "\n",
    "# Paths (adjust these to your actual paths)\n",
    "rfd2_bbs_to_analyze_DIR = f\"{predesign_out_dir}filtered_RFD2/\" # MAKE SURE FILTERED\n",
    "trb_dir = os.path.join(predesign_out_dir, \"filtered_RFD2/\") # IF not in pdb dir\n",
    "params_path_DIR = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/\"\n",
    "\n",
    "# Constant Variables\n",
    "script = f\"{scripts_dir}setup_inpaint_from_rfd2.py\"\n",
    "\n",
    "# Iterate over each ligand name to generate and print the command\n",
    "print (\"Command to run the script:\")\n",
    "for ligand_name in multiple_ligand_names:\n",
    "    ### INPUT STRUCTURES ###\n",
    "    if len(glob.glob(os.path.join(rfd2_bbs_to_analyze_DIR, f\"*{ligand_name}*.pdb\"))) == 0:\n",
    "        continue\n",
    "    pdb_path = os.path.join(rfd2_bbs_to_analyze_DIR, f\"*{ligand_name}*.pdb\")\n",
    "    trb_paths = os.path.join(trb_dir, f\"*{ligand_name}*.trb\")\n",
    "    params_path = os.path.join(params_path_DIR, f\"{ligand_name}.params\")\n",
    "    \n",
    "    # Compile the command for each ligand\n",
    "    command = (f\"python {script} \"\n",
    "               f\"--pdb {pdb_path} \"\n",
    "               f\"--params {params_path} \"\n",
    "               f\"--group {group} \"\n",
    "               f\"{var_flag} \"\n",
    "               f\"--ref_catres {ref_catres} \"\n",
    "               f\"--trb {trb_paths} \"\n",
    "               f\"--nstruct {nstruct} \"\n",
    "               f\"--outdir {inpainting_out_dir}\"\n",
    "              )\n",
    "\n",
    "    # Print the command for the current ligand\n",
    "    print(command)\n",
    "    print()\n",
    "    print(\"Command to move the .json file:\")\n",
    "    print(f\"for f in *.json ; do mv $f ./working_dir/cmds/${{f/cmds_/cmds{ligand_name}_}} ; done\")\n",
    "    print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad780c2f-51c4-41f8-979f-33a2d16c1d98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to run the script:\n",
      "python {software_dir}proteininpainting/inpaint.py --input_json ./working_dir/cmds/cmdsPSZ_0.json\n"
     ]
    }
   ],
   "source": [
    "###########################################\n",
    "### MAKE INPAINTING COMMANDS AND SUBMIT ###\n",
    "###########################################\n",
    "\n",
    "# Constants\n",
    "inpaint_script = f\"{software_dir}proteininpainting/inpaint.py\"\n",
    "## Check this works. I changed the path of checkpoint. If so, remove line 66.\n",
    "\n",
    "# Collect all JSON files in the inpainting output directory\n",
    "json_files = [f for f in glob.glob(\"./working_dir/cmds/*.json\")]\n",
    "\n",
    "# Generate commands for each JSON file\n",
    "print (\"Command to run the script:\")\n",
    "for json_file in json_files:\n",
    "    command = f\"python {inpaint_script} --input_json {json_file}\"\n",
    "    print (command)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85306dbf-37d5-4a3a-92ec-873e68878e98",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Align Input Outputs with RFD2 Outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "235da420-015d-4017-a3e9-e519f4f75c59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to run the script:\n",
      "python {scripts_dir}enzyme_design/align_dif_inpaint_add_ligand.py --pdb ./output/inpainting_out/*PSZ*.pdb --ref ./output/rfdiffusion2_out/filtered_structures/*PSZ*.pdb --outdir ./output/inpainting_out/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##############################################\n",
    "### GENERATE COMMANDS FOR ALIGNING LIGANDS ###\n",
    "##############################################\n",
    "\n",
    "# Set the multiple ligand names\n",
    "multiple_ligand_names = {\"PSZ\"}\n",
    "\n",
    "### INPUT STRUCTURE DIRECTORIES ###\n",
    "filtered_rfd2_bbs_as_ref_DIR = f\"{predesign_out_dir}filtered_RFD2/\"\n",
    "\n",
    "combined_input_ligands_DIR = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/\"\n",
    "params_path_DIR = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/\" #MAKE SURE params .pdb here too\n",
    "\n",
    "aligning_inpainted_str_output_dir = f\"{inpainting_out_dir}alignment/\"\n",
    "os.makedirs(aligning_inpainted_str_output_dir, exist_ok=True)\n",
    "\n",
    "# Script for alignment\n",
    "align_script = f\"{scripts_dir}enzyme_design/align_dif_inpaint_add_ligand.py\"\n",
    "\n",
    "# Iterate over each ligand name to generate and print the command\n",
    "print (\"Command to run the script:\")\n",
    "for ligand_name in multiple_ligand_names:\n",
    "    if len(glob.glob(os.path.join(inpainting_out_dir, f\"*{ligand_name}*.pdb\"))) == 0:\n",
    "        continue\n",
    "        \n",
    "    ### INPUT STRUCTURES ###\n",
    "    pdb_path = os.path.join(inpainting_out_dir, f\"*{ligand_name}*.pdb\")\n",
    "    ref_path = os.path.join(rfdiffusion2_out_dir, \"filtered_structures\", f\"*{ligand_name}*.pdb\") #FILTERED DIFFUSION BBs NOW\n",
    "    params_path = os.path.join(params_path_DIR, f\"{ligand_name}.params\")\n",
    "    \n",
    "    # Compile and print the command for alignment\n",
    "    command_align = (f\"python {align_script} \"\n",
    "                     f\"--pdb {pdb_path} \"\n",
    "                     f\"--ref {ref_path} \"\n",
    "                     f\"--outdir {aligning_inpainted_str_output_dir} \"\n",
    "                     f\"--params {params_path}\")\n",
    "    print(command_align)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd1179e-1e40-4053-8c86-4c439a10c452",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Fix Side Chains Using Predesign on Inpainting Backbones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b969d259-d043-4a88-bb9c-31e361a2eb64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to copy all the .trb files:\n",
      "cp ./output/inpainting_out/*.trb ./output/inpainting_out/alignment\n"
     ]
    }
   ],
   "source": [
    "# Copy all trb files into the alignment folder\n",
    "print(\"Command to copy all the .trb files:\")\n",
    "print(f\"cp {inpainting_out_dir}*.trb {inpainting_out_dir}alignment\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8da570c2-c70e-46ff-9462-d220cae9adf2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to run the script:\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False_0_inp_0.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False_0_inp_1.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False_0_inp_2.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False_0_inp_3.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False_0_inp_4.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_inp_0.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_inp_1.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_inp_2.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_inp_3.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_inp_0.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_inp_1.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_inp_2.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_inp_3.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_8-atomized-bb-False_0_inp_0.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_8-atomized-bb-False_0_inp_1.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_8-atomized-bb-False_0_inp_2.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n",
      "python {scripts_dir}predesign_FMD.py --pdb ./output/inpainting_out/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_8-atomized-bb-False_0_inp_3.pdb --params ./input/theozymes/phenylacetate_4MU_substrate/HHH_to_zinc/ZnO_binding_int/S_enantio/PSZ.params --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --output_dir ./output/predesign_out/inpainting_outputs --scoring {scripts_dir}scoring/esterase_scoring.py\n"
     ]
    }
   ],
   "source": [
    "#############################################################\n",
    "### RUN PREDESIGN SCRIPT TO FIX SIDE CHAINS & MINIMIZE BB ###\n",
    "#############################################################\n",
    "\n",
    "# NOTE: YOU NEED TO RUN THIS IN A SEPARATE DIRECTORY\n",
    "\n",
    "### INPUTS ###\n",
    "commands_name = \"predesign_commands_after_inpainting\"\n",
    "ligand_names = [\"PSZ\"]\n",
    "\n",
    "specific_predesign_output_dir = f'{predesign_out_dir}inpainting_outputs'\n",
    "os.makedirs(specific_predesign_output_dir, exist_ok=True)\n",
    "\n",
    "params_files = {ligand: glob.glob(os.path.join(theozymes_dir, \"phenylacetate_4MU_substrate/HHH_to_zinc/\", # Narrow it down\n",
    "                                               f\"**/*{ligand}*.params\"), recursive=True)[0]\n",
    "                for ligand in ligand_names}\n",
    "\n",
    "cst_files = {\n",
    "    ligand: os.path.join(theozymes_dir, \"cst_files\", f\"{ligand}_3H_NO_DEPROT_RES.cst\")\n",
    "    for ligand in ligand_names}\n",
    "\n",
    "### Constants ###\n",
    "script_path = f\"{scripts_dir}predesign_FMD.py\"\n",
    "scoring_path = f\"{scripts_dir}scoring/esterase_scoring.py\"\n",
    "\n",
    "### Generate Commands ###\n",
    "print (\"Command to run the script:\")\n",
    "for ligand_name in ligand_names:\n",
    "    pdb_files = glob.glob(os.path.join(inpainting_out_dir, \"alignment\", f\"*{ligand_name}*.pdb\"))\n",
    "    params_file = params_files[ligand_name]\n",
    "    cst_file = cst_files[ligand_name]\n",
    "\n",
    "    for pdb_file in pdb_files:\n",
    "        cmd = (\n",
    "            f\"python {script_path} --pdb {pdb_file} \"\n",
    "            f\"--params {params_file} --cstfile {cst_file} --output_dir {specific_predesign_output_dir} --scoring {scoring_path}\"\n",
    "        )\n",
    "        print (cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "58367d88-1726-47b3-a29f-dd825dbb5229",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The real path of the combined file is: /home/donghyo/software/Computational_Design_of_Metallohydrolases_PrivateGitHub/Metallohydrolase_Design_Pipeline_Tutorial/output/predesign_out/inpainting_outputs/scores/cumulative_predesign_analysis_scores.sc\n"
     ]
    }
   ],
   "source": [
    "###########################################\n",
    "### COMBINE ALL PREDESIGN SCORE OUTPUTS ###\n",
    "###########################################\n",
    "\n",
    "### INPUT & OUTPUT ###\n",
    "scores_dir = os.path.join(predesign_out_dir, 'inpainting_outputs/scores')\n",
    "output_file = os.path.join(scores_dir, \"cumulative_predesign_analysis_scores.sc\")\n",
    "\n",
    "sc_files = [os.path.join(scores_dir, f) for f in os.listdir(scores_dir) if f.endswith('.sc')]\n",
    "if not sc_files:\n",
    "    raise ValueError(\"No .sc files found in the current directory.\")\n",
    "\n",
    "data_frames = []\n",
    "for file in sc_files:\n",
    "    if os.path.basename(file) == \"cumulative_predesign_analysis_scores.sc\": continue\n",
    "    df = pd.read_csv(file, sep=r'\\s+')\n",
    "    data_frames.append(df)\n",
    "    \n",
    "combined_df = pd.concat(data_frames, ignore_index=True)\n",
    "combined_df.to_csv(output_file, sep=' ', index=False)\n",
    "\n",
    "# Print the real path of the output file\n",
    "real_path = os.path.realpath(output_file)\n",
    "print(f\"The real path of the combined file is: {real_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0c02818-7ced-41ed-8b59-e8f9a3c9c35b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Filter on All Predesigned Stuctures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d8fbe2e1-5c9e-429f-a9da-3f7e6240b68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame RFD2_predesign_df loaded with 4 rows and 44 columns.\n",
      "\n",
      "DataFrame inpainting_predesign_df loaded with 17 rows and 44 columns.\n",
      "\n",
      "**2 of 4 (50.0%) passed in filtered_RFD2_predesign_df**\n",
      "**4 of 17 (23.529%) passed in filtered_inpainting_predesign_df**\n"
     ]
    }
   ],
   "source": [
    "##################################################################\n",
    "### FILTER THE DATA FRAME BASED ON WHATEVER CRITERIA ARE IDEAL ###\n",
    "##################################################################\n",
    "\n",
    "### HARD FILTERS ###\n",
    "score_per_res = 0\n",
    "all_cst = 7.25 # 7.25\n",
    "\n",
    "### BOUNDARY FILTERS ###\n",
    "L_SASA_upper = 0.5  # relative for ligand only (4MU)\n",
    "L_SASA_lower = 0  # relative for ligand only (4MU)\n",
    "substrate_SASA_upper = 1000 \n",
    "substrate_SASA_lower = 0\n",
    "\n",
    "subdirectories = ['filtered_RFD2', \n",
    "                  'inpainting_outputs']\n",
    "\n",
    "dataframe_names = ['RFD2_predesign_df', \n",
    "                   'inpainting_predesign_df']\n",
    "\n",
    "### AUTOMATED STUFF ###\n",
    "def find_cumulative_files(directory, filename=\"cumulative_predesign_analysis_scores.sc\"):\n",
    "    \"\"\"\n",
    "    Recursively search for files with a specified name within a directory \n",
    "    and its subdirectories using glob.\n",
    "    \"\"\"\n",
    "    file_pattern = os.path.join(directory, \"**\", filename)  # Recursive pattern\n",
    "    return glob.glob(file_pattern, recursive=True)\n",
    "\n",
    "dataframes = {}\n",
    "for subdir, df_name in zip(subdirectories, dataframe_names):\n",
    "    full_path = os.path.join(predesign_out_dir, subdir)\n",
    "    found_files = find_cumulative_files(full_path)\n",
    "    if found_files:\n",
    "        df = pd.read_csv(found_files[0], delim_whitespace=True, comment='#')\n",
    "        dataframes[df_name] = df\n",
    "        globals()[df_name] = df  # Assign DataFrame to a global variable\n",
    "        print(f\"DataFrame {df_name} loaded with {df.shape[0]} rows and {df.shape[1]} columns.\")\n",
    "        print('')\n",
    "\n",
    "            \n",
    "dataframe_dict = {\n",
    "    'RFD2_predesign_df': RFD2_predesign_df,\n",
    "    'inpainting_predesign_df': inpainting_predesign_df\n",
    "}\n",
    "\n",
    "\n",
    "### FILTER AND DISPLAY RESULTS FOR EACH DATA FRAME ###\n",
    "filtered_dataframes = {}\n",
    "\n",
    "for df_name, df in dataframe_dict.items():\n",
    "    filtered_df_name = 'filtered_' + df_name\n",
    "    filter_mask = ((df['all_cst'] <= all_cst) & \n",
    "                   (df['score_per_res'] <= score_per_res) & \n",
    "                   (df['L_SASA'] <= L_SASA_upper) & \n",
    "                   (df['L_SASA'] >= L_SASA_lower) & \n",
    "                   (df['substrate_SASA'] <= substrate_SASA_upper) & \n",
    "                   (df['substrate_SASA'] >= substrate_SASA_lower))\n",
    "    \n",
    "    filtered_df = df[filter_mask].copy()\n",
    "    globals()[filtered_df_name] = filtered_df\n",
    "    filtered_dataframes[filtered_df_name] = filtered_df\n",
    "    filter_rate = float(len(filtered_df)) / float(len(df)) * 100.0\n",
    "\n",
    "    # Output results\n",
    "    print(f'**{len(filtered_df)} of {len(df)} ({round(filter_rate, 3)}%) passed in {filtered_df_name}**')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "198f4fc5-154c-464a-9905-23549068fec4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered_RFD2_predesign_df\n",
      "Number of PDB files to copy: 2\n",
      "./output/predesign_out/filtered_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb\n",
      "[Mon Apr 21 17:26:43 2025] 0th PDB file was copied...\n",
      "./output/predesign_out/filtered_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0.pdb\n",
      "Done\n",
      "filtered_inpainting_predesign_df\n",
      "Number of PDB files to copy: 4\n",
      "./output/predesign_out/inpainting_outputs/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False_0_inp_2_0.pdb\n",
      "[Mon Apr 21 17:26:43 2025] 0th PDB file was copied...\n",
      "./output/predesign_out/inpainting_outputs/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False_0_inp_3_0.pdb\n",
      "./output/predesign_out/inpainting_outputs/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_0-atomized-bb-False_0_inp_4_0.pdb\n",
      "./output/predesign_out/inpainting_outputs/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_inp_3_0.pdb\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "###########################################\n",
    "### COPY FILTERED PREDESIGNED BACKBONES ###\n",
    "###########################################\n",
    "\n",
    "for i, (df_name, filtered_df) in enumerate(filtered_dataframes.items()):\n",
    "    print (df_name)\n",
    "    filtered_predesign_path = os.path.join(predesign_out_dir, subdirectories[i], \"filtered_structures\")\n",
    "    if not os.path.isdir(filtered_predesign_path):\n",
    "        os.mkdir(filtered_predesign_path)\n",
    "    \n",
    "    print (f\"Number of PDB files to copy: {len(filtered_df)}\")\n",
    "    for i, desc in enumerate(filtered_df[\"description\"]):\n",
    "        shutil.copy(desc+\".pdb\", os.path.join(filtered_predesign_path, os.path.basename(desc)+\".pdb\"))\n",
    "        print(os.path.join(filtered_predesign_path, os.path.basename(desc)+\".pdb\"))\n",
    "        if i % 1000 == 0:\n",
    "            print (f\"[{time.ctime()}] {i}th PDB file was copied...\")\n",
    "    print (\"Done\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "905e0ab1-5104-4dc0-9317-6c3de877ab3c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# **II. ProteinMPNN to Generate Sequences for AF2**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c8f0769-3d2b-4bb6-935c-a223725e36e3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Run ProteinMPNN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8699cc39-69d2-486a-a37a-a064d33b21c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to run the script:\n",
      "python {scripts_dir}grab_cat_residues_from_pdb_and_write_fixedAA_jsonl.py --input_dir ./output/predesign_out/filtered_RFD2/filtered_structures/ --output_jsonl ./working_dir/fixed_residue_jsonls/filtered_RFD2.jsonl\n"
     ]
    }
   ],
   "source": [
    "##################################################################\n",
    "### GRAB CATALYTIC RESIDUES AND MAKE FIXED POS .JSONL FOR MPNN ###\n",
    "##################################################################\n",
    "\n",
    "### INPUTS FOR THE FUNCTION ###\n",
    "input_pdb_directory =  f\"{predesign_out_dir}filtered_RFD2/filtered_structures/\"\n",
    "# num 1: f\"{predesign_out_dir}filtered_RFD2/filtered_structures/\"\n",
    "# num 2: f\"{predesign_out_dir}inpainting_outputs/filtered_structures/\"\n",
    "output_jsonl = f\"{fixed_residue_jsonls_dir}filtered_RFD2.jsonl\"\n",
    "# num 1: f\"{fixed_residue_jsonls_dir}filtered_RFD2.jsonl\"\n",
    "# num 2: f\"{fixed_residue_jsonls_dir}inpainting.jsonl\"\n",
    "\n",
    "# Constants\n",
    "script_path = f\"{scripts_dir}grab_cat_residues_from_pdb_and_write_fixedAA_jsonl.py\"\n",
    "\n",
    "# Constructing the command\n",
    "command = (f\"python {script_path} --input_dir {input_pdb_directory} \"\n",
    "           f\"--output_jsonl {output_jsonl}\"\n",
    "          )\n",
    "# Print the command\n",
    "print(\"Command to run the script:\")\n",
    "print(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "23986085-ad11-4a40-9a0f-fb8fa6aa99db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python ./../Software/fastmpnndesign/lib/LigandMPNN/run.py     --model_type protein_mpnn     --omit_AA MCX     --pdb_path ./output/predesign_out/filtered_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb     --out_folder ./output/mpnn_out/predesigned_RFD2     --number_of_batches 10     --temperature 0.1     --batch_size 1     --pack_side_chains 0     --fixed_residues_multi ./working_dir/fixed_residue_jsonls/filtered_RFD2.jsonl     --sc_num_denoising_steps 6     --repack_everything 0     --packed_suffix _packed_tempX10_1_     --bias_AA A:-0.75     --save_stats 0     --checkpoint_protein_mpnn ./../Software/fastmpnndesign/lib/LigandMPNN/model_params/proteinmpnn_v_48_020.pt     --checkpoint_ligand_mpnn ./../Software/fastmpnndesign/lib/LigandMPNN/model_params/ligandmpnn_v_32_010_25.pt     --checkpoint_path_sc ./../Software/fastmpnndesign/lib/LigandMPNN/model_params/ligandmpnn_sc_v_32_002_16.pt\n",
      "python ./../Software/fastmpnndesign/lib/LigandMPNN/run.py     --model_type protein_mpnn     --omit_AA MCX     --pdb_path ./output/predesign_out/filtered_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0.pdb     --out_folder ./output/mpnn_out/predesigned_RFD2     --number_of_batches 10     --temperature 0.1     --batch_size 1     --pack_side_chains 0     --fixed_residues_multi ./working_dir/fixed_residue_jsonls/filtered_RFD2.jsonl     --sc_num_denoising_steps 6     --repack_everything 0     --packed_suffix _packed_tempX10_1_     --bias_AA A:-0.75     --save_stats 0     --checkpoint_protein_mpnn ./../Software/fastmpnndesign/lib/LigandMPNN/model_params/proteinmpnn_v_48_020.pt     --checkpoint_ligand_mpnn ./../Software/fastmpnndesign/lib/LigandMPNN/model_params/ligandmpnn_v_32_010_25.pt     --checkpoint_path_sc ./../Software/fastmpnndesign/lib/LigandMPNN/model_params/ligandmpnn_sc_v_32_002_16.pt\n"
     ]
    }
   ],
   "source": [
    "####################################################################\n",
    "### RUN MPNN FOR GENERATION X OF STRUCTURES | X = NATURAL NUMBER ###\n",
    "####################################################################\n",
    "\n",
    "filtered_RFD2_predesign_df['pdb_path'] = filtered_RFD2_predesign_df['description'] + '.pdb'\n",
    "filtered_inpainting_predesign_df['pdb_path'] = filtered_inpainting_predesign_df['description'] + '.pdb'\n",
    "\n",
    "### INPUT STRUCTURES FOR FILTERED DF ###\n",
    "input_mpnn_pdb_FILTERED_DF = filtered_RFD2_predesign_df\n",
    "# num 1: \"filtered_RFD2_predesign_df\n",
    "# num 2: \"filtered_inpainting_predesign_df\n",
    "\n",
    "pdb_path_column_name = 'pdb_path'\n",
    "\n",
    "### INPUT .jsonl ###\n",
    "fixed_res_jsonl_basename = \"filtered_RFD2.jsonl\" #CHANGE EVERY TIME\n",
    "# num 1: \"filtered_RFD2.jsonl\n",
    "# num 2: \"inpainting.jsonl\n",
    "formatted_fixed_AA_jsonl = os.path.join(fixed_residue_jsonls_dir, fixed_res_jsonl_basename)\n",
    "\n",
    "# options for MPNN sequence design\n",
    "nstruct = 10              # num1: 10, num2: 5, num3: 5\n",
    "fused_mpnn_batch_size = 1\n",
    "temp = 0.1                # num1: 0.1, num2: 0.3, num3: 0.6\n",
    "omit_aa = 'MCX'\n",
    "pack = 0\n",
    "\n",
    "# Where to write commands & output mpnn (mostly unchanged now)\n",
    "updated_mpnn_out_dir = os.path.join(mpnn_out_dir, \"predesigned_RFD2\")\n",
    "# num 1: \"predesigned_RFD2\n",
    "# num 2: \"predesigned_inpainting\n",
    "os.makedirs(updated_mpnn_out_dir, exist_ok=True)\n",
    "\n",
    "# Parameters for fused_mpnn\n",
    "packed_suffix = f'_packed_tempX10_{int(temp*10)}_'\n",
    "mpnn_design_script = './../Software/fastmpnndesign/lib/LigandMPNN/run.py'\n",
    "model_params_path = \"./../Software/fastmpnndesign/lib/LigandMPNN/model_params/\"\n",
    "specific_model_type_for_fused_mpnn = 'protein_mpnn'\n",
    "bias_AA = \"A:-0.75\"\n",
    "sc_num_denoising_steps = 6 \n",
    "repack_everything = 0\n",
    "save_stats = 0 \n",
    "\n",
    "# generate commands\n",
    "\n",
    "for pdb in input_mpnn_pdb_FILTERED_DF[f'{pdb_path_column_name}']: #FOR DATA FRAME\n",
    "    pdb = os.path.join(os.path.dirname(pdb), \"filtered_structures\", os.path.basename(pdb))\n",
    "    cmd = f'python {mpnn_design_script} \\\n",
    "    --model_type {specific_model_type_for_fused_mpnn} \\\n",
    "    --omit_AA {omit_aa} \\\n",
    "    --pdb_path {pdb} \\\n",
    "    --out_folder {updated_mpnn_out_dir} \\\n",
    "    --number_of_batches {nstruct} \\\n",
    "    --temperature {temp} \\\n",
    "    --batch_size {fused_mpnn_batch_size} \\\n",
    "    --pack_side_chains {pack} \\\n",
    "    --fixed_residues_multi {formatted_fixed_AA_jsonl} \\\n",
    "    --sc_num_denoising_steps {sc_num_denoising_steps} \\\n",
    "    --repack_everything {repack_everything} \\\n",
    "    --packed_suffix {packed_suffix} \\\n",
    "    --bias_AA {bias_AA} \\\n",
    "    --save_stats {save_stats} \\\n",
    "    --checkpoint_protein_mpnn {os.path.join(model_params_path, \"proteinmpnn_v_48_020.pt\")} \\\n",
    "    --checkpoint_ligand_mpnn {os.path.join(model_params_path, \"ligandmpnn_v_32_010_25.pt\")} \\\n",
    "    --checkpoint_path_sc {os.path.join(model_params_path, \"ligandmpnn_sc_v_32_002_16.pt\")}'\n",
    "    print(cmd)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5790455c-c197-4a53-b9f8-da3c8d58e957",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Paste Remark 666 Lines into MPNN & Make Updated .jsonl with Catalytic Residues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "5a2ef251-0797-4afc-ab39-02649062fd16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to execute:\n",
      "python {scripts_dir}add_remark666_lines_to_pdb.py --new_pdb_dir ./output/mpnn_out/predesigned_RFD2/backbones/ --reference_old_pdb_dir ./output/predesign_out/filtered_RFD2/ --find_suffixes_from_random_sample 60 --debug \n"
     ]
    }
   ],
   "source": [
    "#########################################################################################\n",
    "### TRANSFER REMARK 666 + IMPORTANT LINES FROM PDBS IN ONE DIRECTORY TO NEW DIRECTORY ###\n",
    "#########################################################################################\n",
    "\n",
    "### INPUT DIRECTORIES ###\n",
    "new_pdb_dir = f'{mpnn_out_dir}predesigned_RFD2/backbones/' #pdbs further in pipeline\n",
    "#num 1: f'{mpnn_out_dir}predesigned_RFD2/backbones/'\n",
    "#num 2: f'{mpnn_out_dir}predesigned_inpainting/backbones/'\n",
    "\n",
    "reference_pdb_dir = f'{predesign_out_dir}filtered_RFD2/' #og pdbs\n",
    "#num 1: f'{predesign_out_dir}filtered_RFD2/'\n",
    "#num 2: f'{predesign_out_dir}inpainting_outputs/'\n",
    "\n",
    "sample_size = 60\n",
    "\n",
    "### CONSTANTS ###\n",
    "script = '{scripts_dir}add_remark666_lines_to_pdb.py'\n",
    "\n",
    "### AUTOMATED COMMAND GENERATION + PRINT ###\n",
    "command = (f\"python {script} \"\n",
    "           f\"--new_pdb_dir {new_pdb_dir} \"\n",
    "           f\"--reference_old_pdb_dir {reference_pdb_dir} \"\n",
    "           f\"--find_suffixes_from_random_sample {sample_size} \"\n",
    "           f\"--debug \" #verbose\n",
    "          )\n",
    "\n",
    "print(\"Command to execute:\")\n",
    "print(command)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b88206b-21b7-4fc7-8044-8257599d6528",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# **III. AlphaFold2 (1st Iteration)**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33ac4906-3789-48f7-8d33-f1b0ddef81ff",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Run AF2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2df2633e",
   "metadata": {},
   "source": [
    "To run SuperFold, you need to download AlphaFold2 model parameters.\n",
    "\n",
    "üìç Steps:\n",
    "1. Make sure `aria2c` is installed:\n",
    "   ```bash\n",
    "   sudo apt install aria2\n",
    "   ```\n",
    "2. Download the parameters by running:\n",
    "   ```bash\n",
    "   bash {software_dir}superfold/scripts/download_all_data.sh {software_dir}superfold/alphafold_weights.pth\n",
    "   ```\n",
    "\n",
    "NOTE: AlphaFold2 parameters are provided under a non-commercial CC BY-NC 4.0 license."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7b49e8c-0262-4f99-8b16-4d4ea1da523f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2/ --model 4 --type monomer_ptm --version monomer --nstruct 1 --max_recycles 6 --num_ensemble 2 --mock_msa_depth 1 --recycle_tol 0.0 --pct_seq_mask 0.15  --output_pae  ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_3.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_4.pdb\n",
      "python {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_8.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9.pdb\n",
      "python {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_1.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_10.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_2.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_3.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_4.pdb\n",
      "python {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_5.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_6.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_7.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_8.pdb ./output/mpnn_out/predesigned_RFD2/backbones/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_9.pdb\n"
     ]
    }
   ],
   "source": [
    "#######################################\n",
    "### predict designs with AlphaFold2 ###\n",
    "#######################################\n",
    "\n",
    "### CONSTANTS ###\n",
    "#apptainer = './containers/crispy.sif' \n",
    "apptainer = 'python' \n",
    "#superfold = '{software_dir}superfold/run_superfold.py'\n",
    "superfold = '{software_dir}superfold/run_superfold.py'\n",
    "\n",
    "### CHANGE THESE EVERY TIME ###\n",
    "af2_inputs_directory = f'{mpnn_out_dir}predesigned_RFD2/backbones/'\n",
    "#num 1: f'{mpnn_out_dir}predesigned_RFD2/backbones/'\n",
    "#num 2: f'{mpnn_out_dir}predesigned_inpainting/backbones/'\n",
    "\n",
    "af2_output_dir = f'{af2_out_dir}predesigned_RFD2/'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting/'\n",
    "os.makedirs(af2_output_dir, exist_ok=True)\n",
    "\n",
    "### ALPHAFOLD PARAMETERS ###\n",
    "model_number = 4 # 4 is recommended, 5 is probably similar\n",
    "model_type = 'monomer_ptm'\n",
    "version = 'monomer'\n",
    "\n",
    "nstruct = 1 # how many structures to generate as outputs, default = 1\n",
    "num_recycles = 6 # default = 3\n",
    "num_ensemble = 2 # default in code is 1 but CASP competition used 8\n",
    "mock_msa_depth = 1 # default is 1, but higher might = better ? goes to +inf\n",
    "recycle_tol = 0.0 # default is 0.0, what convergence rmsd to stop at after recycling\n",
    "\n",
    "pct_seq_mask = 0.15 # 0.15 is default, percent of sequence to make during inference\n",
    "output_summary = '' #'--output_summary' # ENABLE IF YES, otherwise ''\n",
    "output_pae = '--output_pae' # ENABLE IF YES, otherwise ''\n",
    "amber_relax = '' #'--amber_relax' # ENABLE IF YES, otherwise ''\n",
    "\n",
    "### MAKE THE FUNCTION ###\n",
    "cmd = f'{apptainer} {superfold} --out_dir {af2_output_dir} --model {model_number} --type {model_type} --version {version} \\\n",
    "--nstruct {nstruct} --max_recycles {num_recycles} --num_ensemble {num_ensemble} --mock_msa_depth {mock_msa_depth} --recycle_tol {recycle_tol} \\\n",
    "--pct_seq_mask {pct_seq_mask} {output_summary} {output_pae} {amber_relax}' #{initial_guess}'\n",
    "\n",
    "for i, pdb in enumerate(glob.glob(os.path.join(af2_inputs_directory, f\"*.pdb\"))): \n",
    "    if i == 0 or i % 5!= 0:\n",
    "        cmd += f\" {pdb}\"\n",
    "    else:\n",
    "        print(cmd)\n",
    "        cmd = f'{apptainer} {superfold} --out_dir {af2_output_dir} --model {model_number} \\\n",
    "        --type {model_type} --version {version} \\\n",
    "        --nstruct {nstruct} --max_recycles {num_recycles} --num_ensemble {num_ensemble} \\\n",
    "        --mock_msa_depth {mock_msa_depth} --recycle_tol {recycle_tol} \\\n",
    "        --pct_seq_mask {pct_seq_mask} {output_summary} {output_pae} {amber_relax} \\\n",
    "        {pdb}'\n",
    "print(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ce3cc3f-c1be-46d2-99e9-e4046c97663d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Analyze AF2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "122afb76-136a-4f5f-b9df-728336091a21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SUBMIT THIS:\n",
      "python {scripts_dir}af2_parse_processing_multi.py ./output/af2_out/predesigned_RFD2/ --cpus 19\n",
      "\n",
      "OUTPUT CSV FILE HERE:\n",
      "./output/af2_out/predesigned_RFD2/af2_out_parsed.csv\n"
     ]
    }
   ],
   "source": [
    "##################################################################\n",
    "### PARSE & PROCESS AF2 OUTPUT FILES ONLY NEED TO DO THIS ONCE ###\n",
    "##################################################################\n",
    "\n",
    "# INPUT DIRECTORY TO PARSE\n",
    "af2_dir = f'{af2_out_dir}predesigned_RFD2/'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting/'\n",
    "\n",
    "# Constants\n",
    "script = f\"{scripts_dir}af2_parse_processing_multi.py\"\n",
    "optional_cpus = 19  # Optional: specify the number of CPUs to use, or remove this variable to use default logic\n",
    "\n",
    "# Make command\n",
    "command = f\"python {script} {af2_dir}\"\n",
    "if 'optional_cpus' in locals():\n",
    "    command += f\" --cpus {optional_cpus}\"\n",
    "\n",
    "# Print submission\n",
    "print(\"SUBMIT THIS:\")\n",
    "print(command)\n",
    "print('')\n",
    "\n",
    "# Print output .csv path\n",
    "print(\"OUTPUT CSV FILE HERE:\")\n",
    "print(af2_dir + 'af2_out_parsed.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19713797-616c-4e16-b702-7cd2b49cb01a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Filter AF2 structures based on Global Metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "77faf01a-9cdf-4da0-bbfc-2fa00466d813",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10 of 20 (50.0%) with plddt >= 60 and CŒ± rmsd <=3\n",
      "Initial file count in the directory (Should = 0): 20\n",
      "Total files copied: 20\n",
      "\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "############################################\n",
    "### LOAD AND FILTER AF2 BY SUMMARY STATS ###\n",
    "############################################\n",
    "\n",
    "# Correct the path to where your actual CSV file is located\n",
    "csv_file_path = f'{af2_out_dir}predesigned_RFD2/af2_out_parsed.csv'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting/'\n",
    "\n",
    "output_filtered_structures_dir = f'{af2_out_dir}predesigned_RFD2/filtered_structures/'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2/filtered_structures/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting/filtered_structures/'\n",
    "\n",
    "# Select parameters for filtering -> EXTREMELY HARSH IN THIS CASE\n",
    "plddt_cutoff = 60 # in pLDDT units -> usually 70-80\n",
    "rmsd_cutoff = 3 # √Ö, I think this is CŒ± (1.5 is typical, 2.0 might be a stretch but okay for big structures)\n",
    "pae_cutoff = 100 # below 5 is harsh but good\n",
    "pTM_cutoff = 0.5 # Supposedly above 0.75 is good\n",
    "convergence_cutoff = 1.5\n",
    "\n",
    "# Load the CSV file into a DataFrame\n",
    "af2_scores_df = pd.read_csv(csv_file_path)\n",
    "\n",
    "af2_mask = ((af2_scores_df['mean_plddt'] >= plddt_cutoff) & \n",
    "            (af2_scores_df['rmsd_to_input'] <= rmsd_cutoff) &\n",
    "            (af2_scores_df['mean_pae'] <= pae_cutoff) & \n",
    "            (af2_scores_df['pTMscore'] >= pTM_cutoff) &\n",
    "            (af2_scores_df['af2_convergence_tol'] <= convergence_cutoff)\n",
    "           )\n",
    "\n",
    "filtered_af2_scores_df = af2_scores_df[af2_mask]\n",
    "af2_filter_rate = float(len(filtered_af2_scores_df))/float(len(af2_scores_df))*100.0\n",
    "print(f'{len(filtered_af2_scores_df)} of {len(af2_scores_df)} ({round(af2_filter_rate,3)}%) with plddt >= {plddt_cutoff} and CŒ± rmsd <={rmsd_cutoff}')\n",
    "\n",
    "### FUNCTION TO RUN ###\n",
    "def copy_files(df, pdb_col, json_col, output_dir):\n",
    "    existing_files = os.listdir(output_dir)\n",
    "    print(f\"Initial file count in the directory (Should = 0): {len(existing_files)}\")\n",
    "\n",
    "    count = 0\n",
    "    start_time = time.time()\n",
    "    error_files = []\n",
    "\n",
    "    for index, row in df.iterrows():\n",
    "        pdb_path = row[pdb_col]\n",
    "        json_path = row[json_col]\n",
    "\n",
    "        try:\n",
    "            shutil.copy(pdb_path, output_dir)\n",
    "            shutil.copy(json_path, output_dir)\n",
    "            count += 2  # Increment for each PDB and JSON file copied\n",
    "\n",
    "            if count % 2000 == 0:\n",
    "                current_time = time.ctime()\n",
    "                elapsed_time = time.time() - start_time\n",
    "                files_left = (len(df)*2 - count)\n",
    "                average_time_per_file = elapsed_time / count\n",
    "                estimated_time_remaining = max(0, average_time_per_file * files_left)\n",
    "                print(f\"[{current_time}] {count} files copied successfully. Elapsed time: {elapsed_time:.2f} seconds. Average time per file: {average_time_per_file:.4f} seconds. Estimated time remaining: {estimated_time_remaining:.2f} seconds.\")\n",
    "\n",
    "        except Exception as e:\n",
    "            error_files.append((pdb_path, json_path, str(e)))\n",
    "\n",
    "    print(f\"Total files copied: {count}\")\n",
    "    if error_files:\n",
    "        print(\"Failed to copy the following files:\")\n",
    "        for error in error_files:\n",
    "            print(error)\n",
    "\n",
    "### RUN THE SCRIPT ###\n",
    "os.makedirs(output_filtered_structures_dir, exist_ok=True)\n",
    "copy_files(filtered_af2_scores_df, 'pdb_path', 'af2_json_path', output_filtered_structures_dir)\n",
    "print('')\n",
    "print('Done!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2329fa25-d885-4cc0-96cb-1b8b65d0793f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Analyze AF2 on Side Chain Metrics & Do Global Ligand Alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52817e96-8b63-4596-9f9e-2b991341b55f",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Run Both of These Once to Generate Score Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "02e1877b-17e1-40b4-b3aa-b34c7e4ad118",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SUBMIT THIS:\n",
      "python {scripts_dir}af2_parse_processing_multi.py ./output/af2_out/predesigned_RFD2/filtered_structures/ --cpus 19 --optional_basename_for_summary_stats scores.sc\n",
      "\n",
      "OUTPUT CSV FILE HERE:\n",
      "./output/af2_out/predesigned_RFD2/filtered_structures/scores.sc\n"
     ]
    }
   ],
   "source": [
    "##################################################################\n",
    "### PARSE & PROCESS AF2 OUTPUT FILES ONLY NEED TO DO THIS ONCE ###\n",
    "##################################################################\n",
    "\n",
    "# INPUT DIRECTORY TO PARSE\n",
    "af2_directory_to_parse = f'{af2_out_dir}predesigned_RFD2/filtered_structures/'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2/filtered_structures/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting/filtered_structures/'\n",
    "\n",
    "# Constants\n",
    "script = f\"{scripts_dir}af2_parse_processing_multi.py\"\n",
    "optional_cpus = 19  # Optional: specify the number of CPUs to use, or remove this variable to use default logic\n",
    "optional_basename_for_summary_stats = \"scores.sc\"\n",
    "\n",
    "# Make command\n",
    "command = f\"python {script} {af2_directory_to_parse}\"\n",
    "if 'optional_cpus' in locals():\n",
    "    command += f\" --cpus {optional_cpus}\"\n",
    "if 'optional_basename_for_summary_stats' in locals():\n",
    "    command += f\" --optional_basename_for_summary_stats {optional_basename_for_summary_stats}\"\n",
    "\n",
    "# Print submission\n",
    "print(\"SUBMIT THIS:\")\n",
    "print(command)\n",
    "print('')\n",
    "\n",
    "# Print output .csv path\n",
    "print(\"OUTPUT CSV FILE HERE:\")\n",
    "print(af2_directory_to_parse + f'{optional_basename_for_summary_stats}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6e5968e0-1b47-4f19-a95a-3ccc87ec5abf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to execute:\n",
      "python {scripts_dir}sidechain_rmsd_and_info_af2_matching_res.py --scorefile ./output/af2_out/predesigned_RFD2/filtered_structures/scores.sc --ref_pdbs_dir ./output/mpnn_out/predesigned_RFD2/backbones/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --output_file ./output/af2_out/predesigned_RFD2/filtered_structures/updated_scores.sc --max_possible_num_of_suffixes 1 --atom_groups '{\"HIS\": [{\"atoms\": [\"N\", \"CA\", \"C\", \"O\"], \"label\": \"bb_rmsd\"}, {\"atoms\": [\"N\", \"CA\", \"C\", \"O\", \"CB\"], \"label\": \"cb_bb_rmsd\"}, {\"atoms\": [\"CG\", \"CE1\", \"ND1\", \"NE2\", \"CD2\"], \"label\": \"imidazole_rmsd\"}, {\"atoms\": [\"CA\", \"CB\"], \"label\": \"ca_cb_rmsd\"}, {\"atoms\": [\"ND1\", \"NE2\"], \"label\": \"imid_sc_nitrogen_rmsd\"}]}'\n",
      "\n",
      "New scorefile:\n",
      "./output/af2_out/predesigned_RFD2/filtered_structures/updated_scores.sc\n"
     ]
    }
   ],
   "source": [
    "##################################################################################################\n",
    "### CALCULATE SIDE CHAIN RMSDS & OTHER RMSDS FOR CATALYTIC RESIDUES IN FILTERED STRUCTURES DIR ###\n",
    "##################################################################################################\n",
    "\n",
    "### INPUT SCORE PATH + OUTPUT SCORE PATH & REFERENCE FILES ###\n",
    "score_path = os.path.join(f'{af2_out_dir}predesigned_RFD2/filtered_structures', 'scores.sc')\n",
    "output_file = os.path.join(f'{af2_out_dir}predesigned_RFD2/filtered_structures', 'updated_scores.sc')\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2/filtered_structures'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting/filtered_structures'\n",
    "\n",
    "ref_pdbs_dir = os.path.join(f'{mpnn_out_dir}predesigned_RFD2', 'backbones/')\n",
    "#num 1: f'{mpnn_out_dir}predesigned_RFD2/backbones/'\n",
    "#num 2: f'{mpnn_out_dir}predesigned_inpainting/backbones/'\n",
    "\n",
    "### INPUT PARAMS ###\n",
    "params_path = os.path.join(invrot_dir, \"COMBINED_PHENYLACETATE_HHH\", \"COMBINED_PARAMS\", \"*.params\")  # Ensure params include .pdb if needed\n",
    "\n",
    "### INPUT ATOM RMSDS OF CAT RESIDUES TO CALCULATE & NUM OF SUFFIXES ###\n",
    "atom_groups_json = { #MANUALLY INPUT POSSIBLE CATALYTIC RESIDUES!!\n",
    "    \"HIS\": [\n",
    "        {\"atoms\": [\"N\", \"CA\", \"C\", \"O\"], \"label\": \"bb_rmsd\"},\n",
    "        {\"atoms\": [\"N\", \"CA\", \"C\", \"O\", \"CB\"], \"label\": \"cb_bb_rmsd\"},\n",
    "        {\"atoms\": [\"CG\", \"CE1\", \"ND1\", \"NE2\", \"CD2\"], \"label\": \"imidazole_rmsd\"},\n",
    "        {\"atoms\": [\"CA\", \"CB\"], \"label\": \"ca_cb_rmsd\"},\n",
    "        {\"atoms\": [\"ND1\", \"NE2\"], \"label\": \"imid_sc_nitrogen_rmsd\"}\n",
    "    ]\n",
    "}\n",
    "\n",
    "max_possible_num_of_suffixes = 1\n",
    "column_name_for_pdb_path = 'pdb_path'\n",
    "\n",
    "### CONSTANTS ###\n",
    "script = '{scripts_dir}sidechain_rmsd_and_info_af2_matching_res.py'\n",
    "atom_groups = json.dumps(atom_groups_json)\n",
    "\n",
    "# Build the command\n",
    "command = (\n",
    "    f'python {script} --scorefile {score_path} --ref_pdbs_dir {ref_pdbs_dir} --params {params_path} --output_file {output_file} '\n",
    "    f'--max_possible_num_of_suffixes {max_possible_num_of_suffixes} --atom_groups \\'{atom_groups}\\''\n",
    ")\n",
    "print(\"Command to execute:\")\n",
    "print(command)\n",
    "print('')\n",
    "print('New scorefile:')\n",
    "print(f'{output_file}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dff92e0-2854-4dc6-8b25-6f0fb7c9fe3b",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Do a Light Side Chain Filter Before Global Alignment & Harsh SC Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ddc9364f-a543-477f-b98e-b8d29c56441c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                                                                             af2_json_path  \\\n",
      "0   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_prediction_results.json   \n",
      "1  ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_prediction_results.json   \n",
      "2   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_prediction_results.json   \n",
      "3   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_prediction_results.json   \n",
      "4   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_prediction_results.json   \n",
      "5   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_3_model_4_ptm_seed_0_prediction_results.json   \n",
      "6   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_prediction_results.json   \n",
      "7   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_prediction_results.json   \n",
      "8   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_8_model_4_ptm_seed_0_prediction_results.json   \n",
      "9   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_5_model_4_ptm_seed_0_prediction_results.json   \n",
      "\n",
      "                                                                                                                                        pdb_path  \\\n",
      "0   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "1  ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "2   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "3   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "4   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "5   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_3_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "6   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "7   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "8   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_8_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "9   ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_2-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "\n",
      "   mean_plddt  rmsd_to_input  mean_pae_intra_chain  mean_pae  pTMscore  \\\n",
      "0   90.765816       1.187505              4.063441  4.116622  0.740844   \n",
      "1   89.050667       1.644919              4.716823  4.750964  0.735837   \n",
      "2   91.287888       1.096059              3.876372  3.976256  0.738688   \n",
      "3   84.884445       1.748606              5.473359  5.523567  0.702419   \n",
      "4   91.423553       1.368132              3.862613  3.934102  0.743584   \n",
      "5   90.966743       1.004445              3.980876  4.023002  0.741262   \n",
      "6   90.995110       0.977493              3.921531  3.957700  0.741256   \n",
      "7   86.367966       1.300914              5.206219  5.281673  0.715544   \n",
      "8   92.114357       1.063859              3.680240  3.740887  0.748256   \n",
      "9   87.000061       2.641786              5.447616  5.479612  0.838213   \n",
      "\n",
      "   af2_convergence_tol  af2_elapsed_folding_time  cat_HIS1_bb_rmsd  \\\n",
      "0             0.042343                110.829311          0.577407   \n",
      "1             0.041550                160.040541          1.624625   \n",
      "2             0.189567                 36.302838          1.205954   \n",
      "3             0.103899                 36.202080          1.374152   \n",
      "4             0.175769                 36.238962          0.585831   \n",
      "5             0.045493                 36.231157          0.912768   \n",
      "6             0.066977                 36.180618          1.100983   \n",
      "7             0.129313                 36.345173          0.881191   \n",
      "8             0.037367                 36.467609          1.893443   \n",
      "9             0.139926                 37.063108               NaN   \n",
      "\n",
      "   cat_HIS1_cb_bb_rmsd  cat_HIS1_imidazole_rmsd  cat_HIS1_ca_cb_rmsd  \\\n",
      "0             0.645142                 1.616285             0.783781   \n",
      "1             1.723763                 2.838078             1.902931   \n",
      "2             1.321830                 2.616269             1.529052   \n",
      "3             1.453894                 2.755601             1.599929   \n",
      "4             0.644111                 1.667671             0.768093   \n",
      "5             1.052741                 3.081319             1.271622   \n",
      "6             1.228210                 3.241027             1.445409   \n",
      "7             1.011725                 3.410172             1.243940   \n",
      "8             2.217033                 5.216434             2.729788   \n",
      "9                  NaN                      NaN                  NaN   \n",
      "\n",
      "   cat_HIS1_imid_sc_nitrogen_rmsd  cat_HIS2_bb_rmsd  cat_HIS2_cb_bb_rmsd  \\\n",
      "0                        1.755811          0.813298             0.962944   \n",
      "1                        3.065386          1.107936             1.122313   \n",
      "2                        2.927453          1.026377             1.056505   \n",
      "3                        3.071072          0.712577             0.705035   \n",
      "4                        1.707691          0.942284             1.087777   \n",
      "5                        3.364500          0.797419             0.841696   \n",
      "6                        3.653322          0.949580             1.094936   \n",
      "7                        3.724546          0.752280             0.736957   \n",
      "8                        5.345318          2.529770             3.222687   \n",
      "9                             NaN               NaN                  NaN   \n",
      "\n",
      "   cat_HIS2_imidazole_rmsd  cat_HIS2_ca_cb_rmsd  \\\n",
      "0                 2.036191             1.179496   \n",
      "1                 1.763634             1.107203   \n",
      "2                 1.449620             1.026444   \n",
      "3                 1.310746             0.644083   \n",
      "4                 2.319001             1.293884   \n",
      "5                 1.517761             0.818851   \n",
      "6                 1.981024             1.304435   \n",
      "7                 1.330231             0.628011   \n",
      "8                 8.231021             4.015279   \n",
      "9                      NaN                  NaN   \n",
      "\n",
      "   cat_HIS2_imid_sc_nitrogen_rmsd  cat_HIS3_bb_rmsd  cat_HIS3_cb_bb_rmsd  \\\n",
      "0                        2.113378          1.482473             1.501260   \n",
      "1                        1.946631          1.895275             1.904838   \n",
      "2                        1.531902          1.504563             1.510899   \n",
      "3                        1.468343          1.225027             1.235307   \n",
      "4                        2.571307          1.478604             1.474324   \n",
      "5                        1.671755          1.376106             1.421657   \n",
      "6                        2.172214          1.375375             1.396372   \n",
      "7                        1.387933          1.642280             1.662941   \n",
      "8                        8.742316          4.094667             4.153018   \n",
      "9                             NaN               NaN                  NaN   \n",
      "\n",
      "   cat_HIS3_imidazole_rmsd  cat_HIS3_ca_cb_rmsd  \\\n",
      "0                 1.630207             1.516928   \n",
      "1                 1.839041             1.875346   \n",
      "2                 1.312904             1.502277   \n",
      "3                 1.331245             1.212704   \n",
      "4                 1.159433             1.477836   \n",
      "5                 1.375229             1.486152   \n",
      "6                 1.538301             1.415268   \n",
      "7                 1.735688             1.683543   \n",
      "8                 4.181522             4.274958   \n",
      "9                      NaN                  NaN   \n",
      "\n",
      "   cat_HIS3_imid_sc_nitrogen_rmsd  cat_HIS_bb_mean_rmsd  \\\n",
      "0                        1.564893              0.957726   \n",
      "1                        1.742396              1.542612   \n",
      "2                        1.270004              1.245631   \n",
      "3                        1.296578              1.103919   \n",
      "4                        1.041103              1.002240   \n",
      "5                        1.326124              1.028764   \n",
      "6                        1.471370              1.141979   \n",
      "7                        1.688507              1.091917   \n",
      "8                        4.209390              2.839293   \n",
      "9                             NaN                   NaN   \n",
      "\n",
      "   cat_HIS_bb_rms_of_rmsds  cat_HIS_cb_bb_mean_rmsd  \\\n",
      "0                 1.031597                 1.036448   \n",
      "1                 1.576811                 1.583638   \n",
      "2                 1.261148                 1.296412   \n",
      "3                 1.139701                 1.131412   \n",
      "4                 1.067297                 1.068737   \n",
      "5                 1.058724                 1.105365   \n",
      "6                 1.155497                 1.239840   \n",
      "7                 1.160389                 1.137208   \n",
      "8                 2.986144                 3.197579   \n",
      "9                      NaN                      NaN   \n",
      "\n",
      "   cat_HIS_cb_bb_rms_of_rmsds  cat_HIS_ca_cb_ONLY_mean_rmsd  \\\n",
      "0                    1.095026                      1.160068   \n",
      "1                    1.618575                      1.628493   \n",
      "2                    1.309740                      1.352591   \n",
      "3                    1.174289                      1.152239   \n",
      "4                    1.121274                      1.179938   \n",
      "5                    1.131050                      1.192208   \n",
      "6                    1.245959                      1.388370   \n",
      "7                    1.201675                      1.185165   \n",
      "8                    3.293858                      3.673342   \n",
      "9                         NaN                           NaN   \n",
      "\n",
      "   cat_HIS_ca_cb_ONLY_rms_of_rmsds  cat_HIS_imidazole_mean_rmsd  \\\n",
      "0                         1.198136                     1.760894   \n",
      "1                         1.669727                     2.146918   \n",
      "2                         1.372155                     1.792931   \n",
      "3                         1.217274                     1.799197   \n",
      "4                         1.217662                     1.715368   \n",
      "5                         1.224226                     1.991436   \n",
      "6                         1.389693                     2.253451   \n",
      "7                         1.261758                     2.158697   \n",
      "8                         3.734947                     5.876325   \n",
      "9                              NaN                          NaN   \n",
      "\n",
      "   cat_HIS_imidazole_rms_of_rmsds  cat_HIS_imidazole_tip_Ns_mean_rmsd  \\\n",
      "0                        1.771630                            1.811361   \n",
      "1                        2.202057                            2.251471   \n",
      "2                        1.885911                            1.909786   \n",
      "3                        1.922117                            1.945331   \n",
      "4                        1.779811                            1.773367   \n",
      "5                        2.136148                            2.120793   \n",
      "6                        2.366085                            2.432302   \n",
      "7                        2.338903                            2.266996   \n",
      "8                        6.122255                            6.099008   \n",
      "9                             NaN                                 NaN   \n",
      "\n",
      "   cat_HIS_imidazole_tip_Ns_rms_of_rmsds  \n",
      "0                               1.825571  \n",
      "1                               2.325361  \n",
      "2                               2.043658  \n",
      "3                               2.103064  \n",
      "4                               1.880755  \n",
      "5                               2.300234  \n",
      "6                               2.596806  \n",
      "7                               2.493302  \n",
      "8                               6.395818  \n",
      "9                                    NaN  \n"
     ]
    }
   ],
   "source": [
    "##################################################################################\n",
    "### COMPUTE SIDE CHAIN STATISTICS FROM SCORES DF & DO GENTLE/LOGICAL FILTERING ###\n",
    "##################################################################################\n",
    "\n",
    "### INPUT PATH ###\n",
    "scores_filepath = f\"{af2_out_dir}predesigned_RFD2/filtered_structures/updated_scores.sc\"\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2/filtered_structures/updated_scores.sc'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting/filtered_structures/updated_scores.sc'\n",
    "inpaint_or_diffusion = \"RFD2\" #RFD2 or Inpainting\n",
    "\n",
    "### INPUT SPECIFY COLUMN GROUPS FOR CALCULATION ###\n",
    "column_groups = {\n",
    "    'cat_HIS_bb': ['cat_HIS1_bb_rmsd', 'cat_HIS2_bb_rmsd', 'cat_HIS3_bb_rmsd'],\n",
    "    'cat_HIS_cb_bb': ['cat_HIS1_cb_bb_rmsd', 'cat_HIS2_cb_bb_rmsd', 'cat_HIS3_cb_bb_rmsd'],\n",
    "    'cat_HIS_ca_cb_ONLY': ['cat_HIS1_ca_cb_rmsd', 'cat_HIS2_ca_cb_rmsd', 'cat_HIS3_ca_cb_rmsd'],\n",
    "    'cat_HIS_imidazole': ['cat_HIS1_imidazole_rmsd', 'cat_HIS2_imidazole_rmsd', 'cat_HIS3_imidazole_rmsd'],\n",
    "    'cat_HIS_imidazole_tip_Ns': ['cat_HIS1_imid_sc_nitrogen_rmsd', 'cat_HIS2_imid_sc_nitrogen_rmsd', 'cat_HIS3_imid_sc_nitrogen_rmsd']\n",
    "}\n",
    "\n",
    "### FUNCTION TO CALCULATE MEAN RMSD AND RMS OF RMSDs FOR SPECIFIED COLUMN GROUPS ###\n",
    "def calculate_group_rmsd(df, column_groups):\n",
    "    for group_name, columns in column_groups.items():\n",
    "        mean_col_name = f\"{group_name}_mean_rmsd\"\n",
    "        rms_col_name = f\"{group_name}_rms_of_rmsds\"\n",
    "        df[mean_col_name] = df[columns].mean(axis=1)\n",
    "        df[rms_col_name] = np.sqrt((df[columns]**2).mean(axis=1))\n",
    "    print (df)\n",
    "\n",
    "### EXECUTE PRINTING ###\n",
    "sidechain_rmsd_df = pd.read_csv(scores_filepath, sep=',', na_values='')\n",
    "calculate_group_rmsd(sidechain_rmsd_df, column_groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "55925550-1256-494a-9c8f-ed714cf7df49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########################## SIDE CHAIN RMSD METRICS FILTER ##########################\n",
      "8 of 10 (80.000%) have CATALYTIC HIS BB RMS OF RMSDS <= 2\n",
      "8 of 10 (80.000%) have CATALYTIC HIS CB BB RMS OF RMSDS <= 2\n",
      "8 of 10 (80.000%) have CATALYTIC HIS CA CB ONLY RMS OF RMSDS <= 2\n",
      "8 of 10 (80.000%) have CATALYTIC HIS IMIDAZOLE RMS OF RMSDS <= 3\n",
      "8 of 10 (80.000%) have CATALYTIC HIS IMIDAZOLE TIP NS RMS OF RMSDS <= 3\n",
      "\n",
      "8 of 10 (80.000%) pass at least one filter!\n",
      "####################################################################################\n",
      "Columns used in filtering: cat_HIS_bb_rms_of_rmsds, cat_HIS_cb_bb_rms_of_rmsds, cat_HIS_ca_cb_ONLY_rms_of_rmsds, cat_HIS_imidazole_rms_of_rmsds, cat_HIS_imidazole_tip_Ns_rms_of_rmsds\n"
     ]
    }
   ],
   "source": [
    "##################################################################################\n",
    "### COMPUTE SIDE CHAIN STATISTICS FROM SCORES DF & DO GENTLE/LOGICAL FILTERING ###\n",
    "##################################################################################\n",
    "\n",
    "### NEW FILTER FROM SIDE CHAIN 2 ###\n",
    "cat_HIS_bb_rms_of_rmsds_cutoff = 2 #2 # √Ö\n",
    "cat_HIS_cb_bb_rms_of_rmsds_cutoff = 2 # √Ö\n",
    "cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff = 2 # √Ö\n",
    "cat_HIS_imidazole_rms_of_rmsds_cutoff = 3 # √Ö\n",
    "cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff = 3 # √Ö\n",
    "\n",
    "# Individual filter calculations with specific naming\n",
    "filter_masks = {\n",
    "    'cat_HIS_bb_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_bb_rms_of_rmsds'] <= cat_HIS_bb_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_cb_bb_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_cb_bb_rms_of_rmsds'] <= cat_HIS_cb_bb_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_ca_cb_ONLY_rms_of_rmsds'] <= cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_imidazole_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_imidazole_rms_of_rmsds'] <= cat_HIS_imidazole_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_imidazole_tip_Ns_rms_of_rmsds'] <= cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff\n",
    "}\n",
    "\n",
    "# Calculate pass counts and rates for each filter\n",
    "filter_results = {filter_name: (mask.sum(), len(sidechain_rmsd_df), mask.mean() * 100) for filter_name, mask in filter_masks.items()}\n",
    "\n",
    "# Overall filtering result using OR logic\n",
    "cumulative_filter_mask_or = filter_masks['cat_HIS_bb_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_cb_bb_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_imidazole_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask']\n",
    "\n",
    "cumulative_pass_count_or = cumulative_filter_mask_or.sum()\n",
    "cumulative_total_or = len(sidechain_rmsd_df)\n",
    "cumulative_pass_rate_or = cumulative_pass_count_or / cumulative_total_or * 100\n",
    "\n",
    "### DO THE PRINT STATEMENT (FANCY) ###\n",
    "print('########################## SIDE CHAIN RMSD METRICS FILTER ##########################')\n",
    "for filter_name, (pass_count, total, pass_rate) in filter_results.items():\n",
    "    specific_criteria = filter_name.replace('_mask', '').replace('_', ' ').replace('cat', 'catalytic').upper()\n",
    "    cutoff_value = {\n",
    "        'cat_HIS_bb_rms_of_rmsds_mask': cat_HIS_bb_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_cb_bb_rms_of_rmsds_mask': cat_HIS_cb_bb_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask': cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_imidazole_rms_of_rmsds_mask': cat_HIS_imidazole_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask': cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff\n",
    "    }[filter_name]\n",
    "    print(f'{pass_count} of {total} ({pass_rate:.3f}%) have {specific_criteria} <= {cutoff_value}')\n",
    "\n",
    "# Maintaining the cumulative print statement from the provided script\n",
    "print('')\n",
    "print(f'{cumulative_pass_count_or} of {cumulative_total_or} ({cumulative_pass_rate_or:.3f}%) pass at least one filter!')\n",
    "print('####################################################################################')\n",
    "\n",
    "### MAKE THE FINAL DATA FRAME ###\n",
    "final_filtered_df = sidechain_rmsd_df[cumulative_filter_mask_or]\n",
    "\n",
    "### PRINT OUT COLUMN NAMES USED IN FILTERING ###\n",
    "filtered_columns = ', '.join([key.replace('_mask', '') for key in filter_masks.keys()])\n",
    "print(f'Columns used in filtering: {filtered_columns}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc044fa4-1758-4971-96aa-a631da2a4bab",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create Alignment with Lightly Filtered Structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "359ebf2b-917a-41ac-acb0-1d8e4b3051ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed.pdb --ref ./output/predesign_out/filtered_RFD2/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb --outdir ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed.pdb --ref ./output/predesign_out/filtered_RFD2/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb --outdir ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed.pdb --ref ./output/predesign_out/filtered_RFD2/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb --outdir ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed.pdb --ref ./output/predesign_out/filtered_RFD2/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb --outdir ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed.pdb --ref ./output/predesign_out/filtered_RFD2/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb --outdir ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_3_model_4_ptm_seed_0_unrelaxed.pdb --ref ./output/predesign_out/filtered_RFD2/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb --outdir ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed.pdb --ref ./output/predesign_out/filtered_RFD2/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb --outdir ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed.pdb --ref ./output/predesign_out/filtered_RFD2/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0.pdb --outdir ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "8 rows processed, 0 remaining. 4 ref pdb files available for matching. Elapsed time: 0.00 s. Estimated time left: 0.00 s.\n"
     ]
    }
   ],
   "source": [
    "##############################################\n",
    "### GENERATE COMMANDS FOR ALIGNING LIGANDS ###\n",
    "##############################################\n",
    "\n",
    "import time\n",
    "\n",
    "### INPUTS ###\n",
    "column_of_pdb_paths = 'pdb_path' # COLUMN OF DF W/PDB PATHS\n",
    "\n",
    "ref_pdbs_to_match = f'{predesign_out_dir}filtered_RFD2/'\n",
    "#num1: f'{predesign_out_dir}filtered_RFD2/'\n",
    "#num2: f'{predesign_out_dir}inpainting_outputs/'\n",
    "\n",
    "params_path = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/\"\n",
    "\n",
    "align_af2_out_directory = os.path.join(f\"{af2_out_dir}\", \n",
    "                                       \"predesigned_RFD2\", # num1: predesigned_RFD2, num2: predesigned_inpainting\n",
    "                                       \"filtered_structures\",\n",
    "                                       \"alignment/\")\n",
    "### CONSTANTS ###\n",
    "align_script = f\"{scripts_dir}align_af2_with_inputs_and_copy_ligand.py\"\n",
    "\n",
    "### REFERENCE PDB PROCESSING ###\n",
    "ref_pdb_files = glob.glob(os.path.join(ref_pdbs_to_match, '*.pdb'))\n",
    "ref_pdb_map = {os.path.basename(p).replace('.pdb', ''): p for p in ref_pdb_files}\n",
    "ref_pdb_usage = Counter()\n",
    "\n",
    "### AUTOMATED COMMAND WRITING & REFERENCE MATCHING ###\n",
    "os.makedirs(align_af2_out_directory, exist_ok=True)\n",
    "max_usages_per_ref_pdb = 30  # Number of times a reference pdb can be used\n",
    "\n",
    "total_rows = len(final_filtered_df)\n",
    "rows_processed = 0\n",
    "update_interval = 1000\n",
    "\n",
    "start_time, num_cmds = time.time(), 0\n",
    "for index, row in final_filtered_df.iterrows():\n",
    "    pdb_path = row[column_of_pdb_paths]\n",
    "    pdb_name = os.path.basename(pdb_path).split('.')[0]\n",
    "\n",
    "    # Find matching reference PDB\n",
    "    matching_ref = None\n",
    "    for ref_name, ref_path in ref_pdb_map.items():\n",
    "        if pdb_name.startswith(ref_name) and ref_pdb_usage[ref_name] < max_usages_per_ref_pdb:\n",
    "            matching_ref = ref_path\n",
    "            ref_pdb_usage[ref_name] += 1\n",
    "            if ref_pdb_usage[ref_name] >= max_usages_per_ref_pdb:\n",
    "                del ref_pdb_map[ref_name]  # Remove from future considerations\n",
    "            break\n",
    "\n",
    "    if matching_ref:\n",
    "        command_align = (f\"python {align_script} \"\n",
    "                         f\"--pdb {pdb_path} \"\n",
    "                         f\"--ref {matching_ref} \"\n",
    "                         f\"--outdir {align_af2_out_directory} \"\n",
    "                         f\"--params {os.path.join(params_path, '*.params')} \"\n",
    "                         f\"--fix_catres\")\n",
    "        print (command_align)\n",
    "        num_cmds += 1\n",
    "\n",
    "    rows_processed += 1\n",
    "    if rows_processed % update_interval == 0 or rows_processed == total_rows:\n",
    "        elapsed_time = time.time() - start_time\n",
    "        time_per_row = elapsed_time / rows_processed\n",
    "        estimated_time_remaining = time_per_row * (total_rows - rows_processed)\n",
    "        print(f\"{rows_processed} rows processed, {total_rows - rows_processed} remaining. \"\n",
    "              f\"{len(ref_pdb_map)} ref pdb files available for matching. \"\n",
    "              f\"Elapsed time: {elapsed_time:.2f} s. Estimated time left: {estimated_time_remaining:.2f} s.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef55dfe",
   "metadata": {},
   "source": [
    "# **IV. FastMPNNdesign**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab6951a8-0a08-4830-98c4-e8af7cdf3546",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Run FastMPNNdesign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "53859d90-9abd-40fe-941e-b9d2a6067871",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run the command below:\n",
      "python {scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ.pdb --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --nstruct 2 --mpnn --detect_pocket --bias_atoms H1 --bias_AAs ED --position_bias 4  --scoring {scripts_dir}scoring/esterase_scoring.py --outdir ./output/fastMPNNdesign/RFD2/\n",
      "python {scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ.pdb --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --nstruct 2 --mpnn --detect_pocket --bias_atoms H1 --bias_AAs ED --position_bias 4  --scoring {scripts_dir}scoring/esterase_scoring.py --outdir ./output/fastMPNNdesign/RFD2/\n",
      "python {scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ.pdb --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --nstruct 2 --mpnn --detect_pocket --bias_atoms H1 --bias_AAs ED --position_bias 4  --scoring {scripts_dir}scoring/esterase_scoring.py --outdir ./output/fastMPNNdesign/RFD2/\n",
      "python {scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_3_model_4_ptm_seed_0_unrelaxed_PSZ.pdb --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --nstruct 2 --mpnn --detect_pocket --bias_atoms H1 --bias_AAs ED --position_bias 4  --scoring {scripts_dir}scoring/esterase_scoring.py --outdir ./output/fastMPNNdesign/RFD2/\n",
      "python {scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ.pdb --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --nstruct 2 --mpnn --detect_pocket --bias_atoms H1 --bias_AAs ED --position_bias 4  --scoring {scripts_dir}scoring/esterase_scoring.py --outdir ./output/fastMPNNdesign/RFD2/\n",
      "python {scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ.pdb --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --nstruct 2 --mpnn --detect_pocket --bias_atoms H1 --bias_AAs ED --position_bias 4  --scoring {scripts_dir}scoring/esterase_scoring.py --outdir ./output/fastMPNNdesign/RFD2/\n",
      "python {scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ.pdb --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --nstruct 2 --mpnn --detect_pocket --bias_atoms H1 --bias_AAs ED --position_bias 4  --scoring {scripts_dir}scoring/esterase_scoring.py --outdir ./output/fastMPNNdesign/RFD2/\n",
      "python {scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py --pdb ./output/af2_out/predesigned_RFD2/filtered_structures/alignment/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ.pdb --cstfile ./input/theozymes/cst_files/PSZ_3H_NO_DEPROT_RES.cst --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/PSZ.params --nstruct 2 --mpnn --detect_pocket --bias_atoms H1 --bias_AAs ED --position_bias 4  --scoring {scripts_dir}scoring/esterase_scoring.py --outdir ./output/fastMPNNdesign/RFD2/\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "### FastMPNN ###\n",
    "################\n",
    "\n",
    "### FILTERED & ALIGNED AF2 INPUT DIRECTORY ###\n",
    "input_af2_aligned_str_dir = f'{af2_out_dir}predesigned_RFD2/filtered_structures/alignment/'\n",
    "#num1: predesigned_RFD2\n",
    "#num2: predesigned_inpainting\n",
    "\n",
    "### OUTPUT DIRECTORY ###\n",
    "output_directory = os.path.join(fastMPNNdesign_dir, 'RFD2/') #num1: RFD2, #num2: inpainting\n",
    "os.makedirs(output_directory, exist_ok=True)\n",
    "\n",
    "### INPUT LIGANDS ###\n",
    "multiple_ligand_names = [\"PSZ\"]\n",
    "\n",
    "### INPUT CST & PARAMS DIRECTORIES ###\n",
    "cst_files_dir = f\"{theozymes_dir}cst_files/\"\n",
    "params_and_theozyme_pdb_dir = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/\"\n",
    "\n",
    "### INPUT PARAMETERS ###\n",
    "nstruct = 2\n",
    "mpnn_flag = '--mpnn'\n",
    "detect_pocket = '--detect_pocket'\n",
    "bias_atoms = \"H1\"\n",
    "bias_AAs = \"ED\"\n",
    "bias_weight = 4\n",
    "filter_flag = '' #--filter\n",
    "\n",
    "### CONSTANTS ###\n",
    "fastmpnn_script = f\"{scripts_dir}enzyme_design/fastmpnndesign_ZnEsterase.py\"\n",
    "scoring_file = f\"{scripts_dir}scoring/esterase_scoring.py\"\n",
    "\n",
    "# Generate commands\n",
    "for ligand_name in multiple_ligand_names:\n",
    "    # Find all PDB files for the current ligand\n",
    "    pdb_files = glob.glob(os.path.join(input_af2_aligned_str_dir, f\"*{ligand_name}*.pdb\"))\n",
    "    if len(pdb_files) == 0:\n",
    "        continue\n",
    "\n",
    "    # Find cst and params files, handling cases where they might not be found\n",
    "    cst_files = glob.glob(os.path.join(cst_files_dir, f\"*{ligand_name}*.cst\"))\n",
    "    if cst_files:\n",
    "        cst_file = cst_files[0]\n",
    "    else:\n",
    "        print(f\"No CST files found for {ligand_name}.\")\n",
    "        continue\n",
    "\n",
    "    params_files = glob.glob(os.path.join(params_and_theozyme_pdb_dir, f\"*{ligand_name}*.params\"))\n",
    "    if params_files:\n",
    "        params_file = params_files[0]\n",
    "    else:\n",
    "        print(f\"No PARAMS files found for {ligand_name}.\")\n",
    "        continue\n",
    "\n",
    "    # Generate a command for each pdb file\n",
    "    print (\"Run the command below:\")\n",
    "    for pdb_file in pdb_files:\n",
    "        command = (\n",
    "            f\"python {fastmpnn_script} --pdb {pdb_file} \"\n",
    "            f\"--cstfile {cst_file} --params {params_file} --nstruct {nstruct} {mpnn_flag} {detect_pocket} \"\n",
    "            f\"--bias_atoms {bias_atoms} --bias_AAs {bias_AAs} --position_bias {bias_weight} {filter_flag} --scoring {scoring_file} --outdir {output_directory}\"\n",
    "        )\n",
    "        print (command)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "659f9169-66fd-4132-936c-5cb2cceaa014",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Analyze FastMPNNdesign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7a30d271-0a73-486e-96b3-843d8675bb74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 files were found!\n",
      "DataFrame RFD2_df loaded with 160 rows and 44 columns.\n",
      "\n",
      "22 files were found!\n",
      "DataFrame inpainting_df loaded with 438 rows and 44 columns.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "####################################################\n",
    "### LOAD FASTMPNN SCORES AND GRAPH SUMMARY STATS ###\n",
    "####################################################\n",
    "\n",
    "### INPUTS ###\n",
    "base_dir_for_fastmpnn = fastMPNNdesign_dir\n",
    "\n",
    "subdirectories = ['RFD2', 'inpainting']\n",
    "dataframe_names = [x+\"_df\" for x in subdirectories]\n",
    "\n",
    "dataframes = {}\n",
    "for subdir, df_name in zip(subdirectories, dataframe_names):\n",
    "    found_files = glob.glob(os.path.join(base_dir_for_fastmpnn, subdir, \"scores\", \"*.sc\"))\n",
    "    print (f'{len(found_files)} files were found!')\n",
    "    if found_files:\n",
    "        dfs = []\n",
    "        for found_file in found_files:\n",
    "            df = pd.read_csv(found_file, delim_whitespace=True, comment='#')\n",
    "            dfs.append(df)\n",
    "        dfs = pd.concat(dfs, ignore_index=True)\n",
    "        dataframes[df_name] = dfs\n",
    "        globals()[df_name] = dfs  # Assign DataFrame to a global variable\n",
    "        print(f\"DataFrame {df_name} loaded with {dfs.shape[0]} rows and {dfs.shape[1]} columns.\")\n",
    "        print('')\n",
    "        dfs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6c5986b-6e14-4568-b50e-38a592a922e4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Filter FastMPNNdesign structures "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bb840a70-ca5b-4c2a-9a44-bdc15becbd56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[RFD2_df_P*Z] (160 designs)\n",
      "# [ L_SASA          <= 0.2   ]: 160    (100.0%)\n",
      "# [ L_SASA          >= 0     ]: 160    (100.0%)\n",
      "# [ substrate_SASA  <= 60    ]: 160    (100.0%)\n",
      "# [ substrate_SASA  >= 1     ]: 82     (51.2%)\n",
      "# [ all_cst         <= 10    ]: 139    (86.9%)\n",
      "# [ H2O_hbond       >= 0     ]: 160    (100.0%)\n",
      "# [ cms_per_atom    >= 6     ]: 160    (100.0%)\n",
      "# [ corrected_ddg   <= -20   ]: 160    (100.0%)\n",
      "# [ nlr_totrms      <= 1.0   ]: 159    (99.4%)\n",
      "# [ score_per_res   <  0.0   ]: 160    (100.0%)\n",
      "# [           ALL            ]: 67 (41.9%)\n",
      "67 of 160 (41.875%) passed\n",
      "\n",
      "DataFrame RFD2_df saved to ./working_dir/important_dfs/RFD2_fastmpnn_stats.csv\n",
      "DataFrame inpainting_df saved to ./working_dir/important_dfs/inpainting_fastmpnn_stats.csv\n"
     ]
    }
   ],
   "source": [
    "##################################################################\n",
    "### FILTER THE DATA FRAME BASED ON WHATEVER CRITERIA ARE IDEAL ###\n",
    "##################################################################\n",
    "\n",
    "show_df_head = False # True = show first 5 rows of df; False = do not \n",
    "\n",
    "dataframe_name = \"RFD2_df\"  # num1: RFD2_df, num2: inpainting_df\n",
    "target_df = eval(dataframe_name)\n",
    "labels_for_print = ['description', 'all_cst', 'H2O_hbond', 'oxy_hbond', 'cms_per_atom', 'corrected_ddg',\\\n",
    "                    'nlr_totrms', 'L_SASA', 'substrate_SASA', 'O2_hbond']\n",
    "\n",
    "### MANDATORY (HARD-CODED) FILTERS ###\n",
    "all_cst = 10 # 7\n",
    "H2O_hbond = 0 # hydrogen bond with H of water molecule\n",
    "O2_hbond = 1 # hydrogen bond with negatively charged oxygen\n",
    "\n",
    "### OPTIONAL FILTERS ###\n",
    "cms_per_atom = 6 # contact molecular surface\n",
    "corrected_ddg = -20 # ddg\n",
    "nlr_totrms = 1.0 # no-ligand-repack\n",
    "score_per_res = 0.0 # usually always want < 0\n",
    "\n",
    "### SASA UPPER + LOWER FILTERS ###\n",
    "L_SASA_lower = 0 # ligand exposure\n",
    "L_SASA_upper = 0.2 # ligand exposure\n",
    "substrate_SASA_lower = 1 # substrate exposure\n",
    "substrate_SASA_upper = 60 # substrate exposure\n",
    "\n",
    "\n",
    "### FILTER ###\n",
    "filters_P_Z = {\n",
    "               (\"L_SASA\", \"<=\"): L_SASA_upper, \n",
    "               (\"L_SASA\", \">=\"): L_SASA_lower, \n",
    "               (\"substrate_SASA\", \"<=\"): substrate_SASA_upper, \n",
    "               (\"substrate_SASA\", \">=\"): substrate_SASA_lower, \n",
    "               (\"all_cst\", \"<=\"): all_cst, \n",
    "               (\"H2O_hbond\", \">=\"): H2O_hbond, \n",
    "               (\"cms_per_atom\", \">=\"): cms_per_atom, \n",
    "               (\"corrected_ddg\", \"<=\"): corrected_ddg, \n",
    "               (\"nlr_totrms\", \"<=\"): nlr_totrms,\n",
    "               (\"score_per_res\", \"<\"): score_per_res\n",
    "              }\n",
    "\n",
    "def filtering_df(dataframe_name, filters, target_df, print_stat=True):\n",
    "    if print_stat: print (f'[{dataframe_name}] ({len(target_df)} designs)')\n",
    "    \n",
    "    converted_filters = []\n",
    "    for key, condition in filters.items():\n",
    "        key, operator = key\n",
    "        converted_filter = f\"target_df['{key}'] {operator} {condition}\"\n",
    "        converted_filters.append(converted_filter)\n",
    "        if print_stat: \n",
    "            print(f'# [ {key.ljust(15)} {operator.ljust(2)} {str(condition).ljust(6)}]: {str(eval(converted_filter).to_list().count(True)).ljust(6)} ({eval(converted_filter).to_list().count(True)/len(target_df)*100:.1f}%)')\n",
    "    \n",
    "    filtered_df = target_df[eval(\" & \".join([\"(\"+x+\")\" for x in converted_filters]))]\n",
    "    if print_stat:\n",
    "        print(f'# [           ALL            ]: {len(filtered_df)} ({len(filtered_df)/len(target_df)*100:.1f}%)')\n",
    "    return filtered_df\n",
    "\n",
    "\n",
    "filtered_df = filtering_df(dataframe_name+\"_P*Z\", filters_P_Z, target_df)\n",
    "filter_rate = float(len(filtered_df))/float(len(target_df))*100.0\n",
    "\n",
    "print(f'{len(filtered_df)} of {len(target_df)} ({round(filter_rate,3)}%) passed')\n",
    "print()\n",
    "\n",
    "# Display the filtered DataFrame to verify the result\n",
    "if show_df_head:\n",
    "    n_descriptions = 40  # Adjust based on how many descriptions you want to print\n",
    "    filtered_df.sort_values(by='all_cst', ascending=False, inplace=True)\n",
    "    top_descriptions = (filtered_df['description']+\".pdb\").head(n_descriptions).tolist()\n",
    "    print(' '.join(top_descriptions))\n",
    "    print(filtered_df.head(10)[labels_for_print])\n",
    "\n",
    "### SAVE DATAFRAMES ###\n",
    "for subdir, df_name in zip(subdirectories, dataframe_names):\n",
    "    if not df_name in dataframes:\n",
    "        continue\n",
    "    save_path = os.path.join(important_dfs_dir, f'{subdir}_fastmpnn_stats.csv')\n",
    "    dataframes[df_name].to_csv(save_path, index=False)\n",
    "    print(f\"DataFrame {df_name} saved to {save_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1138f25d-014c-40ad-ba64-58ffa0f12e7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial file count in the directory (Should = 0): 0\n",
      "Number of PDBs to Copy: 3168\n",
      "Copying Files to the Directory: ./output/fastMPNNdesign/inpainting/filtered_structures\n",
      "[Wed Apr 23 12:04:29 2025] 1 files copied successfully. Elapsed time: 0.02 seconds. Average time per file: 0.0190 seconds. Estimated time remaining: 10.92 seconds.\n",
      "[Wed Apr 23 12:04:29 2025] 10 files copied successfully. Elapsed time: 0.09 seconds. Average time per file: 0.0091 seconds. Estimated time remaining: 5.16 seconds.\n",
      "[Wed Apr 23 12:04:29 2025] 100 files copied successfully. Elapsed time: 0.93 seconds. Average time per file: 0.0093 seconds. Estimated time remaining: 4.44 seconds.\n",
      "[Wed Apr 23 12:04:38 2025] 1000 files copied successfully. Elapsed time: 9.68 seconds. Average time per file: 0.0097 seconds. Estimated time remaining: 0.00 seconds.\n",
      "Total files copied: 3168\n",
      "\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "#######################################################################\n",
    "### COPY FILTERED STRUCTURES INTO SEPARATE DIRECTORY FOR ALPHAFOLD2 ###\n",
    "#######################################################################\n",
    "\n",
    "### INPUTS ###\n",
    "dataframe_name = \"filtered_df\"\n",
    "input_filtered_structures_dir = f'{fastMPNNdesign_dir}inpainting/'\n",
    "#num 1: f'{fastMPNNdesign_dir}RFD2/'\n",
    "#num 2: f'{fastMPNNdesign_dir}inpainting/'\n",
    "\n",
    "output_filtered_structures_dir = os.path.join(input_filtered_structures_dir, \"filtered_structures\")\n",
    "target_df = eval(dataframe_name)\n",
    "\n",
    "### FUNCTION TO RUN ###s\n",
    "def copy_filtered_fastmpnn_files(df, output_dir):\n",
    "    existing_files = os.listdir(output_dir)\n",
    "    print(f\"Initial file count in the directory (Should = 0): {len(existing_files)}\")\n",
    "    if existing_files:\n",
    "        print('')\n",
    "        print(f\"WARNING: The directory {output_dir} is not empty. It contains {len(existing_files)} files! Consider clearing that entire directory...\")\n",
    "        print('')\n",
    "\n",
    "    descs_to_be_copied = df[\"description\"].to_list()\n",
    "    pdbs_to_be_copied = [os.path.join(input_filtered_structures_dir, x)+\".pdb\" for x in descs_to_be_copied]\n",
    "    seqs_files = os.listdir(os.path.join(input_filtered_structures_dir, \"seqs\"))\n",
    "    for seqs_file in seqs_files:\n",
    "        if seqs_file[:seqs_file.index(\"_T0.\")] in descs_to_be_copied:\n",
    "            pdbs_to_be_copied.append(os.path.join(input_filtered_structures_dir, \"seqs\", seqs_file))\n",
    "    \n",
    "    print(f\"Number of PDBs to Copy: {len(pdbs_to_be_copied)}\")\n",
    "    print(f\"Copying Files to the Directory: {output_dir}\")\n",
    "    \n",
    "    count = 0\n",
    "    start_time = time.time()\n",
    "    error_files = []\n",
    "\n",
    "    for pdb_path in pdbs_to_be_copied:\n",
    "\n",
    "        try:\n",
    "            shutil.copy(pdb_path, output_dir)\n",
    "            count += 1\n",
    "\n",
    "            if np.log10(count) % 1 == 0:\n",
    "                current_time = time.ctime()\n",
    "                elapsed_time = time.time() - start_time\n",
    "                files_left = (len(df)*2 - count)\n",
    "                average_time_per_file = elapsed_time / count\n",
    "                estimated_time_remaining = max(0, average_time_per_file * files_left)\n",
    "                print(f\"[{current_time}] {count} files copied successfully. Elapsed time: {elapsed_time:.2f} seconds. Average time per file: {average_time_per_file:.4f} seconds. Estimated time remaining: {estimated_time_remaining:.2f} seconds.\")\n",
    "\n",
    "        except Exception as e:\n",
    "            error_files.append((pdb_path, str(e)))\n",
    "\n",
    "    print(f\"Total files copied: {count}\")\n",
    "    if error_files:\n",
    "        print(\"Failed to copy the following files:\")\n",
    "        for error in error_files:\n",
    "            print(error)\n",
    "\n",
    "### RUN THE SCRIPT ###\n",
    "os.makedirs(output_filtered_structures_dir, exist_ok=True)\n",
    "copy_filtered_fastmpnn_files(target_df, output_filtered_structures_dir)\n",
    "print('')\n",
    "print('Done!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8ffd9a3-f20e-498e-80a8-4fca974ac41b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# **V. AlphaFold2 (2nd Iteration)**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43924155-c72e-4cdc-9a29-3f5b80573d75",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Run AF2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "91eb8d5e-e079-4864-b259-596178fc0979",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4 --type monomer_ptm --version monomer --nstruct 1 --max_recycles 6 --num_ensemble 2 --mock_msa_depth 1 --recycle_tol 0.0 --pct_seq_mask 0.15  --output_pae  ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_2_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_2_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_5_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_5_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_6_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_7_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_0_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_2_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_3_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_6_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_7_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_1.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_4_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_0.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_5_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.1_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_6_T0.2_4.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_2.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_3.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_7_T0.2_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_4.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_0.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_1.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_2.pdb\n",
      "./containers/crispy.sif {software_dir}superfold/run_superfold.py --out_dir ./output/af2_out/predesigned_RFD2_i2/ --model 4         --type monomer_ptm --version monomer         --nstruct 1 --max_recycles 6 --num_ensemble 2         --mock_msa_depth 1 --recycle_tol 0.0         --pct_seq_mask 0.15  --output_pae          ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_3.pdb ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_9_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_4.pdb\n"
     ]
    }
   ],
   "source": [
    "#######################################\n",
    "### predict designs with AlphaFold2 ###\n",
    "#######################################\n",
    "\n",
    "### CONSTANTS ###\n",
    "apptainer = './containers/crispy.sif' \n",
    "superfold = '{software_dir}superfold/run_superfold.py'\n",
    "\n",
    "### CHANGE THESE EVERY TIME ###\n",
    "af2_inputs_directory = f'{fastMPNNdesign_dir}RFD2/filtered_structures/'\n",
    "#num 1: f'{fastMPNNdesign_dir}RFD2/filtered_structures/'\n",
    "#num 2: f'{fastMPNNdesign_dir}inpainting/filtered_structures/'\n",
    "\n",
    "af2_output_dir = f'{af2_out_dir}predesigned_RFD2_i2/'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2_i2/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting_i2/'\n",
    "os.makedirs(af2_output_dir, exist_ok=True)\n",
    "\n",
    "### ALPHAFOLD PARAMETERS ###\n",
    "model_number = 4 # 4 is recommended, 5 is probably similar\n",
    "model_type = 'monomer_ptm'\n",
    "version = 'monomer'\n",
    "\n",
    "nstruct = 1 # how many structures to generate as outputs, default = 1\n",
    "num_recycles = 6 # default = 3\n",
    "num_ensemble = 2 # default in code is 1 but CASP competition used 8\n",
    "mock_msa_depth = 1 # default is 1, but higher might = better ? goes to +inf\n",
    "recycle_tol = 0.0 # default is 0.0, what convergence rmsd to stop at after recycling\n",
    "\n",
    "pct_seq_mask = 0.15 # 0.15 is default, percent of sequence to make during inference\n",
    "output_summary = '' #'--output_summary' # ENABLE IF YES, otherwise ''\n",
    "output_pae = '--output_pae' # ENABLE IF YES, otherwise ''\n",
    "amber_relax = '' #'--amber_relax' # ENABLE IF YES, otherwise ''\n",
    "initial_guess = '' #'--initial_guess {pdb}' # ENABLE IF YES, otherwise ''\n",
    "\n",
    "### MAKE THE FUNCTION ###\n",
    "cmd = f'{apptainer} {superfold} --out_dir {af2_output_dir} --model {model_number} --type {model_type} --version {version} \\\n",
    "--nstruct {nstruct} --max_recycles {num_recycles} --num_ensemble {num_ensemble} --mock_msa_depth {mock_msa_depth} --recycle_tol {recycle_tol} \\\n",
    "--pct_seq_mask {pct_seq_mask} {output_summary} {output_pae} {amber_relax}' #{initial_guess}'\n",
    "\n",
    "for i, pdb in enumerate(glob.glob(os.path.join(af2_inputs_directory, f\"*.pdb\"))): \n",
    "    if i == 0 or i % 5 != 0:\n",
    "        cmd += f\" {pdb}\"\n",
    "    else:\n",
    "        print(cmd)\n",
    "        cmd = f'{apptainer} {superfold} --out_dir {af2_output_dir} --model {model_number} \\\n",
    "        --type {model_type} --version {version} \\\n",
    "        --nstruct {nstruct} --max_recycles {num_recycles} --num_ensemble {num_ensemble} \\\n",
    "        --mock_msa_depth {mock_msa_depth} --recycle_tol {recycle_tol} \\\n",
    "        --pct_seq_mask {pct_seq_mask} {output_summary} {output_pae} {amber_relax} \\\n",
    "        {pdb}' # {initial_guess} {pdb}\n",
    "print(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e9921e-166e-4db0-ac13-32616e94da8b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## AF2 Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "625166b2-a607-4a50-8e14-fb7d77e188d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SUBMIT THIS:\n",
      "python {scripts_dir}af2_parse_processing_multi.py ./output/af2_out/predesigned_RFD2_i2/ --cpus 19\n",
      "\n",
      "OUTPUT CSV FILE HERE:\n",
      "./output/af2_out/predesigned_RFD2_i2/af2_out_parsed.csv\n"
     ]
    }
   ],
   "source": [
    "##################################################################\n",
    "### PARSE & PROCESS AF2 OUTPUT FILES ONLY NEED TO DO THIS ONCE ###\n",
    "##################################################################\n",
    "\n",
    "# INPUT DIRECTORY TO PARSE\n",
    "af2_dir = f'{af2_out_dir}predesigned_RFD2_i2/'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2_i2/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting_i2/'\n",
    "\n",
    "# Constants\n",
    "script = f\"{scripts_dir}af2_parse_processing_multi.py\"\n",
    "optional_cpus = 19  # Optional: specify the number of CPUs to use, or remove this variable to use default logic\n",
    "\n",
    "# Make command\n",
    "command = f\"python {script} {af2_dir}\"\n",
    "if 'optional_cpus' in locals():\n",
    "    command += f\" --cpus {optional_cpus}\"\n",
    "\n",
    "# Print submission\n",
    "print(\"SUBMIT THIS:\")\n",
    "print(command)\n",
    "print('')\n",
    "\n",
    "# Print output .csv path\n",
    "print(\"OUTPUT CSV FILE HERE:\")\n",
    "print(af2_dir + 'af2_out_parsed.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9c4f128-3d09-49c3-8775-d1f0bd23258e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Filter AF2 structures based on Global Metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b74dd750-fd27-4b4e-a51d-707b7e11bd1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21 of 51 (41.176%) with plddt >= 60 and CŒ± rmsd <=3\n",
      "Initial file count in the directory (Should = 0): 0\n",
      "Total files copied: 42\n",
      "\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "############################################\n",
    "### LOAD AND FILTER AF2 BY SUMMARY STATS ###\n",
    "############################################\n",
    "\n",
    "# Correct the path to where your actual CSV file is located\n",
    "csv_file_path = f'{af2_out_dir}predesigned_RFD2_i2/af2_out_parsed.csv'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2_i2/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting_i2/'\n",
    "\n",
    "output_filtered_structures_dir = f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2/filtered_structures_i2/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting/filtered_structures_i2/'\n",
    "\n",
    "# Select parameters for filtering -> EXTREMELY HARSH IN THIS CASE\n",
    "plddt_cutoff = 60 # in pLDDT units -> usually 70-80\n",
    "rmsd_cutoff = 3 # √Ö, I think this is CŒ± (1.5 is typical, 2.0 might be a stretch but okay for big structures)\n",
    "pae_cutoff = 100 # below 5 is harsh but good\n",
    "pTM_cutoff = 0.5 # Supposedly above 0.75 is good\n",
    "convergence_cutoff = 1.5\n",
    "\n",
    "# Load the CSV file into a DataFrame\n",
    "af2_scores_df = pd.read_csv(csv_file_path)\n",
    "\n",
    "af2_mask = ((af2_scores_df['mean_plddt'] >= plddt_cutoff) & \n",
    "            (af2_scores_df['rmsd_to_input'] <= rmsd_cutoff) &\n",
    "            (af2_scores_df['mean_pae'] <= pae_cutoff) & \n",
    "            (af2_scores_df['pTMscore'] >= pTM_cutoff) &\n",
    "            (af2_scores_df['af2_convergence_tol'] <= convergence_cutoff)\n",
    "           )\n",
    "\n",
    "filtered_af2_scores_df = af2_scores_df[af2_mask]\n",
    "af2_filter_rate = float(len(filtered_af2_scores_df))/float(len(af2_scores_df))*100.0\n",
    "print(f'{len(filtered_af2_scores_df)} of {len(af2_scores_df)} ({round(af2_filter_rate,3)}%) with plddt >= {plddt_cutoff} and CŒ± rmsd <={rmsd_cutoff}')\n",
    "\n",
    "### FUNCTION TO RUN ###\n",
    "def copy_files(df, pdb_col, json_col, output_dir):\n",
    "    existing_files = os.listdir(output_dir)\n",
    "    print(f\"Initial file count in the directory (Should = 0): {len(existing_files)}\")\n",
    "\n",
    "    count = 0\n",
    "    start_time = time.time()\n",
    "    error_files = []\n",
    "\n",
    "    for index, row in df.iterrows():\n",
    "        pdb_path = row[pdb_col]\n",
    "        json_path = row[json_col]\n",
    "\n",
    "        try:\n",
    "            shutil.copy(pdb_path, output_dir)\n",
    "            shutil.copy(json_path, output_dir)\n",
    "            count += 2  # Increment for each PDB and JSON file copied\n",
    "\n",
    "            if count % 2000 == 0:\n",
    "                current_time = time.ctime()\n",
    "                elapsed_time = time.time() - start_time\n",
    "                files_left = (len(df)*2 - count)\n",
    "                average_time_per_file = elapsed_time / count\n",
    "                estimated_time_remaining = max(0, average_time_per_file * files_left)\n",
    "                print(f\"[{current_time}] {count} files copied successfully. Elapsed time: {elapsed_time:.2f} seconds. Average time per file: {average_time_per_file:.4f} seconds. Estimated time remaining: {estimated_time_remaining:.2f} seconds.\")\n",
    "\n",
    "        except Exception as e:\n",
    "            error_files.append((pdb_path, json_path, str(e)))\n",
    "\n",
    "    print(f\"Total files copied: {count}\")\n",
    "    if error_files:\n",
    "        print(\"Failed to copy the following files:\")\n",
    "        for error in error_files:\n",
    "            print(error)\n",
    "\n",
    "### RUN THE SCRIPT ###\n",
    "os.makedirs(output_filtered_structures_dir, exist_ok=True)\n",
    "copy_files(filtered_af2_scores_df, 'pdb_path', 'af2_json_path', output_filtered_structures_dir)\n",
    "print('')\n",
    "print('Done!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22944ca3-f0b4-43a9-8e82-dfe646e90a54",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Analyze AF2 on Side Chain Metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e49e6f5-6c12-4cd6-8b51-42f5e209b2e9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Run Both of These Once to Generate Score Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bfb72134-6a7a-4e45-86c6-34cbe57802be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SUBMIT THIS:\n",
      "python {scripts_dir}af2_parse_processing_multi.py ./output/af2_out/predesigned_RFD2_i2/filtered_structures/ --cpus 19 --optional_basename_for_summary_stats scores.sc\n",
      "\n",
      "OUTPUT CSV FILE HERE:\n",
      "./output/af2_out/predesigned_RFD2_i2/filtered_structures/scores.sc\n"
     ]
    }
   ],
   "source": [
    "##################################################################\n",
    "### PARSE & PROCESS AF2 OUTPUT FILES ONLY NEED TO DO THIS ONCE ###\n",
    "##################################################################\n",
    "\n",
    "# INPUT DIRECTORY TO PARSE\n",
    "af2_directory_to_parse = f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting_i2/filtered_structures/'\n",
    "\n",
    "# Constants\n",
    "script = f\"{scripts_dir}af2_parse_processing_multi.py\"\n",
    "optional_cpus = 19  # Optional: specify the number of CPUs to use, or remove this variable to use default logic\n",
    "optional_basename_for_summary_stats = \"scores.sc\"\n",
    "\n",
    "# Make command\n",
    "command = f\"python {script} {af2_directory_to_parse}\"\n",
    "if 'optional_cpus' in locals():\n",
    "    command += f\" --cpus {optional_cpus}\"\n",
    "if 'optional_basename_for_summary_stats' in locals():\n",
    "    command += f\" --optional_basename_for_summary_stats {optional_basename_for_summary_stats}\"\n",
    "\n",
    "# Print submission\n",
    "print(\"SUBMIT THIS:\")\n",
    "print(command)\n",
    "print('')\n",
    "\n",
    "# Print output .csv path\n",
    "print(\"OUTPUT CSV FILE HERE:\")\n",
    "print(af2_directory_to_parse + f'{optional_basename_for_summary_stats}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f2530363-b35c-4169-90a3-a07119ae5a97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command to execute:\n",
      "python {scripts_dir}sidechain_rmsd_and_info_af2_matching_res.py --scorefile ./output/af2_out/predesigned_RFD2_i2/filtered_structures/scores.sc --ref_pdbs_dir ./output/fastMPNNdesign/RFD2/filtered_structures/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --output_file ./output/af2_out/predesigned_RFD2_i2/filtered_structures/updated_scores.sc --max_possible_num_of_suffixes 1 --atom_groups '{\"HIS\": [{\"atoms\": [\"N\", \"CA\", \"C\", \"O\"], \"label\": \"bb_rmsd\"}, {\"atoms\": [\"N\", \"CA\", \"C\", \"O\", \"CB\"], \"label\": \"cb_bb_rmsd\"}, {\"atoms\": [\"CG\", \"CE1\", \"ND1\", \"NE2\", \"CD2\"], \"label\": \"imidazole_rmsd\"}, {\"atoms\": [\"CA\", \"CB\"], \"label\": \"ca_cb_rmsd\"}, {\"atoms\": [\"ND1\", \"NE2\"], \"label\": \"imid_sc_nitrogen_rmsd\"}]}'\n",
      "\n",
      "New scorefile:\n",
      "./output/af2_out/predesigned_RFD2_i2/filtered_structures/updated_scores.sc\n"
     ]
    }
   ],
   "source": [
    "##################################################################################################\n",
    "### CALCULATE SIDE CHAIN RMSDS & OTHER RMSDS FOR CATALYTIC RESIDUES IN FILTERED STRUCTURES DIR ###\n",
    "##################################################################################################\n",
    "\n",
    "### INPUT SCORE PATH + OUTPUT SCORE PATH & REFERENCE FILES ###\n",
    "score_path = os.path.join(f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures', 'scores.sc')\n",
    "output_file = os.path.join(f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures', 'updated_scores.sc')\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting_i2/filtered_structures'\n",
    "\n",
    "ref_pdbs_dir = os.path.join(f'{fastMPNNdesign_dir}RFD2', 'filtered_structures/')\n",
    "#num 1: f'{fastMPNNdesign_dir}RFD2/filtered_structures/'\n",
    "#num 2: f'{fastMPNNdesign_dir}inpainting/filtered_structures/'\n",
    "\n",
    "### INPUT PARAMS ###\n",
    "params_path = os.path.join(invrot_dir, \"COMBINED_PHENYLACETATE_HHH\", \"COMBINED_PARAMS\", \"*.params\")  # Ensure params include .pdb if needed\n",
    "\n",
    "### INPUT ATOM RMSDS OF CAT RESIDUES TO CALCULATE & NUM OF SUFFIXES ###\n",
    "atom_groups_json = { #MANUALLY INPUT POSSIBLE CATALYTIC RESIDUES!!\n",
    "    \"HIS\": [\n",
    "        {\"atoms\": [\"N\", \"CA\", \"C\", \"O\"], \"label\": \"bb_rmsd\"},\n",
    "        {\"atoms\": [\"N\", \"CA\", \"C\", \"O\", \"CB\"], \"label\": \"cb_bb_rmsd\"},\n",
    "        {\"atoms\": [\"CG\", \"CE1\", \"ND1\", \"NE2\", \"CD2\"], \"label\": \"imidazole_rmsd\"},\n",
    "        {\"atoms\": [\"CA\", \"CB\"], \"label\": \"ca_cb_rmsd\"},\n",
    "        {\"atoms\": [\"ND1\", \"NE2\"], \"label\": \"imid_sc_nitrogen_rmsd\"}\n",
    "    ]\n",
    "}\n",
    "\n",
    "max_possible_num_of_suffixes = 1\n",
    "column_name_for_pdb_path = 'pdb_path'\n",
    "\n",
    "### CONSTANTS ###\n",
    "script = '{scripts_dir}sidechain_rmsd_and_info_af2_matching_res.py'\n",
    "atom_groups = json.dumps(atom_groups_json)\n",
    "\n",
    "# Build the command\n",
    "command = (\n",
    "    f'python {script} --scorefile {score_path} --ref_pdbs_dir {ref_pdbs_dir} --params {params_path} --output_file {output_file} '\n",
    "    f'--max_possible_num_of_suffixes {max_possible_num_of_suffixes} --atom_groups \\'{atom_groups}\\''\n",
    ")\n",
    "print(\"Command to execute:\")\n",
    "print(command)\n",
    "print('')\n",
    "print('New scorefile:')\n",
    "print(f'{output_file}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74bc48da-9c56-4f57-819e-28f4ba8080de",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Do a Light Side Chain Filter Before Global Alignment & Harsh SC Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4f0b88f2-a0b9-4437-89a9-51658c845569",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################################################\n",
    "### COMPUTE SIDE CHAIN STATISTICS FROM SCORES DF & DO GENTLE/LOGICAL FILTERING ###\n",
    "##################################################################################\n",
    "\n",
    "### INPUT PATH ###\n",
    "scores_filepath = f\"{af2_out_dir}predesigned_RFD2_i2/filtered_structures/updated_scores.sc\"\n",
    "inpaint_or_diffusion = \"RFD2\" #RFD2 or Inpainting\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/updated_scores.sc'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting_i2/filtered_structures/updated_scores.sc'\n",
    "\n",
    "### INPUT SPECIFY COLUMN GROUPS FOR CALCULATION ###\n",
    "column_groups = {\n",
    "    'cat_HIS_bb': ['cat_HIS1_bb_rmsd', 'cat_HIS2_bb_rmsd', 'cat_HIS3_bb_rmsd'],\n",
    "    'cat_HIS_cb_bb': ['cat_HIS1_cb_bb_rmsd', 'cat_HIS2_cb_bb_rmsd', 'cat_HIS3_cb_bb_rmsd'],\n",
    "    'cat_HIS_ca_cb_ONLY': ['cat_HIS1_ca_cb_rmsd', 'cat_HIS2_ca_cb_rmsd', 'cat_HIS3_ca_cb_rmsd'],\n",
    "    'cat_HIS_imidazole': ['cat_HIS1_imidazole_rmsd', 'cat_HIS2_imidazole_rmsd', 'cat_HIS3_imidazole_rmsd'],\n",
    "    'cat_HIS_imidazole_tip_Ns': ['cat_HIS1_imid_sc_nitrogen_rmsd', 'cat_HIS2_imid_sc_nitrogen_rmsd', 'cat_HIS3_imid_sc_nitrogen_rmsd']\n",
    "}\n",
    "\n",
    "### FUNCTION TO CALCULATE MEAN RMSD AND RMS OF RMSDs FOR SPECIFIED COLUMN GROUPS ###\n",
    "def calculate_group_rmsd(df, column_groups):\n",
    "    for group_name, columns in column_groups.items():\n",
    "        mean_col_name = f\"{group_name}_mean_rmsd\"\n",
    "        rms_col_name = f\"{group_name}_rms_of_rmsds\"\n",
    "        df[mean_col_name] = df[columns].mean(axis=1)\n",
    "        df[rms_col_name] = np.sqrt((df[columns]**2).mean(axis=1))\n",
    "\n",
    "### EXECUTE PLOTTING AND PRINTING ###\n",
    "sidechain_rmsd_df = pd.read_csv(scores_filepath, sep=',', na_values='')\n",
    "calculate_group_rmsd(sidechain_rmsd_df, column_groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fa495ae7-e01f-42e9-a0c5-46db5d7f99b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########################## SIDE CHAIN RMSD METRICS FILTER ##########################\n",
      "15 of 21 (71.429%) have CATALYTIC HIS BB RMS OF RMSDS <= 0.85\n",
      "14 of 21 (66.667%) have CATALYTIC HIS CB BB RMS OF RMSDS <= 0.85\n",
      "14 of 21 (66.667%) have CATALYTIC HIS CA CB ONLY RMS OF RMSDS <= 0.85\n",
      "0 of 21 (0.000%) have CATALYTIC HIS IMIDAZOLE RMS OF RMSDS <= 1.25\n",
      "0 of 21 (0.000%) have CATALYTIC HIS IMIDAZOLE TIP NS RMS OF RMSDS <= 1.25\n",
      "\n",
      "16 of 21 (76.190%) pass at least one filter!\n",
      "####################################################################################\n",
      "Columns used in filtering: cat_HIS_bb_rms_of_rmsds, cat_HIS_cb_bb_rms_of_rmsds, cat_HIS_ca_cb_ONLY_rms_of_rmsds, cat_HIS_imidazole_rms_of_rmsds, cat_HIS_imidazole_tip_Ns_rms_of_rmsds\n"
     ]
    }
   ],
   "source": [
    "##################################################################################\n",
    "### COMPUTE SIDE CHAIN STATISTICS FROM SCORES DF & DO GENTLE/LOGICAL FILTERING ###\n",
    "##################################################################################\n",
    "\n",
    "### NEW FILTER FROM SIDE CHAIN 2 ###\n",
    "cat_HIS_bb_rms_of_rmsds_cutoff = 0.85 #2 # √Ö\n",
    "cat_HIS_cb_bb_rms_of_rmsds_cutoff = 0.85 # √Ö\n",
    "cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff = 0.85 # √Ö\n",
    "cat_HIS_imidazole_rms_of_rmsds_cutoff = 1.25 # √Ö\n",
    "cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff = 1.25 # √Ö\n",
    "\n",
    "# Individual filter calculations with specific naming\n",
    "filter_masks = {\n",
    "    'cat_HIS_bb_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_bb_rms_of_rmsds'] <= cat_HIS_bb_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_cb_bb_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_cb_bb_rms_of_rmsds'] <= cat_HIS_cb_bb_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_ca_cb_ONLY_rms_of_rmsds'] <= cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_imidazole_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_imidazole_rms_of_rmsds'] <= cat_HIS_imidazole_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask': sidechain_rmsd_df['cat_HIS_imidazole_tip_Ns_rms_of_rmsds'] <= cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff\n",
    "}\n",
    "\n",
    "# Calculate pass counts and rates for each filter\n",
    "filter_results = {filter_name: (mask.sum(), len(sidechain_rmsd_df), mask.mean() * 100) for filter_name, mask in filter_masks.items()}\n",
    "\n",
    "# Overall filtering result using OR logic\n",
    "cumulative_filter_mask_or = filter_masks['cat_HIS_bb_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_cb_bb_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_imidazole_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask']\n",
    "\n",
    "cumulative_pass_count_or = cumulative_filter_mask_or.sum()\n",
    "cumulative_total_or = len(sidechain_rmsd_df)\n",
    "cumulative_pass_rate_or = cumulative_pass_count_or / cumulative_total_or * 100\n",
    "\n",
    "### DO THE PRINT STATEMENT (FANCY) ###\n",
    "print('########################## SIDE CHAIN RMSD METRICS FILTER ##########################')\n",
    "for filter_name, (pass_count, total, pass_rate) in filter_results.items():\n",
    "    specific_criteria = filter_name.replace('_mask', '').replace('_', ' ').replace('cat', 'catalytic').upper()\n",
    "    cutoff_value = {\n",
    "        'cat_HIS_bb_rms_of_rmsds_mask': cat_HIS_bb_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_cb_bb_rms_of_rmsds_mask': cat_HIS_cb_bb_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask': cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_imidazole_rms_of_rmsds_mask': cat_HIS_imidazole_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask': cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff\n",
    "    }[filter_name]\n",
    "    print(f'{pass_count} of {total} ({pass_rate:.3f}%) have {specific_criteria} <= {cutoff_value}')\n",
    "\n",
    "# Maintaining the cumulative print statement from the provided script\n",
    "print('')\n",
    "print(f'{cumulative_pass_count_or} of {cumulative_total_or} ({cumulative_pass_rate_or:.3f}%) pass at least one filter!')\n",
    "print('####################################################################################')\n",
    "\n",
    "### MAKE THE FINAL DATA FRAME ###\n",
    "final_filtered_df = sidechain_rmsd_df[cumulative_filter_mask_or]\n",
    "\n",
    "### PRINT OUT COLUMN NAMES USED IN FILTERING ###\n",
    "filtered_columns = ', '.join([key.replace('_mask', '') for key in filter_masks.keys()])\n",
    "print(f'Columns used in filtering: {filtered_columns}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "348836a3-522f-426c-a19f-e6a9c686a438",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Apo Structure Relax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "eb0f9951-bf6a-47f7-b1bb-47c16459f56c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_3_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_1_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_3_T0.1_2_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_0_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.1_3_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_8_T0.2_2_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_9_T0.1_4_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_0_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n",
      "python {scripts_dir}pyfastrelax_and_hbond_APO.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_1_model_4_ptm_seed_0_unrelaxed_PSZ_0_4_T0.1_2_model_4_ptm_seed_0_unrelaxed.pdb --out_dir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/\n"
     ]
    }
   ],
   "source": [
    "#######################################################\n",
    "### RUN PYFAST RELAX W A BUNCH OF STAT CALCULATIONS ### \n",
    "### (MOSTLY JUST SIDE CHAIN PACKING)                ###  \n",
    "#######################################################\n",
    "\n",
    "### GENERATION OF RELAX (USUALLY ONLY INPUT) ###\n",
    "\n",
    "# Regulars, do not change much\n",
    "script = '{scripts_dir}pyfastrelax_and_hbond_APO.py'\n",
    "\n",
    "# Change Only if Necessary | INPUT DIR #mpnn_out_dir & packed/---> (OLD GENERATION, ASSUMES PREVIOUS Gen)\n",
    "input_directory_pyfast_relax = f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/'\n",
    "#num1: f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/' \n",
    "#num2: f'{af2_out_dir}predesigned_inpainting_i2/filtered_structures/' \n",
    "\n",
    "# Change Only if Necessary | OUTPUT RELAX DIRECTORY\n",
    "pyfast_relax_out_directory = f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/APO_RELAXED/'\n",
    "#num1: f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/APO_RELAXED/'\n",
    "#num2: f'{af2_out_dir}predesigned_inpainting_i2/filtered_structures/APO_RELAXED/'\n",
    "os.makedirs(pyfast_relax_out_directory, exist_ok=True)\n",
    "\n",
    "# Change Only if Necessary | Generation Format Should Update Accordingly\n",
    "for pdb in glob.glob(f'{input_directory_pyfast_relax}*pdb'):  \n",
    "    cmd_txt = (f'python {script} '\n",
    "               f'--pdb {pdb} '\n",
    "               f'--out_dir {pyfast_relax_out_directory}')\n",
    "    print (cmd_txt) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13e43187-5cd0-44e7-a946-fcc6f2127258",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COMMAND:\n",
      "python {scripts_dir}pyfastrelax_stat_processingAPO.py ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/ --nproc 300\n",
      "\n",
      "OUTPUT FILE:\n",
      "./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/parsed_energy_data.csv\n"
     ]
    }
   ],
   "source": [
    "###############################\n",
    "### ANALYZE RELAXED OUTPUTS ###\n",
    "###############################\n",
    "\n",
    "# Define the constant variables\n",
    "script_path = f\"{scripts_dir}pyfastrelax_stat_processingAPO.py\"\n",
    "\n",
    "# CHANGE EVERY TIME \n",
    "part_diffusion_relax_out_directory_path = f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/APO_RELAXED/'\n",
    "#num1: f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/APO_RELAXED/'\n",
    "#num2: f'{af2_out_dir}predesigned_inpainting_i2/filtered_structures/APO_RELAXED/'\n",
    "number_of_cores = '--nproc 10'\n",
    "\n",
    "# Format the command string\n",
    "command = f\"python {script_path} {part_diffusion_relax_out_directory_path} {number_of_cores}\"\n",
    "\n",
    "# Print the command\n",
    "print('COMMAND:')\n",
    "print(command)\n",
    "print('')\n",
    "print('OUTPUT FILE:')\n",
    "print(f'{part_diffusion_relax_out_directory_path}parsed_energy_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b96b3b97-acd3-43bb-b83a-d1c1b5fda6e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rows matching criteria: 13 out of 13 total rows\n"
     ]
    }
   ],
   "source": [
    "#############################################################\n",
    "### FILTER THE DATA FRAME BASED ON THE CRITERIA ARE IDEAL ###\n",
    "#############################################################\n",
    "\n",
    "# Dynamically create file path and DataFrame names using the generation number\n",
    "file_path = f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/APO_RELAXED/parsed_energy_data.csv'\n",
    "#num 1: f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/APO_RELAXED/parsed_energy_data.csv'\n",
    "#num 2: f'{af2_out_dir}predesigned_inpainting_i2/filtered_structures/APO_RELAXED/parsed_energy_data.csv'\n",
    "\n",
    "relaxed_stats_APO_df = pd.read_csv(file_path)\n",
    "\n",
    "### FILTER THE DATA FRAME ###\n",
    "filtered_relaxed_stats_APO_df = relaxed_stats_APO_df[\n",
    "    ((relaxed_stats_APO_df['sap_score'] <= 75) |\n",
    "    (relaxed_stats_APO_df['net_charge_in_design_not_w_his'].abs() >= 20)) &\n",
    "    (relaxed_stats_APO_df['hydrophobic_exposure_sasa_in_design'] <= 1500)\n",
    "]\n",
    "\n",
    "# Print the count of rows in the filtered DataFrame and the total in the original DataFrame\n",
    "print(f\"Rows matching criteria: {filtered_relaxed_stats_APO_df.shape[0]} out of {relaxed_stats_APO_df.shape[0]} total rows\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a2d3f4d-d74f-4c3c-9402-a2a465137849",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Copy Ligand Coords into Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8f5e0054-d998-4511-a028-7e4580daabdc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13 rows processed, 0 remaining. 737 ref pdb files available for matching. Elapsed time: 0.00 s. Estimated time left: 0.00 s.\n",
      "\n",
      "Run below commands:\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_T0.2_3_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n",
      "python {scripts_dir}align_af2_with_inputs_and_copy_ligand.py --pdb ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb --ref ./output/fastMPNNdesign/RFD2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1.pdb --outdir ./output/af2_out/predesigned_RFD2_i2/filtered_structures/alignment/ --params ./input/invrot/COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/*.params --fix_catres\n"
     ]
    }
   ],
   "source": [
    "##############################################\n",
    "### GENERATE COMMANDS FOR ALIGNING LIGANDS ###\n",
    "##############################################\n",
    "\n",
    "import time\n",
    "\n",
    "### INPUTS ###\n",
    "filtered_af2_sc_AND_global_scores_df = filtered_relaxed_stats_APO_df # INPUT DF\n",
    "column_of_pdb_paths = 'pdb_path' # COLUMN OF DF W/PDB PATHS\n",
    "\n",
    "ref_pdbs_dir = os.path.join(fastMPNNdesign_dir, 'RFD2', 'filtered_structures/')\n",
    "#num 1: RFD2\n",
    "#num 2: inpainting\n",
    "\n",
    "params_path = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/\"\n",
    "\n",
    "align_af2_out_directory = os.path.join(f\"{af2_out_dir}\", \"predesigned_RFD2_i2\", \"filtered_structures\", \"alignment/\")\n",
    "# num1: predesigned_RFD2_i2\n",
    "# num2: predesigned_inpainting_i2\n",
    "\n",
    "### CONSTANTS ###\n",
    "align_script = f\"{scripts_dir}align_af2_with_inputs_and_copy_ligand.py\"\n",
    "\n",
    "### REFERENCE PDB PROCESSING ###\n",
    "ref_pdb_files = glob.glob(os.path.join(ref_pdbs_dir, '*.pdb'))\n",
    "ref_pdb_map = {os.path.basename(p).replace('.pdb', ''): p for p in ref_pdb_files}\n",
    "ref_pdb_usage = Counter()\n",
    "\n",
    "### AUTOMATED COMMAND WRITING & REFERENCE MATCHING ###\n",
    "os.makedirs(align_af2_out_directory, exist_ok=True)\n",
    "max_usages_per_ref_pdb = 30  # Number of times a reference pdb can be used\n",
    "\n",
    "total_rows = len(filtered_af2_sc_AND_global_scores_df)\n",
    "rows_processed = 0\n",
    "update_interval = 1000\n",
    "\n",
    "start_time = time.time()\n",
    "cmds = []\n",
    "for index, row in filtered_af2_sc_AND_global_scores_df.iterrows():\n",
    "    pdb_path = row[column_of_pdb_paths]\n",
    "    pdb_name = os.path.basename(pdb_path).split('.')[0]\n",
    "\n",
    "    # Find matching reference PDB\n",
    "    matching_ref = None\n",
    "    for ref_name, ref_path in ref_pdb_map.items():\n",
    "        if pdb_name.startswith(ref_name) and ref_pdb_usage[ref_name] < max_usages_per_ref_pdb:\n",
    "            matching_ref = ref_path\n",
    "            ref_pdb_usage[ref_name] += 1\n",
    "            if ref_pdb_usage[ref_name] >= max_usages_per_ref_pdb:\n",
    "                del ref_pdb_map[ref_name]  # Remove from future considerations\n",
    "            break\n",
    "\n",
    "    if matching_ref:\n",
    "        command_align = (f\"python {align_script} \"\n",
    "                         f\"--pdb {pdb_path} \"\n",
    "                         f\"--ref {matching_ref} \"\n",
    "                         f\"--outdir {align_af2_out_directory} \"\n",
    "                         f\"--params {os.path.join(params_path, '*.params')} \"\n",
    "                         f\"--fix_catres\")\n",
    "        cmds.append(command_align)\n",
    "\n",
    "    rows_processed += 1\n",
    "    if rows_processed % update_interval == 0 or rows_processed == total_rows:\n",
    "        elapsed_time = time.time() - start_time\n",
    "        time_per_row = elapsed_time / rows_processed\n",
    "        estimated_time_remaining = time_per_row * (total_rows - rows_processed)\n",
    "        print(f\"{rows_processed} rows processed, {total_rows - rows_processed} remaining. \"\n",
    "              f\"{len(ref_pdb_map)} ref pdb files available for matching. \"\n",
    "              f\"Elapsed time: {elapsed_time:.2f} s. Estimated time left: {estimated_time_remaining:.2f} s.\")\n",
    "print('')\n",
    "print(\"Run below commands:\")\n",
    "print(\"\\n\".join(cmds))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fc9eeae-ba01-4db7-af24-8c4090ee9376",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Sequence Clustering & Filtering Down"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1e3b002e-8cde-416c-bc68-97509fa26694",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AF2 DataFrame before merge:\n",
      "(21, 30)\n",
      "\n",
      "APO Relax DataFrame before merge:\n",
      "(13, 19)\n",
      "\n",
      "Merged DataFrame:\n",
      "(13, 48)\n",
      "\n",
      "Total number of rows in the merged DataFrame: 13\n",
      "Number of unique groups: 8\n",
      "Number of groups with more than 3 rows: 0 (Total rows: 0)\n",
      "Number of groups with more than 5 rows: 0 (Total rows: 0)\n",
      "Number of groups with more than 7 rows: 0 (Total rows: 0)\n",
      "\n",
      "Number of rows in each group:\n",
      "12 of 13 (92.308%) with plddt >= 72.5, CŒ± rmsd <= 1.75√Ö, convergence <= 0.75√Ö, pae <= 7.25, and pTM >= 0.725\n",
      "\n",
      "########################## SIDE CHAIN RMSD METRICS FILTER ##########################\n",
      "11 of 12 (91.667%) have CATALYTIC HIS BB RMS OF RMSDS <= 0.85\n",
      "10 of 12 (83.333%) have CATALYTIC HIS CB BB RMS OF RMSDS <= 0.85\n",
      "10 of 12 (83.333%) have CATALYTIC HIS CA CB ONLY RMS OF RMSDS <= 0.85\n",
      "0 of 12 (0.000%) have CATALYTIC HIS IMIDAZOLE RMS OF RMSDS <= 1.25\n",
      "0 of 12 (0.000%) have CATALYTIC HIS IMIDAZOLE TIP NS RMS OF RMSDS <= 1.25\n",
      "\n",
      "12 of 12 (100.000%) pass at least one filter!\n",
      "####################################################################################\n",
      "Columns used in filtering: cat_HIS_bb_rms_of_rmsds, cat_HIS_cb_bb_rms_of_rmsds, cat_HIS_ca_cb_ONLY_rms_of_rmsds, cat_HIS_imidazole_rms_of_rmsds, cat_HIS_imidazole_tip_Ns_rms_of_rmsds\n",
      "\n",
      "METRICS FOR CHECKING WITH BELOW CALCULATIONS/LOGIC:\n",
      "Number of unique groups: 7 (Total rows: 12)\n",
      "Number of groups with 3+ rows: 1 (Total rows: 3)\n",
      "Number of groups with more than 4+ rows: 0 (Total rows: 0)\n",
      "\n",
      "\n",
      "Number of structures before top 3 filtering: 12\n",
      "Number of structures after top 3 filtering: 12\n",
      "Number of structures filtered out: 0\n",
      "\n",
      "Number of unique groups: 7\n",
      "Number of groups with 3 rows: 1 (Total rows: 3)\n",
      "Number of groups with more than 3 rows: 0 (Total rows: 0)\n",
      "\n",
      "Number of unique fastmpnn groups: 1\n",
      "                                                                                                                                                                                    af2_json_path_af2_out  \\\n",
      "0         ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_model_4_ptm_seed_0_prediction_results.json   \n",
      "1  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3_model_4_ptm_seed_0_prediction_results.json   \n",
      "2  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0_model_4_ptm_seed_0_prediction_results.json   \n",
      "3  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2_model_4_ptm_seed_0_prediction_results.json   \n",
      "4  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2_model_4_ptm_seed_0_prediction_results.json   \n",
      "\n",
      "                                                                                                                                                                               pdb_path_af2_out  \\\n",
      "0         ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "1  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "2  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "3  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "4  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "\n",
      "   mean_plddt_af2_out  rmsd_to_input_af2_out  mean_pae_intra_chain_af2_out  \\\n",
      "0           88.812935               1.467121                      4.755084   \n",
      "1           90.589630               0.769541                      4.312035   \n",
      "2           89.767685               1.196374                      4.476934   \n",
      "3           86.710754               1.101084                      5.225068   \n",
      "4           89.375534               1.745819                      4.608910   \n",
      "\n",
      "   mean_pae_af2_out  pTMscore_af2_out  af2_convergence_tol_af2_out  \\\n",
      "0          4.775782          0.857008                     0.210543   \n",
      "1          4.346132          0.858776                     0.074986   \n",
      "2          4.506941          0.858870                     0.069758   \n",
      "3          5.301823          0.832086                     0.121731   \n",
      "4          4.667342          0.846754                     0.137402   \n",
      "\n",
      "   af2_elapsed_folding_time_af2_out  cat_HIS1_bb_rmsd_af2_out  \\\n",
      "0                         38.773224                  0.740811   \n",
      "1                        191.917191                  0.632856   \n",
      "2                         38.735088                  0.563488   \n",
      "3                         38.667376                  0.510518   \n",
      "4                         38.836154                  0.823901   \n",
      "\n",
      "   cat_HIS1_cb_bb_rmsd_af2_out  cat_HIS1_imidazole_rmsd_af2_out  \\\n",
      "0                     0.685227                         1.172572   \n",
      "1                     0.669494                         2.692293   \n",
      "2                     0.569319                         3.403390   \n",
      "3                     0.526570                         3.132519   \n",
      "4                     0.847165                         3.066490   \n",
      "\n",
      "   cat_HIS1_ca_cb_rmsd_af2_out  cat_HIS1_imid_sc_nitrogen_rmsd_af2_out  \\\n",
      "0                     0.484510                                1.271909   \n",
      "1                     0.734541                                2.930216   \n",
      "2                     0.563973                                3.737960   \n",
      "3                     0.543779                                3.432823   \n",
      "4                     0.886033                                3.355527   \n",
      "\n",
      "   cat_HIS2_bb_rmsd_af2_out  cat_HIS2_cb_bb_rmsd_af2_out  \\\n",
      "0                  1.110435                     1.126091   \n",
      "1                  1.179181                     1.162210   \n",
      "2                  1.229525                     1.208171   \n",
      "3                  0.950403                     0.936527   \n",
      "4                  1.000834                     0.990085   \n",
      "\n",
      "   cat_HIS2_imidazole_rmsd_af2_out  cat_HIS2_ca_cb_rmsd_af2_out  \\\n",
      "0                         2.270202                     1.205363   \n",
      "1                         1.909446                     1.070859   \n",
      "2                         2.015856                     1.121826   \n",
      "3                         1.835581                     0.838954   \n",
      "4                         1.967654                     0.949140   \n",
      "\n",
      "   cat_HIS2_imid_sc_nitrogen_rmsd_af2_out  cat_HIS3_bb_rmsd_af2_out  \\\n",
      "0                                2.471441                  0.561985   \n",
      "1                                1.999151                  0.482263   \n",
      "2                                2.249754                  0.466438   \n",
      "3                                1.894889                  0.418669   \n",
      "4                                2.134985                  0.600578   \n",
      "\n",
      "   cat_HIS3_cb_bb_rmsd_af2_out  cat_HIS3_imidazole_rmsd_af2_out  \\\n",
      "0                     0.703234                         2.044484   \n",
      "1                     0.491957                         1.977817   \n",
      "2                     0.445323                         1.859227   \n",
      "3                     0.409542                         1.799065   \n",
      "4                     0.592681                         1.776751   \n",
      "\n",
      "   cat_HIS3_ca_cb_rmsd_af2_out  cat_HIS3_imid_sc_nitrogen_rmsd_af2_out  \\\n",
      "0                     0.902210                                2.008068   \n",
      "1                     0.481437                                2.019554   \n",
      "2                     0.335648                                1.915595   \n",
      "3                     0.339045                                1.866265   \n",
      "4                     0.495931                                1.756439   \n",
      "\n",
      "   cat_HIS_bb_rms_of_rmsds_af2_out  cat_HIS_cb_bb_rms_of_rmsds_af2_out  \\\n",
      "0                         0.836200                            0.862584   \n",
      "1                         0.821290                            0.824819   \n",
      "2                         0.825997                            0.812837   \n",
      "3                         0.668126                            0.663848   \n",
      "4                         0.824859                            0.826484   \n",
      "\n",
      "   cat_HIS_ca_cb_ONLY_rms_of_rmsds_af2_out  \\\n",
      "0                                 0.913169   \n",
      "1                                 0.799598   \n",
      "2                                 0.750382   \n",
      "3                                 0.609505   \n",
      "4                                 0.802469   \n",
      "\n",
      "   cat_HIS_imidazole_rms_of_rmsds_af2_out  \\\n",
      "0                                1.889326   \n",
      "1                                2.221575   \n",
      "2                                2.523454   \n",
      "3                                2.339421   \n",
      "4                                2.340361   \n",
      "\n",
      "   cat_HIS_imidazole_tip_Ns_rms_of_rmsds_af2_out  \\\n",
      "0                                       1.979740   \n",
      "1                                       2.356648   \n",
      "2                                       2.750954   \n",
      "3                                       2.507178   \n",
      "4                                       2.510166   \n",
      "\n",
      "                                                                               af2_names_for_matching  \\\n",
      "0         zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1   \n",
      "1  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3   \n",
      "2  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0   \n",
      "3  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2   \n",
      "4  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2   \n",
      "\n",
      "                                                                                                                                                                                            pdb_path_af2_apo_relax_out  \\\n",
      "0         ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb   \n",
      "1  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb   \n",
      "2  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb   \n",
      "3  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb   \n",
      "4  ./output/af2_out/predesigned_RFD2_i2/filtered_structures/APO_RELAXED/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2_model_4_ptm_seed_0_unrelaxed_relaxedAPO.pdb   \n",
      "\n",
      "                                                                                                                                                                     secondary_structure_af2_apo_relax_out  \\\n",
      "0  EEEEEE   HHHHHHHHHHHHHTT  EEEEEE GGGGGGTGGG  EEEEEEE  STT      EEHHHHHHHHHHHHHHHHHTT EEEEEEE       HHHHHHHHHHHHHHHHHHHHHHHHHHHHTT   SEEEEEEEE   HHHHHHHHHTTTTT  EEEEEE  TT  HHHHHHHHHHHHHHHHHHHHTTHHHHH   \n",
      "1  EEEEEE   HHHHHHHHHHHHHTT  EEEEEE GGGGGGTGGG  EEEEEEE   TT  TTTB  HHHHHHHHHHHHHHHHHTT EEEEEEE   EE  HHHHHHHHHHHHHHHHHHHHHHHHHHHHTT    EEEEEEEE   HHHHHHHHHTTTTT  EEEEEE   EE HHHHHHHHHHHHHHHHHHHHTTTTGGG   \n",
      "2  EEEEEE   HHHHHHHHHHHHHTT  EEEEEE GGGGGGTGGG  EEEEEEEET TT  TTTEE HHHHHHHHHHHHHHHHHTT EEEEEEE       HHHHHHHHHHHHHHHHHHHHHHHHHHHHTT    EEEEEEEE   GHHHHHHHHTTGGG  EEEEEEE   S  THHHHHHHHHHHHHHHHHHTTTTGGG   \n",
      "3  EEEEEE   HHHHHHHHHHHHHTT  EEEEEE GGGGGGTGGG  EEEEEEEE  TT  TTTEE HHHHHHHHHHHHHHHHHTT EEEEEEE       HHHHHHHHHHHHHHHHHHHHHHHHHHHHHT   SEEEEEEEE   GHHHHHHHHTTTTT  EEEEEEE   S HHHHHHHHHHHHHHHHHHHHTT GGGT   \n",
      "4  EEEEEE   HHHHHHHHHHHHHTT  EEEEEE GGGGGGTGGG  EEEEEEE B TT TT B   HHHHHHHHHHHHHHHHHTT EEEEEEE       HHHHHHHHHHHHHHHHHHHHHHHHHHHHHT   SEEEEEEEE   HHHHHHHHHTTTTT  EEEEEE  TT  HHHHHHHHHHHHHHHHHHHHTTHHHHH   \n",
      "\n",
      "                                                                                                                                                   secondary_structure_dssp_reduced_alphabet_af2_apo_relax_out  \\\n",
      "0  LLLEEEEEELLLHHHHHHHHHHHHHLLLLEEEEEELHHHHHHLHHHLLEEEEEEELLLLLLLLLLLEEHHHHHHHHHHHHHHHHHLLLEEEEEEELLLLLLLHHHHHHHHHHHHHHHHHHHHHHHHHHHHLLLLLLEEEEEEEELLLHHHHHHHHHLLLLLLLEEEEEELLLLLLHHHHHHHHHHHHHHHHHHHHLLHHHHHL   \n",
      "1  LLLEEEEEELLLHHHHHHHHHHHHHLLLLEEEEEELHHHHHHLHHHLLEEEEEEELLLLLLLLLLELLHHHHHHHHHHHHHHHHHLLLEEEEEEELLLEELLHHHHHHHHHHHHHHHHHHHHHHHHHHHHLLLLLLEEEEEEEELLLHHHHHHHHHLLLLLLLEEEEEELLLEELHHHHHHHHHHHHHHHHHHHHLLLLHHHL   \n",
      "2  LLLEEEEEELLLHHHHHHHHHHHHHLLLLEEEEEELHHHHHHLHHHLLEEEEEEEELLLLLLLLLEELHHHHHHHHHHHHHHHHHLLLEEEEEEELLLLLLLHHHHHHHHHHHHHHHHHHHHHHHHHHHHLLLLLLEEEEEEEELLLHHHHHHHHHLLHHHLLEEEEEEELLLLLLLHHHHHHHHHHHHHHHHHHLLLLHHHL   \n",
      "3  LLLEEEEEELLLHHHHHHHHHHHHHLLLLEEEEEELHHHHHHLHHHLLEEEEEEEELLLLLLLLLEELHHHHHHHHHHHHHHHHHLLLEEEEEEELLLLLLLHHHHHHHHHHHHHHHHHHHHHHHHHHHHHLLLLLEEEEEEEELLLHHHHHHHHHLLLLLLLEEEEEEELLLLLHHHHHHHHHHHHHHHHHHHHLLLHHHLL   \n",
      "4  LLLEEEEEELLLHHHHHHHHHHHHHLLLLEEEEEELHHHHHHLHHHLLEEEEEEELELLLLLLLELLLHHHHHHHHHHHHHHHHHLLLEEEEEEELLLLLLLHHHHHHHHHHHHHHHHHHHHHHHHHHHHHLLLLLEEEEEEEELLLHHHHHHHHHLLLLLLLEEEEEELLLLLLHHHHHHHHHHHHHHHHHHHHLLHHHHHL   \n",
      "\n",
      "   sap_score_af2_apo_relax_out  \\\n",
      "0                      47.0713   \n",
      "1                      47.2152   \n",
      "2                      47.6003   \n",
      "3                      44.7762   \n",
      "4                      49.9606   \n",
      "\n",
      "   aliphatic_residues_in_design_af2_apo_relax_out  \\\n",
      "0                                              74   \n",
      "1                                              71   \n",
      "2                                              71   \n",
      "3                                              71   \n",
      "4                                              72   \n",
      "\n",
      "   bad_torsion_preproline_af2_apo_relax_out  \\\n",
      "0                                         2   \n",
      "1                                         2   \n",
      "2                                         2   \n",
      "3                                         2   \n",
      "4                                         2   \n",
      "\n",
      "   holes_in_design_lower_is_better_af2_apo_relax_out  \\\n",
      "0                                          -0.713032   \n",
      "1                                          -1.409680   \n",
      "2                                          -1.130610   \n",
      "3                                          -1.495660   \n",
      "4                                          -0.698413   \n",
      "\n",
      "   hydrophobic_exposure_sasa_in_design_af2_apo_relax_out  \\\n",
      "0                                                725.773   \n",
      "1                                                599.103   \n",
      "2                                                559.828   \n",
      "3                                                624.529   \n",
      "4                                                760.928   \n",
      "\n",
      "   hydrophobic_residues_in_design_af2_apo_relax_out  \\\n",
      "0                                                64   \n",
      "1                                                67   \n",
      "2                                                66   \n",
      "3                                                66   \n",
      "4                                                65   \n",
      "\n",
      "   longest_cont_apolar_seg_af2_apo_relax_out  \\\n",
      "0                                         10   \n",
      "1                                         10   \n",
      "2                                         10   \n",
      "3                                         10   \n",
      "4                                          5   \n",
      "\n",
      "   longest_cont_polar_seg_af2_apo_relax_out  \\\n",
      "0                                         5   \n",
      "1                                         7   \n",
      "2                                         7   \n",
      "3                                         7   \n",
      "4                                         5   \n",
      "\n",
      "   net_charge_in_design_not_w_his_af2_apo_relax_out  \\\n",
      "0                                               -13   \n",
      "1                                               -10   \n",
      "2                                               -12   \n",
      "3                                               -10   \n",
      "4                                               -12   \n",
      "\n",
      "   number_dssp_helices_in_design_af2_apo_relax_out  \\\n",
      "0                                                8   \n",
      "1                                                8   \n",
      "2                                                9   \n",
      "3                                                8   \n",
      "4                                                8   \n",
      "\n",
      "   number_dssp_loops_in_design_af2_apo_relax_out  \\\n",
      "0                                             15   \n",
      "1                                             18   \n",
      "2                                             17   \n",
      "3                                             16   \n",
      "4                                             17   \n",
      "\n",
      "   number_dssp_sheets_in_design_af2_apo_relax_out  \\\n",
      "0                                               6   \n",
      "1                                               6   \n",
      "2                                               6   \n",
      "3                                               6   \n",
      "4                                               6   \n",
      "\n",
      "   total_pose_sasa_af2_apo_relax_out  \\\n",
      "0                            9718.16   \n",
      "1                            9757.14   \n",
      "2                            9704.83   \n",
      "3                            9376.68   \n",
      "4                            9755.36   \n",
      "\n",
      "   total_residues_in_design_plus_ligand_af2_apo_relax_out  \\\n",
      "0                                                     203   \n",
      "1                                                     203   \n",
      "2                                                     203   \n",
      "3                                                     203   \n",
      "4                                                     203   \n",
      "\n",
      "   total_rosetta_energy_metric_af2_apo_relax_out  \\\n",
      "0                                       -559.592   \n",
      "1                                       -598.550   \n",
      "2                                       -584.716   \n",
      "3                                       -601.053   \n",
      "4                                       -607.741   \n",
      "\n",
      "                                                                                                    second_shell_seq_clustering  \\\n",
      "0  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_model_4_ptm_seed_0_unrelaxed.pdb   \n",
      "1                                   zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8   \n",
      "2                                   zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8   \n",
      "3                                   zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8   \n",
      "4                                   zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9   \n",
      "\n",
      "   normalized_rmsd  normalized_pTMscore  combined_score  \\\n",
      "0              NaN                  NaN             NaN   \n",
      "1          0.00000             0.996482        0.001759   \n",
      "2          1.00000             1.000000        0.500000   \n",
      "3          0.77675             0.000000        0.888375   \n",
      "4          1.00000             1.000000        0.500000   \n",
      "\n",
      "                                      fastmpnn_clustering  \n",
      "0  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10  \n",
      "1  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10  \n",
      "2  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10  \n",
      "3  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10  \n",
      "4  zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1763/346897955.py:219: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  shell_filtered_megadata_df = filtered_af2_sc_AND_global_scores_df.groupby('second_shell_seq_clustering').apply(lambda x: x if len(x) < 3 else keep_top_3(x)).reset_index(drop=True)\n"
     ]
    }
   ],
   "source": [
    "##########################################################################################################################\n",
    "### DO FILTERING BY SIDECHAINS & GLOBAL METRICS (DAMAGE CONTROL IN THIS CASE) AND SEQUENCE CLUSTER & FILTER DOWN TOP 3 ###\n",
    "##########################################################################################################################\n",
    "\n",
    "### INPUTS ###\n",
    "af2_file = f\"{af2_out_dir}predesigned_RFD2_i2/filtered_structures/updated_scores.sc\"\n",
    "# num1: predesigned_RFD2_i2\n",
    "# num2: predesigned_inpainting_i2\n",
    "apo_relax_file = f'{af2_out_dir}predesigned_RFD2_i2/filtered_structures/APO_RELAXED/parsed_energy_data.csv'\n",
    "# num1: predesigned_RFD2_i2\n",
    "# num2: predesigned_inpainting_i2\n",
    "\n",
    "# Specify column groups for calculation\n",
    "column_groups = {\n",
    "    'cat_HIS_bb': ['cat_HIS1_bb_rmsd', 'cat_HIS2_bb_rmsd', 'cat_HIS3_bb_rmsd'],\n",
    "    'cat_HIS_cb_bb': ['cat_HIS1_cb_bb_rmsd', 'cat_HIS2_cb_bb_rmsd', 'cat_HIS3_cb_bb_rmsd'],\n",
    "    'cat_HIS_ca_cb_ONLY': ['cat_HIS1_ca_cb_rmsd', 'cat_HIS2_ca_cb_rmsd', 'cat_HIS3_ca_cb_rmsd'],\n",
    "    'cat_HIS_imidazole': ['cat_HIS1_imidazole_rmsd', 'cat_HIS2_imidazole_rmsd', 'cat_HIS3_imidazole_rmsd'],\n",
    "    'cat_HIS_imidazole_tip_Ns': ['cat_HIS1_imid_sc_nitrogen_rmsd', 'cat_HIS2_imid_sc_nitrogen_rmsd', 'cat_HIS3_imid_sc_nitrogen_rmsd']\n",
    "}\n",
    "\n",
    "### FUNCTION TO CALCULATE MEAN RMSD AND RMS OF RMSDs FOR SPECIFIED COLUMN GROUPS ###\n",
    "def calculate_group_rmsd(df, column_groups):\n",
    "    for group_name, columns in column_groups.items():\n",
    "        rms_col_name = f\"{group_name}_rms_of_rmsds\"\n",
    "        df[rms_col_name] = np.sqrt((df[columns]**2).mean(axis=1))\n",
    "\n",
    "### FUNCTION TO EXTRACT BASE NAME FOR MATCHING ###\n",
    "def extract_base_name_for_matching(pdb_path):\n",
    "    base_name = os.path.basename(pdb_path)\n",
    "    parts = base_name.split('_model_4_ptm_seed_0_unrelaxed')\n",
    "    if len(parts) > 2:\n",
    "        return '_model_4_ptm_seed_0_unrelaxed'.join(parts[:2])\n",
    "    return base_name\n",
    "\n",
    "### FUNCTION TO EXTRACT BASE NAME ###\n",
    "def extract_base_name(pdb_path):\n",
    "    base_name = os.path.basename(pdb_path)\n",
    "    if '_T0' in base_name:\n",
    "        return base_name.split('_T0')[0]\n",
    "    elif '_model_4_ptm_seed_0_unrelaxed_relaxedAPO_' in base_name:\n",
    "        return base_name.split('_model_4_ptm_seed_0_unrelaxed_relaxedAPO_')[0]\n",
    "    else:\n",
    "        return base_name\n",
    "\n",
    "### LOAD DATA FRAMES ###\n",
    "# Load AF2 DataFrame\n",
    "af2_df = pd.read_csv(af2_file, sep=',', na_values='')\n",
    "calculate_group_rmsd(af2_df, column_groups)\n",
    "af2_df = af2_df.add_suffix('_af2_out')\n",
    "af2_df['af2_names_for_matching'] = af2_df['pdb_path_af2_out'].apply(extract_base_name_for_matching)\n",
    "\n",
    "# Load APO Relax DataFrame\n",
    "apo_relax_df = pd.read_csv(apo_relax_file)\n",
    "apo_relax_df = apo_relax_df.add_suffix('_af2_apo_relax_out')\n",
    "apo_relax_df['af2_names_for_matching'] = apo_relax_df['pdb_path_af2_apo_relax_out'].apply(extract_base_name_for_matching)\n",
    "\n",
    "# Print the shape of each DataFrame before merging\n",
    "print(\"AF2 DataFrame before merge:\")\n",
    "print(af2_df.shape)\n",
    "print(\"\\nAPO Relax DataFrame before merge:\")\n",
    "print(apo_relax_df.shape)\n",
    "\n",
    "### MERGE DATA FRAMES ###\n",
    "# First merge af2_df and apo_relax_df\n",
    "merged_df = pd.merge(af2_df, apo_relax_df, on='af2_names_for_matching', suffixes=('_af2', '_apo_relax'))\n",
    "\n",
    "# Print the shape of the merged DataFrame\n",
    "print(\"\\nMerged DataFrame:\")\n",
    "print(merged_df.shape)\n",
    "\n",
    "### CREATE SECOND SHELL SEQ CLUSTERING COLUMN ###\n",
    "merged_df['second_shell_seq_clustering'] = merged_df['pdb_path_af2_out'].apply(extract_base_name)\n",
    "\n",
    "### GROUP BY SECOND SHELL SEQ CLUSTERING ###\n",
    "grouped = merged_df.groupby('second_shell_seq_clustering')\n",
    "\n",
    "# Get the number of unique groups\n",
    "num_unique_groups = grouped.ngroups\n",
    "\n",
    "# Count the number of rows in each group\n",
    "group_counts = grouped.size()\n",
    "\n",
    "# Count how many groups have more than 3, 5, and 7 rows\n",
    "groups_more_than_3 = group_counts[group_counts > 3].count()\n",
    "groups_more_than_5 = group_counts[group_counts > 5].count()\n",
    "groups_more_than_7 = group_counts[group_counts > 7].count()\n",
    "\n",
    "### PRINT RESULTS ###\n",
    "print(f'\\nTotal number of rows in the merged DataFrame: {len(merged_df)}')\n",
    "print(f'Number of unique groups: {num_unique_groups}')\n",
    "print(f'Number of groups with more than 3 rows: {groups_more_than_3} (Total rows: {group_counts[group_counts > 3].sum()})')\n",
    "print(f'Number of groups with more than 5 rows: {groups_more_than_5} (Total rows: {group_counts[group_counts > 5].sum()})')\n",
    "print(f'Number of groups with more than 7 rows: {groups_more_than_7} (Total rows: {group_counts[group_counts > 7].sum()})')\n",
    "\n",
    "# Optionally, print the number of rows in each group\n",
    "print(\"\\nNumber of rows in each group:\")\n",
    "#print(group_counts)\n",
    "\n",
    "mega_dataframe = merged_df\n",
    "\n",
    "# Now apply the filtering to mega_dataframe\n",
    "##################################################################################\n",
    "### COMPUTE SIDE CHAIN STATISTICS FROM SCORES DF & DO GENTLE/LOGICAL FILTERING ###\n",
    "##################################################################################\n",
    "\n",
    "### OLD FILTER FROM GLOBAL STUFF (EVERYTHING SHOULD PASS THIS ONE) ###\n",
    "plddt_cutoff = 72.5  # in pLDDT units -> usually 70-80\n",
    "rmsd_cutoff = 1.75  # √Ö, I think this is CŒ± (1.5 is typical, 2.0 might be a stretch but okay for big structures)\n",
    "pae_cutoff = 7.25  # below 5 is harsh but good -> not applicable here\n",
    "pTM_cutoff = 0.725  # Supposedly above 0.75 is really good & above 0.5 is on the side of probs correct\n",
    "convergence_cutoff = 0.75  # Get it in a reasonable range\n",
    "\n",
    "# Adjust column names with suffixes\n",
    "global_metric_filter_af2_mask = (\n",
    "    (mega_dataframe['mean_plddt_af2_out'] >= plddt_cutoff) & \n",
    "    (mega_dataframe['rmsd_to_input_af2_out'] <= rmsd_cutoff) &\n",
    "    (mega_dataframe['mean_pae_af2_out'] <= pae_cutoff) & \n",
    "    (mega_dataframe['pTMscore_af2_out'] >= pTM_cutoff) &\n",
    "    (mega_dataframe['af2_convergence_tol_af2_out'] <= convergence_cutoff)\n",
    ")\n",
    "\n",
    "filtered_af2_global_scores_df = mega_dataframe[global_metric_filter_af2_mask]\n",
    "global_af2_filter_rate = float(len(filtered_af2_global_scores_df)) / float(len(mega_dataframe)) * 100.0\n",
    "print(f'{len(filtered_af2_global_scores_df)} of {len(mega_dataframe)} ({round(global_af2_filter_rate, 3)}%) with plddt >= {plddt_cutoff}, CŒ± rmsd <= {rmsd_cutoff}√Ö, convergence <= {convergence_cutoff}√Ö, pae <= {pae_cutoff}, and pTM >= {pTM_cutoff}')\n",
    "print('')\n",
    "\n",
    "### NEW FILTER FROM SIDE CHAIN 2 ###\n",
    "cat_HIS_bb_rms_of_rmsds_cutoff = 0.85  # √Ö\n",
    "cat_HIS_cb_bb_rms_of_rmsds_cutoff = 0.85  # √Ö\n",
    "cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff = 0.85  # √Ö\n",
    "cat_HIS_imidazole_rms_of_rmsds_cutoff = 1.25  # √Ö\n",
    "cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff = 1.25  # √Ö\n",
    "\n",
    "# Adjust column names with suffixes\n",
    "filter_masks = {\n",
    "    'cat_HIS_bb_rms_of_rmsds_mask': filtered_af2_global_scores_df['cat_HIS_bb_rms_of_rmsds_af2_out'] <= cat_HIS_bb_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_cb_bb_rms_of_rmsds_mask': filtered_af2_global_scores_df['cat_HIS_cb_bb_rms_of_rmsds_af2_out'] <= cat_HIS_cb_bb_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask': filtered_af2_global_scores_df['cat_HIS_ca_cb_ONLY_rms_of_rmsds_af2_out'] <= cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_imidazole_rms_of_rmsds_mask': filtered_af2_global_scores_df['cat_HIS_imidazole_rms_of_rmsds_af2_out'] <= cat_HIS_imidazole_rms_of_rmsds_cutoff,\n",
    "    'cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask': filtered_af2_global_scores_df['cat_HIS_imidazole_tip_Ns_rms_of_rmsds_af2_out'] <= cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff\n",
    "}\n",
    "\n",
    "# Calculate pass counts and rates for each filter\n",
    "filter_results = {filter_name: (mask.sum(), len(filtered_af2_global_scores_df), mask.mean() * 100) for filter_name, mask in filter_masks.items()}\n",
    "\n",
    "# Overall filtering result using OR logic\n",
    "cumulative_filter_mask_or = filter_masks['cat_HIS_bb_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_cb_bb_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_imidazole_rms_of_rmsds_mask'] | \\\n",
    "                            filter_masks['cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask']\n",
    "\n",
    "cumulative_pass_count_or = cumulative_filter_mask_or.sum()\n",
    "cumulative_total_or = len(filtered_af2_global_scores_df)\n",
    "cumulative_pass_rate_or = cumulative_pass_count_or / cumulative_total_or * 100\n",
    "\n",
    "### DO THE PRINT STATEMENT (FANCY) ###\n",
    "print('########################## SIDE CHAIN RMSD METRICS FILTER ##########################')\n",
    "for filter_name, (pass_count, total, pass_rate) in filter_results.items():\n",
    "    specific_criteria = filter_name.replace('_mask', '').replace('_', ' ').replace('cat', 'catalytic').upper()\n",
    "    cutoff_value = {\n",
    "        'cat_HIS_bb_rms_of_rmsds_mask': cat_HIS_bb_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_cb_bb_rms_of_rmsds_mask': cat_HIS_cb_bb_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_ca_cb_ONLY_rms_of_rmsds_mask': cat_HIS_ca_cb_ONLY_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_imidazole_rms_of_rmsds_mask': cat_HIS_imidazole_rms_of_rmsds_cutoff,\n",
    "        'cat_HIS_imidazole_tip_Ns_rms_of_rmsds_mask': cat_HIS_imidazole_tip_Ns_rms_of_rmsds_cutoff\n",
    "    }[filter_name]\n",
    "    print(f'{pass_count} of {total} ({pass_rate:.3f}%) have {specific_criteria} <= {cutoff_value}')\n",
    "\n",
    "# Maintaining the cumulative print statement from the provided script\n",
    "print('')\n",
    "print(f'{cumulative_pass_count_or} of {cumulative_total_or} ({cumulative_pass_rate_or:.3f}%) pass at least one filter!')\n",
    "print('####################################################################################')\n",
    "\n",
    "### MAKE THE FINAL DATA FRAME ###\n",
    "final_filtered_df = filtered_af2_global_scores_df[cumulative_filter_mask_or]\n",
    "filtered_af2_sc_AND_global_scores_df = final_filtered_df\n",
    "\n",
    "### PRINT OUT COLUMN NAMES USED IN FILTERING ###\n",
    "filtered_columns = ', '.join([key.replace('_mask', '') for key in filter_masks.keys()])\n",
    "print(f'Columns used in filtering: {filtered_columns}')\n",
    "\n",
    "### GROUP BY SECOND SHELL SEQ CLUSTERING ###\n",
    "grouped = filtered_af2_sc_AND_global_scores_df.groupby('second_shell_seq_clustering')\n",
    "\n",
    "# Get the number of unique groups\n",
    "num_unique_groups = grouped.ngroups\n",
    "\n",
    "# Count the number of rows in each group\n",
    "group_counts = grouped.size()\n",
    "\n",
    "# Count how many groups have more than 3, 5, and 7 rows\n",
    "groups_3 = group_counts[group_counts >= 3].count()\n",
    "groups_more_than_3 = group_counts[group_counts > 3].count()\n",
    "print('')\n",
    "print('METRICS FOR CHECKING WITH BELOW CALCULATIONS/LOGIC:')\n",
    "print(f'Number of unique groups: {num_unique_groups} (Total rows: {len(filtered_af2_sc_AND_global_scores_df)})')\n",
    "print(f'Number of groups with 3+ rows: {groups_3} (Total rows: {group_counts[group_counts >= 3].sum()})')\n",
    "print(f'Number of groups with more than 4+ rows: {groups_more_than_3} (Total rows: {group_counts[group_counts > 3].sum()})')\n",
    "print('')\n",
    "\n",
    "### TOP 3 STRUCTURES PER GROUP BY COMBINED SCORE ###\n",
    "# Normalize rmsd_to_input_af2_out and pTMscore_af2_out within each group\n",
    "def normalize_series(series):\n",
    "    return (series - series.min()) / (series.max() - series.min())\n",
    "\n",
    "filtered_af2_sc_AND_global_scores_df['normalized_rmsd'] = filtered_af2_sc_AND_global_scores_df.groupby('second_shell_seq_clustering')['rmsd_to_input_af2_out'].transform(normalize_series)\n",
    "filtered_af2_sc_AND_global_scores_df['normalized_pTMscore'] = filtered_af2_sc_AND_global_scores_df.groupby('second_shell_seq_clustering')['pTMscore_af2_out'].transform(normalize_series)\n",
    "\n",
    "# Create a combined score (50/50 weighting of normalized_rmsd and normalized_pTMscore)\n",
    "filtered_af2_sc_AND_global_scores_df['combined_score'] = 0.5 * filtered_af2_sc_AND_global_scores_df['normalized_rmsd'] + 0.5 * (1 - filtered_af2_sc_AND_global_scores_df['normalized_pTMscore'])\n",
    "\n",
    "# Sort by combined_score within each group and keep the top 3 structures, or all structures if there are fewer than 3\n",
    "def keep_top_3(group):\n",
    "    return group.nsmallest(3, 'combined_score')\n",
    "\n",
    "# Ensure all unique groups are represented\n",
    "shell_filtered_megadata_df = filtered_af2_sc_AND_global_scores_df.groupby('second_shell_seq_clustering').apply(lambda x: x if len(x) < 3 else keep_top_3(x)).reset_index(drop=True)\n",
    "\n",
    "### PRINT FILTERING RESULTS ###\n",
    "num_structures_before = len(filtered_af2_sc_AND_global_scores_df)\n",
    "num_structures_after = len(shell_filtered_megadata_df)\n",
    "print(f'\\nNumber of structures before top 3 filtering: {num_structures_before}')\n",
    "print(f'Number of structures after top 3 filtering: {num_structures_after}')\n",
    "print(f'Number of structures filtered out: {num_structures_before - num_structures_after}')\n",
    "print('')\n",
    "\n",
    "### GROUP BY SECOND SHELL SEQ CLUSTERING ###\n",
    "grouped = shell_filtered_megadata_df.groupby('second_shell_seq_clustering')\n",
    "\n",
    "# Get the number of unique groups\n",
    "num_unique_groups = grouped.ngroups\n",
    "\n",
    "# Count the number of rows in each group\n",
    "group_counts = grouped.size()\n",
    "\n",
    "# Count how many groups have 3 rows\n",
    "groups_3 = group_counts[group_counts == 3].count()\n",
    "groups_more_than_3 = group_counts[group_counts > 3].count()\n",
    "print(f'Number of unique groups: {num_unique_groups}')\n",
    "print(f'Number of groups with 3 rows: {groups_3} (Total rows: {group_counts[group_counts == 3].sum()})')\n",
    "print(f'Number of groups with more than 3 rows: {groups_more_than_3} (Total rows: {group_counts[group_counts > 3].sum()})')\n",
    "\n",
    "### CREATE FASTMPNN_CLUSTERING COLUMN ###\n",
    "def extract_fastmpnn_clustering(seq_clustering):\n",
    "    return seq_clustering.split('_model_')[0]\n",
    "\n",
    "shell_filtered_megadata_df['fastmpnn_clustering'] = shell_filtered_megadata_df['second_shell_seq_clustering'].apply(extract_fastmpnn_clustering)\n",
    "\n",
    "# Find the number of unique fastmpnn groups\n",
    "num_unique_fastmpnn_groups = shell_filtered_megadata_df['fastmpnn_clustering'].nunique()\n",
    "print('')\n",
    "print(f'Number of unique fastmpnn groups: {num_unique_fastmpnn_groups}')\n",
    "\n",
    "# Optionally, show the final filtered DataFrame\n",
    "print(shell_filtered_megadata_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e8d0e4e4-5592-4ea9-af57-0f965b96aaf0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in df_sorted: Index(['af2_json_path_af2_out', 'pdb_path_af2_out', 'mean_plddt_af2_out',\n",
      "       'rmsd_to_input_af2_out', 'mean_pae_intra_chain_af2_out',\n",
      "       'mean_pae_af2_out', 'pTMscore_af2_out', 'af2_convergence_tol_af2_out',\n",
      "       'af2_elapsed_folding_time_af2_out', 'cat_HIS1_bb_rmsd_af2_out',\n",
      "       'cat_HIS1_cb_bb_rmsd_af2_out', 'cat_HIS1_imidazole_rmsd_af2_out',\n",
      "       'cat_HIS1_ca_cb_rmsd_af2_out', 'cat_HIS1_imid_sc_nitrogen_rmsd_af2_out',\n",
      "       'cat_HIS2_bb_rmsd_af2_out', 'cat_HIS2_cb_bb_rmsd_af2_out',\n",
      "       'cat_HIS2_imidazole_rmsd_af2_out', 'cat_HIS2_ca_cb_rmsd_af2_out',\n",
      "       'cat_HIS2_imid_sc_nitrogen_rmsd_af2_out', 'cat_HIS3_bb_rmsd_af2_out',\n",
      "       'cat_HIS3_cb_bb_rmsd_af2_out', 'cat_HIS3_imidazole_rmsd_af2_out',\n",
      "       'cat_HIS3_ca_cb_rmsd_af2_out', 'cat_HIS3_imid_sc_nitrogen_rmsd_af2_out',\n",
      "       'cat_HIS_bb_rms_of_rmsds_af2_out', 'cat_HIS_cb_bb_rms_of_rmsds_af2_out',\n",
      "       'cat_HIS_ca_cb_ONLY_rms_of_rmsds_af2_out',\n",
      "       'cat_HIS_imidazole_rms_of_rmsds_af2_out',\n",
      "       'cat_HIS_imidazole_tip_Ns_rms_of_rmsds_af2_out',\n",
      "       'af2_names_for_matching', 'pdb_path_af2_apo_relax_out',\n",
      "       'secondary_structure_af2_apo_relax_out',\n",
      "       'secondary_structure_dssp_reduced_alphabet_af2_apo_relax_out',\n",
      "       'sap_score_af2_apo_relax_out',\n",
      "       'aliphatic_residues_in_design_af2_apo_relax_out',\n",
      "       'bad_torsion_preproline_af2_apo_relax_out',\n",
      "       'holes_in_design_lower_is_better_af2_apo_relax_out',\n",
      "       'hydrophobic_exposure_sasa_in_design_af2_apo_relax_out',\n",
      "       'hydrophobic_residues_in_design_af2_apo_relax_out',\n",
      "       'longest_cont_apolar_seg_af2_apo_relax_out',\n",
      "       'longest_cont_polar_seg_af2_apo_relax_out',\n",
      "       'net_charge_in_design_not_w_his_af2_apo_relax_out',\n",
      "       'number_dssp_helices_in_design_af2_apo_relax_out',\n",
      "       'number_dssp_loops_in_design_af2_apo_relax_out',\n",
      "       'number_dssp_sheets_in_design_af2_apo_relax_out',\n",
      "       'total_pose_sasa_af2_apo_relax_out',\n",
      "       'total_residues_in_design_plus_ligand_af2_apo_relax_out',\n",
      "       'total_rosetta_energy_metric_af2_apo_relax_out',\n",
      "       'second_shell_seq_clustering', 'normalized_rmsd', 'normalized_pTMscore',\n",
      "       'combined_score', 'fastmpnn_clustering'],\n",
      "      dtype='object')\n",
      "Starting process...\n",
      "Total rows before filtering: 12\n",
      "Number of unique fastmpnn_clustering before filtering: 1\n",
      "Filtered groups with 1 or more entries: 12\n",
      "Number of unique fastmpnn_clustering after filtering groups with 1 or more entries: 1\n",
      "Randomly selected sequence from ./output/af2_out/predesigned_RFD2_i2/filtered_structures/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_model_4_ptm_seed_0_unrelaxed.pdb: KKSLKVVHVGPNEEANKLVEKGLREGGVEAEVISVKPEEVEEYVPEADILGFHVYLNDPTQPKELALSKEAVEKHSDYLIELAAENNVDLHFGVSTGVGKFTPEEWNKLSREAVDLLGELLKKAKEVREKKGSNALIIGVFLGISPDPEIVKLAYEKSEPYADYVLVKTAPNDATRDDLLALIYEIAYEAGKDINSGKFEKLR\n",
      "\n",
      "Processing fastmpnn_clustering zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10 with 12 entries\n",
      "Final processing complete.\n",
      "Total sequences flagged for consecutive charged residues: 0\n",
      "Total sequences flagged for repetitive amino acids: 0\n",
      "Total sequences flagged for hydrophobic stretches: 0\n",
      "Total sequences flagged for specific patterns: 0\n",
      "Number of rows in the final DataFrame: 10\n",
      "Number of unique fastmpnn_clustering after filtering: 1\n",
      "Number of sequences with exact duplicates in the final DataFrame: 0\n",
      "\n",
      "fastmpnn_clustering zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10 has 10 files\n",
      "Number of rows in the new DataFrame: 10\n",
      "Number of unique fastmpnn_clustering after filtering: 1\n"
     ]
    }
   ],
   "source": [
    "###################################################################################################\n",
    "### GET RID OF THE SHITTY SEQUENCES TO START FILTERING DOWN (IF ALL ARE SHIT TAKE THE BEST ONE) ###\n",
    "###################################################################################################\n",
    "\n",
    "### INPUTS ###\n",
    "df_sorted = shell_filtered_megadata_df\n",
    "pdb_column = 'pdb_path_af2_out'\n",
    "grouping_column = 'fastmpnn_clustering'\n",
    "ptm_column = 'pTMscore_af2_out'\n",
    "rmsd_column = 'rmsd_to_input_af2_out'\n",
    "similarity_threshold = 0.975 # SEQUENCE SIMILARITY TO DEFINE A CLUSTER\n",
    "\n",
    "# Function to extract amino acid sequences from PDB files, considering only the first sequence if multiple are present\n",
    "def extract_sequences(pdb_paths):\n",
    "    sequences = {}\n",
    "    pdb_parser = PDBParser(QUIET=True)\n",
    "    pp_builder = PPBuilder()\n",
    "    \n",
    "    for pdb_path in pdb_paths:\n",
    "        try:\n",
    "            structure = pdb_parser.get_structure(pdb_path, pdb_path)\n",
    "            for pp in pp_builder.build_peptides(structure):\n",
    "                seq = pp.get_sequence()\n",
    "                sequences[pdb_path] = str(seq)\n",
    "                break  # Only take the first sequence\n",
    "        except Exception as e:\n",
    "            print(f\"Failed to read {pdb_path}: {e}\")\n",
    "    \n",
    "    if sequences:\n",
    "        random_path = random.choice(list(sequences.keys()))\n",
    "        print(f\"Randomly selected sequence from {random_path}: {sequences[random_path]}\")\n",
    "    else:\n",
    "        print(\"No sequences extracted.\")\n",
    "    \n",
    "    return sequences\n",
    "\n",
    "# Function to calculate similarity between two sequences\n",
    "def calculate_similarity(seq1, seq2):\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "    best_alignment = alignments[0]\n",
    "    matches = best_alignment[2]\n",
    "    similarity = matches / max(len(seq1), len(seq2))\n",
    "    return similarity\n",
    "\n",
    "# Function to cluster sequences by similarity\n",
    "def cluster_sequences(group, threshold):\n",
    "    clusters = []\n",
    "    for index, row in group.iterrows():\n",
    "        seq = row['sequence']\n",
    "        placed = False\n",
    "        for cluster in clusters:\n",
    "            if calculate_similarity(seq, cluster['representative']['sequence']) >= threshold:\n",
    "                cluster['members'].append(row)\n",
    "                placed = True\n",
    "                break\n",
    "        if not placed:\n",
    "            clusters.append({'representative': row, 'members': [row]})\n",
    "    return clusters\n",
    "\n",
    "# Function to flag sequences with repetitive amino acids, excessive same-charge residues, long hydrophobic stretches, or specific patterns\n",
    "def flag_sequences(seq):\n",
    "    repetitive_pattern = re.compile(r\"(.)\\1{4,}\")  # flag sequences with four or more consecutive identical amino acids\n",
    "    negative_charged_pattern = re.compile(r\"(D{5,}|E{5,}|(DE){5,})\")  # flag sequences with four or more consecutive negatively charged residues\n",
    "    positive_charged_pattern = re.compile(r\"(K{5,}|R{5,}|(KR){5,})\")  # flag sequences with four or more consecutive positively charged residues\n",
    "    hydrophobic_pattern = re.compile(r\"[AVILMFYWP]{7,}\")  # flag sequences with six or more consecutive hydrophobic residues\n",
    "    specific_pattern = re.compile(r\"(AANDENYALAA)\")  # flag sequences with RR or KR\n",
    "    \n",
    "    repetitive_match = repetitive_pattern.search(seq)\n",
    "    negative_charged_match = negative_charged_pattern.search(seq)\n",
    "    positive_charged_match = positive_charged_pattern.search(seq)\n",
    "    hydrophobic_match = hydrophobic_pattern.search(seq)\n",
    "    specific_match = specific_pattern.search(seq)\n",
    "    \n",
    "    is_repetitive = bool(repetitive_match)\n",
    "    is_negatively_charged = bool(negative_charged_match)\n",
    "    is_positively_charged = bool(positive_charged_match)\n",
    "    is_hydrophobic = bool(hydrophobic_match)\n",
    "    is_specific = bool(specific_match)\n",
    "    \n",
    "    flagged_parts = []\n",
    "    violations = 0\n",
    "    if is_repetitive:\n",
    "        flagged_parts.append(f\"Repetitive: {repetitive_match.group()}\")\n",
    "        violations += 1\n",
    "    if is_negatively_charged:\n",
    "        flagged_parts.append(f\"Negatively Charged: {negative_charged_match.group()}\")\n",
    "        violations += 1\n",
    "    if is_positively_charged:\n",
    "        flagged_parts.append(f\"Positively Charged: {positive_charged_match.group()}\")\n",
    "        violations += 1\n",
    "    if is_hydrophobic:\n",
    "        flagged_parts.append(f\"Hydrophobic: {hydrophobic_match.group()}\")\n",
    "        violations += 1\n",
    "    if is_specific:\n",
    "        flagged_parts.append(f\"Specific Pattern: {specific_match.group()}\")\n",
    "        violations += 1\n",
    "\n",
    "    if flagged_parts:\n",
    "        flagged_str = \", \".join(flagged_parts)\n",
    "        print(f\"Flagging sequence: {seq} - {flagged_str}\")\n",
    "    \n",
    "    return is_repetitive or is_negatively_charged or is_positively_charged or is_hydrophobic or is_specific, flagged_parts, violations\n",
    "\n",
    "# Main function to process the DataFrame\n",
    "def process_dataframe(df_sorted, similarity_threshold=0.85):\n",
    "    print(\"Starting process...\")\n",
    "    print(f\"Total rows before filtering: {len(df_sorted)}\")\n",
    "    print(f\"Number of unique {grouping_column} before filtering: {df_sorted[grouping_column].nunique()}\")\n",
    "    \n",
    "    group_sizes = df_sorted.groupby(grouping_column).size()\n",
    "    df_filtered_groups = df_sorted[df_sorted[grouping_column].isin(group_sizes[group_sizes >= 1].index)].copy()\n",
    "    print(f\"Filtered groups with 1 or more entries: {len(df_filtered_groups)}\")\n",
    "    print(f\"Number of unique {grouping_column} after filtering groups with 1 or more entries: {df_filtered_groups[grouping_column].nunique()}\")\n",
    "\n",
    "    pdb_paths = df_filtered_groups[pdb_column].unique()\n",
    "    sequences = extract_sequences(pdb_paths)\n",
    "    df_filtered_groups['sequence'] = df_filtered_groups[pdb_column].map(sequences)\n",
    "    df_filtered_groups['sequence_to_repair'] = np.nan\n",
    "\n",
    "    final_df = pd.DataFrame()\n",
    "    subgroup_counts = {}\n",
    "    flagged_repetitive_sequences = []\n",
    "    flagged_charged_sequences = []\n",
    "    flagged_hydrophobic_sequences = []\n",
    "    flagged_specific_sequences = []\n",
    "\n",
    "    if df_filtered_groups.empty:\n",
    "        print(\"No groups with 1 or more entries. Exiting...\")\n",
    "        return pd.DataFrame()\n",
    "\n",
    "    for group_value, group in df_filtered_groups.groupby(grouping_column):\n",
    "        print('')\n",
    "        print(f\"Processing {grouping_column} {group_value} with {len(group)} entries\")\n",
    "        group['flag'], group['flagged_parts'], group['violations'] = zip(*group['sequence'].apply(lambda seq: flag_sequences(seq) if isinstance(seq, str) else (True, [], 0)))\n",
    "        \n",
    "        # Check if all entries are flagged\n",
    "        if group['flag'].all():\n",
    "            print(f\"All sequences in group {group_value} are flagged.\")\n",
    "            group['pTM_rmsd_ratio'] = group[ptm_column] / group[rmsd_column]\n",
    "            min_violations = group['violations'].min()\n",
    "            group = group[group['violations'] == min_violations]  # Keep rows with the least violations\n",
    "            if not group.empty:\n",
    "                selected = group.loc[group['pTM_rmsd_ratio'].idxmax()]\n",
    "                flagged_part_str = \", \".join(selected['flagged_parts'])\n",
    "                selected['sequence_to_repair'] = flagged_part_str\n",
    "                final_df = pd.concat([final_df, pd.DataFrame([selected])], ignore_index=True)\n",
    "            continue\n",
    "\n",
    "        flagged_repetitive_sequences += group[group['flag']]['sequence'].tolist()\n",
    "        flagged_charged_sequences += group[group['flag']]['sequence'].tolist()\n",
    "        flagged_hydrophobic_sequences += group[group['flag']]['sequence'].tolist()\n",
    "        flagged_specific_sequences += group[group['flag']]['sequence'].tolist()\n",
    "        group = group[~group['flag']]\n",
    "\n",
    "        if group.empty:\n",
    "            print(f\"No valid sequences left after flagging in group {group_value}\")\n",
    "            print('')\n",
    "            continue\n",
    "\n",
    "        clusters = cluster_sequences(group, similarity_threshold)\n",
    "        if not clusters:\n",
    "            print(f\"No clusters formed in group {group_value}\")\n",
    "            print('')\n",
    "            continue\n",
    "\n",
    "        subgroup_counts[group_value] = len(clusters)\n",
    "        for cluster in clusters:\n",
    "            # SELECT BEST CLUSTER BASED ON LOWEST RMSD\n",
    "            selected = min(cluster['members'], key=lambda x: x[rmsd_column])\n",
    "            final_df = pd.concat([final_df, pd.DataFrame([selected])], ignore_index=True)\n",
    "\n",
    "    print(\"Final processing complete.\")\n",
    "    print(f\"Total sequences flagged for consecutive charged residues: {len(set(flagged_charged_sequences))}\")\n",
    "    print(f\"Total sequences flagged for repetitive amino acids: {len(set(flagged_repetitive_sequences))}\")\n",
    "    print(f\"Total sequences flagged for hydrophobic stretches: {len(set(flagged_hydrophobic_sequences))}\")\n",
    "    print(f\"Total sequences flagged for specific patterns: {len(set(flagged_specific_sequences))}\")\n",
    "    print(f\"Number of rows in the final DataFrame: {len(final_df)}\")\n",
    "    print(f\"Number of unique {grouping_column} after filtering: {final_df[grouping_column].nunique() if not final_df.empty else 0}\")\n",
    "    print(f\"Number of sequences with exact duplicates in the final DataFrame: {final_df.duplicated(subset=['sequence']).sum()}\")\n",
    "    print('')\n",
    "    \n",
    "    # Print the number of subgroups for each group\n",
    "    if subgroup_counts:\n",
    "        for group_value, count in subgroup_counts.items():\n",
    "            print(f\"{grouping_column} {group_value} has {count} files\")\n",
    "    else:\n",
    "        print(\"No data available in the final DataFrame after processing.\")\n",
    "    \n",
    "    return final_df\n",
    "\n",
    "# Print column names to verify the presence of grouping_column\n",
    "print(\"Columns in df_sorted:\", df_sorted.columns)\n",
    "\n",
    "# Check for the presence of grouping_column in the DataFrame\n",
    "if grouping_column not in df_sorted.columns:\n",
    "    raise ValueError(f\"DataFrame does not contain '{grouping_column}'. Please check your DataFrame columns.\")\n",
    "\n",
    "# Proceed with the rest of the script if the column check is fine\n",
    "final_df = process_dataframe(df_sorted, similarity_threshold)\n",
    "\n",
    "# Optionally, display the resulting DataFrame\n",
    "if not final_df.empty:\n",
    "    print(f\"Number of rows in the new DataFrame: {len(final_df)}\")\n",
    "    if grouping_column in final_df.columns:\n",
    "        print(f\"Number of unique {grouping_column} after filtering: {final_df[grouping_column].nunique()}\")\n",
    "    else:\n",
    "        print(\"No grouping column in final DataFrame; possibly all data was filtered out.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "deaf1013-429d-447e-ba62-1a72827e1c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv(f'{important_dfs_dir}RFD2_sequence_clustered_post_fastmpnn_af2_pre_placer.csv', index=False)\n",
    "# num1:f'{important_dfs_dir}RFD2_sequence_clustered_post_fastmpnn_af2_pre_placer.csv'\n",
    "# num2: f'{important_dfs_dir}inpainting_sequence_clustered_post_fastmpnn_af2_pre_placer.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60f6441d-867e-49cf-95c3-810ad9b8d5d1",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Copy AF2 + LIGAND into PLACER Input Directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "56cabdcc-055c-41ef-b054-04db2ed4f986",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of files to copy: 12\n",
      "All 12 files have been copied to ./output/placer_input_structures/RFD2/holo.\n",
      "Total number of files that were not found: 0\n"
     ]
    }
   ],
   "source": [
    "######################################################################################################################\n",
    "### COPY APO RELAXED AF2 STRUCTURES WITH LIGAND COORDS INTO FILE (WE NEED TO PULL ALIGNED APO STRUCTURES FOR THIS) ###\n",
    "######################################################################################################################\n",
    "\n",
    "align_af2_out_directory = os.path.join(f\"{af2_out_dir}\", \"predesigned_RFD2_i2\", \"filtered_structures/alignment/\")\n",
    "# num1: predesigned_RFD2_i2\n",
    "# num2: predesigned_inpainting_i2\n",
    "\n",
    "# Define the destination directory\n",
    "destination_dir = os.path.join(placer_input_structures_dir, \"RFD2\", \"holo\")\n",
    "# num1: RFD2\n",
    "# num2: inpainting\n",
    "os.makedirs(destination_dir, exist_ok=True)\n",
    "\n",
    "# Get the list of files to copy from the DataFrame\n",
    "files_to_copy = shell_filtered_megadata_df['pdb_path_af2_apo_relax_out'].tolist()\n",
    "total_files = len(files_to_copy)\n",
    "\n",
    "print(f\"Total number of files to copy: {total_files}\")\n",
    "\n",
    "# Initialize counters\n",
    "copied_files_count = 0\n",
    "missing_files = []\n",
    "\n",
    "# Copy the files and print updates after every 1000 files\n",
    "for i, file_path in enumerate(files_to_copy):\n",
    "    # Extract the basename and remove the .pdb extension\n",
    "    base_name = os.path.basename(file_path).replace('.pdb', '')\n",
    "    \n",
    "    # Search for the corresponding files in align_af2_out_directory\n",
    "    search_pattern = os.path.join(align_af2_out_directory, f\"{base_name}*.pdb\")\n",
    "    matched_files = glob.glob(search_pattern)\n",
    "    \n",
    "    if matched_files:\n",
    "        for matched_file in matched_files:\n",
    "            shutil.copy(matched_file, destination_dir)\n",
    "            copied_files_count += 1\n",
    "    else:\n",
    "        missing_files.append(search_pattern)\n",
    "    \n",
    "    if (i + 1) % 1000 == 0:\n",
    "        remaining_files = total_files - (i + 1)\n",
    "        print(f\"Copied {copied_files_count} files. {remaining_files} files remaining.\")\n",
    "\n",
    "# Final update after all files are copied\n",
    "print(f\"All {copied_files_count} files have been copied to {destination_dir}.\")\n",
    "print(f\"Total number of files that were not found: {len(missing_files)}\")\n",
    "if missing_files:\n",
    "    print(\"Files not found:\")\n",
    "    for missing in missing_files:\n",
    "        print(missing)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a38512c-a346-484b-a9d0-e7538e2108e0",
   "metadata": {
    "tags": []
   },
   "source": [
    "# **VI. PLACER**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7571f32-6fcc-4589-a7e4-01637d8908f5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Run PLACER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e79dc8e7-e7d0-43d4-b574-060699ca9155",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of files to generate substrate-binding structures: 12\n",
      "1 th PDB processed, 13 remaining. Elapsed time: 0.00 s. Estimated time left: 0.00 s.\n",
      "10 th PDB processed, 4 remaining. Elapsed time: 0.49 s. Estimated time left: 0.10 s.\n",
      "12 th PDB processed. Done.\n"
     ]
    }
   ],
   "source": [
    "###############################################\n",
    "### GENERATING SUBSTRATE-BINDING STRUCTURES ###\n",
    "###############################################\n",
    "\n",
    "input_pdb_dir = os.path.join(placer_input_structures_dir, \"RFD2\", \"holo\")\n",
    "#num1: f\"{placer_input_structures_dir}RFD2\"\n",
    "#num2: f\"{placer_input_structures_dir}inpainting\"\n",
    "\n",
    "subst_pdb_dir = f\"{placer_input_structures_dir}RFD2/subst_only/\"\n",
    "#num1: f\"{placer_input_structures_dir}RFD2/subst_only/\"\n",
    "#num1: f\"{placer_input_structures_dir}inpainting/subst_only/\"\n",
    "os.makedirs(subst_pdb_dir, exist_ok=True)\n",
    "\n",
    "subst_atom_list = [\"remove\", \"ZN1\", \"O1\", \"H1\"]\n",
    "\n",
    "script = '{scripts_dir}modify_atoms_in_pdb.py'\n",
    "\n",
    "input_list = glob.glob(os.path.join(input_pdb_dir, \"*.pdb\"))\n",
    "print(f\"Total number of files to generate substrate-binding structures: {len(input_list)}\")\n",
    "\n",
    "input_pdb_list = glob.glob(os.path.join(input_pdb_dir, \"*.pdb\"))\n",
    "start_time = time.time()\n",
    "for i, input_pdb in enumerate(input_pdb_list):\n",
    "    if np.log10(i+1)%1 == 0:        \n",
    "        elapsed_time = time.time() - start_time\n",
    "        time_per_row = elapsed_time / (i+1)\n",
    "        estimated_time_remaining = time_per_row * (len(input_pdb_list) - i-1)\n",
    "        print(f\"{i+1} th PDB processed, {len(input_pdb_list) - i+1} remaining. \"\n",
    "              f\"Elapsed time: {elapsed_time:.2f} s. Estimated time left: {estimated_time_remaining:.2f} s.\")\n",
    "    elif i+1 == len(input_pdb_list):\n",
    "        print (f\"{i+1} th PDB processed. Done.\")\n",
    "        \n",
    "    bn = os.path.basename(input_pdb)\n",
    "    subst_pdb_out = os.path.join(subst_pdb_dir, bn.replace(\".pdb\", \"_subst.pdb\"))\n",
    "    \n",
    "    for atom_list in [subst_atom_list]:\n",
    "        os.system(f\"python {script} --input {input_pdb} --output {subst_pdb_out} --atoms_to_remove {' '.join(atom_list[1:])}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b5dc24cb-c0fe-4774-9755-383287e67842",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run below commands:\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_0_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_2_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_8_T0.1_3_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.1_2_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_0_9_T0.2_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.1_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_0_T0.2_0_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.1_4_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n",
      "python {software_dir}PLACER/run_PLACER.py -f ./output/placer_input_structures/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.pdb -n 40 --odir ./output/placer_out/RFD2/subst_only/ --ocsv ./output/placer_out/RFD2/subst_only/zn_hydrolase_PSZ_H_H_H_1_63_0_1-atomized-bb-False_0_10_model_4_ptm_seed_0_unrelaxed_PSZ_1_1_T0.2_3_model_4_ptm_seed_0_unrelaxed_relaxedAPO_PSZ_subst.csv\n"
     ]
    }
   ],
   "source": [
    "########################################\n",
    "### SUBSTRATE-BINDING DESIGN PLACER ###\n",
    "########################################\n",
    "\n",
    "# IMPORTANT INPUTS \n",
    "placer_input_structures = f\"{placer_input_structures_dir}RFD2/subst_only/\"\n",
    "#num1: f\"{placer_input_structures_dir}RFD2/subst_only/\"\n",
    "#num1: f\"{placer_input_structures_dir}inpainting/subst_only/\"\n",
    "\n",
    "placer_out = f'{placer_out_dir}RFD2/subst_only/' # Update occassionally\n",
    "#num1: f'{placer_out_dir}RFD2/subst_only/'\n",
    "#num1: f'{placer_out_dir}inpainting/subst_only/'\n",
    "\n",
    "num_samples = 40\n",
    "\n",
    "# Mostly Stay constant\n",
    "os.makedirs(placer_out,exist_ok=True)\n",
    "script = '{software_dir}PLACER/run_PLACER.py'\n",
    "\n",
    "# Make PLACER Commands\n",
    "print (\"Run below commands:\")\n",
    "for pdb in glob.glob(f'{placer_input_structures}*pdb'): \n",
    "    bn = os.path.basename(pdb).replace('.pdb','')\n",
    "    cmd_txt = (f'python {script} ' #KEEP python\n",
    "               f'-f {pdb} '\n",
    "               f'-n {num_samples} '\n",
    "               f'--odir {placer_out} '\n",
    "               f'--ocsv {placer_out}{bn}.csv')\n",
    "    print (cmd_txt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d26932f7-4b5a-457e-aa99-4fa40f38e1f3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## PLACER Analysis Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a16ce638-281c-4e9d-8cb7-cbe7493f2029",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################################################\n",
    "### PLACER ANALYSIS FOR SUBSTRATE-BINDING DESIGN ###\n",
    "#####################################################\n",
    "\n",
    "script = '{scripts_dir}process_placer.py'\n",
    "\n",
    "multiple_ligand_names = {\"PSZ\"}\n",
    "lig_atom = \"C1 C2 C3 C4 C5 C6 C7 C8 C9 C10 C11 C12 C13 C14 C15 C16 C17 C18 O2 O3 O4 O5\"\n",
    "\n",
    "params_path_DIR = f\"{invrot_dir}COMBINED_PHENYLACETATE_HHH/COMBINED_PARAMS/\"\n",
    "params_path = glob.glob(os.path.join(params_path_DIR, \"*.params\"))\n",
    "\n",
    "placer_input_path = f\"{placer_input_structures_dir}inpainting/subst_only/\"\n",
    "#num1: f\"{placer_input_structures_dir}RFD2/subst_only/\"\n",
    "#num1: f\"{placer_input_structures_dir}inpainting/subst_only/\"\n",
    "\n",
    "placer_output_path = f'{placer_out_dir}inpainting/subst_only/' # Update occassionally\n",
    "#num1: f'{placer_out_dir}RFD2/subst_only/'\n",
    "#num1: f'{placer_out_dir}inpainting/subst_only/'\n",
    "\n",
    "for placer_output_pdb in glob.glob(os.path.join(placer_output_path, \"*.pdb\")):\n",
    "    bn = os.path.basename(placer_output_pdb)\n",
    "    placer_input = os.path.join(placer_input_path, bn.replace(\"_model.pdb\", \".pdb\"))\n",
    "    placer_output_csv = os.path.join(placer_output_path, bn.replace(\"_model.pdb\", \".csv\"))\n",
    "    score_output = os.path.join(placer_output_path, bn.replace(\"_model.pdb\", \"_updated_score.csv\"))\n",
    "\n",
    "    lig_name = list(filter(lambda lig: lig in bn, multiple_ligand_names))[0]\n",
    "\n",
    "    cmd_txt = (f'python {script} '\n",
    "           f'--pdb {placer_input} '\n",
    "           f'--scorefile {placer_output_csv} '\n",
    "           f'--lig_name {lig_name} '\n",
    "           f'--lig_atom {lig_atom} '    \n",
    "           f'--params {\" \".join(params_path)} '\n",
    "           f'--placer_pdb_path {placer_output_path} '\n",
    "           f'--scorefile_out {score_output} \\n')\n",
    "    print(cmd_txt)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1467f383-40dc-44e2-95a0-6e9e8493609b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## PLACER Analysis Substrate-binding Designs Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8493d08-3aa3-4b7e-b67d-3c116e38a002",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################\n",
    "### COMBINE THE ABOVE TWO DATA FRAMES AND SAVE ###\n",
    "##################################################\n",
    "\n",
    "# Define the paths\n",
    "placer_subst_input_path = f\"{placer_input_structures_dir}inpainting/subst_only/\"\n",
    "#num1: f\"{placer_input_structures_dir}RFD2/subst_only/\"\n",
    "#num2: f\"{placer_input_structures_dir}inpainting/subst_only/\"\n",
    "\n",
    "placer_subst_output_path = f'{placer_out_dir}inpainting/subst_only/' # Update occassionally\n",
    "#num1: f'{placer_out_dir}RFD2/subst_only/'\n",
    "#num2: f'{placer_out_dir}inpainting/subst_only/'\n",
    "\n",
    "file_paths = glob.glob(os.path.join(placer_subst_output_path, \"*updated_score.csv\"))\n",
    "print(\"Number of found updated score files:\", len(file_paths))\n",
    "\n",
    "df_list = []\n",
    "file_names = []\n",
    "for fi in file_paths:\n",
    "    df = pd.read_csv(fi, sep=\",\")\n",
    "    if not df.empty:\n",
    "        df_list.append(df)\n",
    "        file_names.append(os.path.basename(fi))\n",
    "placer_subst_df = pd.concat(df_list, ignore_index=True)\n",
    "print(\"Size of placer_subst_df\", len(placer_subst_df))\n",
    "\n",
    "# Remove unnamed columns if they exist\n",
    "placer_subst_df = placer_subst_df.loc[:, ~placer_subst_df.columns.str.contains('^Unnamed')]\n",
    "\n",
    "# Add the 'name' column\n",
    "placer_subst_df[\"name\"] = placer_subst_df[\"description\"].str[:-6]\n",
    "\n",
    "# Create new columns\n",
    "subst_top_pdb_paths = [os.path.join(placer_subst_output_path, str(el).replace(\"_subst\", \"_subst_model.pdb\")) for el in placer_subst_df['description'].tolist()]\n",
    "pdb_path_subst_placer = subst_top_pdb_paths\n",
    "pdb_path_subst_placer_INPUT = [os.path.join(placer_subst_input_path, os.path.basename(path).replace(\"_model.pdb\", \".pdb\")) for path in subst_top_pdb_paths]\n",
    "\n",
    "# Add new columns to merged_df\n",
    "placer_subst_df[\"pdb_path_subst_placer\"] = pdb_path_subst_placer\n",
    "placer_subst_df[\"pdb_path_subst_placer_INPUT\"] = pdb_path_subst_placer_INPUT\n",
    "\n",
    "# Define the save path\n",
    "save_path = f'{important_dfs_dir}inpainting_placer_subst_combined_df.csv'\n",
    "#num1: RFD2\n",
    "#num2: inpainting\n",
    "\n",
    "# Save the new dataframe\n",
    "placer_subst_df.to_csv(save_path, index=False)\n",
    "\n",
    "# Print the save location\n",
    "print(f\"Dataframe saved to: {save_path}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyrosetta",
   "language": "python",
   "name": "pyrosetta"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
